<!DOCTYPE html>
<head>
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>16. R Markdown | systemPipeR Website</title>
<link rel="stylesheet" href="css/syntax.css">


<link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css">-->
<link rel="stylesheet" href="css/modern-business.css">
<link rel="stylesheet" href="css/lavish-bootstrap.css">
<link rel="stylesheet" href="css/customstyles.css">
<link rel="stylesheet" href="css/theme-blue.css">

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="js/jquery.navgoco.min.js"></script>


<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="js/toc.js"></script>
<script src="js/customscripts.js"></script>

<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

<link rel="shortcut icon" href="images/favicon.ico">

<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->

<link rel="alternate" type="application/rss+xml" title="" href="http://localhost:4000feed.xml">

    <script>
        $(document).ready(function() {
            // Initialize navgoco with default options
            $("#mysidebar").navgoco({
                caretHtml: '',
                accordion: true,
                openClass: 'active', // open
                save: false, // leave false or nav highlighting doesn't work right
                cookie: {
                    name: 'navgoco',
                    expires: false,
                    path: '/'
                },
                slide: {
                    duration: 400,
                    easing: 'swing'
                }
            });

            $("#collapseAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', false);
            });

            $("#expandAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', true);
            });

        });

    </script>
    <script>
        $(function () {
            $('[data-toggle="tooltip"]').tooltip()
        })
    </script>
    <script>
        $(document).ready(function() {
            $("#tg-sb-link").click(function() {
                $("#tg-sb-sidebar").toggle();
                $("#tg-sb-content").toggleClass('col-md-9');
                $("#tg-sb-content").toggleClass('col-md-12');
                $("#tg-sb-icon").toggleClass('fa-toggle-on');
                $("#tg-sb-icon").toggleClass('fa-toggle-off');
            });
        });
    </script>
    

</head>
<body>
<!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="fa fa-home fa-lg navbar-brand" href="index.html">&nbsp;<span class="projectTitle"> systemPipeR Project Website</span></a>
        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- toggle sidebar button -->
                <li><a id="tg-sb-link" href="#"><i id="tg-sb-icon" class="fa fa-toggle-on"></i> Nav</a></li>
                <!-- entries without drop-downs appear here -->
                
                
                
                <li><a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html" target="_blank">Bioc</a></li>
                
                
                
                <li><a href="https://github.com/tgirke/systemPipeR" target="_blank">GitHub Repo</a></li>
                
                
                
                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        <li><a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual" target="_blank">R & Bioconductor Manual</a></li>
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        <li><a href="http://hpcc.ucr.edu/" target="_blank">HPC Center (HPCC)</a></li>
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                    </ul>
                </li>
                
                
                
			<li>



  <a class="email" title="Submit feedback" href="#" onclick="javascript:window.location='mailto:thomas.girke@ucr.edu?subject=systemPipeR Website feedback&body=I have some feedback about the 16. R Markdown page: ' + window.location.href;"><i class="fa fa-envelope-o"></i> Feedback</a>

</li>

		
                <!--comment out this block if you want to hide search-->
                <li>
                    <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                            SimpleJekyllSearch.init({
                                searchInput: document.getElementById('search-input'),
                                resultsContainer: document.getElementById('results-container'),
                                dataSource: 'search.json',
                                searchResultTemplate: '<li><a href="{url}" title="16. R Markdown">{title}</a></li>',
                    noResultsText: 'No results found.',
                            limit: 10,
                            fuzzy: true,
                    })
                    </script>
                    <!--end search-->
                </li>
            </ul>
        </div>
        </div>
        <!-- /.container -->
</nav>

<!-- Page Content -->
<div class="container">
    <div class="col-lg-12">&nbsp;</div>
    <!-- Content Row -->
    <div class="row">
        <!-- Sidebar Column -->
        <div class="col-md-3" id="tg-sb-sidebar">

          












<ul id="mysidebar" class="nav">
    <li class="sidebarTitle"><center><font color="#26466D" size="6">systemPipeR</font></center> <center><font color="#8b9299" size="3">...</font></center></li>
    
    
    
        
    
    <li>
        <a href="#">Home</a>
        <ul>
            
            
            
            <li><a href="index.html">Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_syllabus.html">Syllabus</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Internal</a>
        <ul>
            
            
            
            <li><a href="mydoc_internal.html">Internal Resources</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Schedule</a>
        <ul>
            
            
            
            <li><a href="mydoc_schedule.html">Course Schedule</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Lecture Slides</a>
        <ul>
            
            
            
            <li><a href="mydoc_slides_01.html">Course Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_02.html">Molbiol Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_03.html">Databases and Software</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_04.html">DNA Sequencing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_05.html">Introduction to R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_06.html">Sequence Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_07.html">Programming in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_08.html">Multiple Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_09.html">Short Read Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_10.html">NGS Analysis Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_11.html">Gene Expression Analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_12.html">NGS Workflows</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_13.html">ChIP-Seq Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_14.html">VAR-Seq Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_15.html">Sequence Assembly</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_16.html">Gene Annotation and Ontologies</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_17.html">Cluster Analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_18.html">Profile HMMs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_19.html">Introduction to Phylogenetics</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Homework</a>
        <ul>
            
            
            
            <li><a href="mydoc_homework_01.html">HW1 - Online Tools</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_02.html">HW2 - Biocluster/Linux</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_03.html">HW3 - Introduction to R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_04.html">HW4 - Pairwise Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_09.html">HW5 - Programming in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_08.html">HW6 - Sequence Analysis</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">GitHub</a>
        <ul>
            
            
            
            <li><a href="mydoc_tutorial_01.html">Table of Content</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#github-in-gen242">GitHub in GEN242</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#installing-git">Installation</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#git-basics-from-command-line">Git from Command-line</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#github-basics-from-command-line">GitHub from Command-line</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#using-github-from-rstudio">GitHub from RStudio</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Linux & HPCC Cluster</a>
        <ul>
            
            
            
            <li><a href="mydoc_tutorial_02.html">Table of Content</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#hpcc-cluster-overview">HPCC Cluster</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#linux-basics">Linux Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#software-and-module-system">Software & Module System</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#big-data-storage">Big Data Storage</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#queuing-system-slurm">Queueing System Slurm</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#textcode-editors">Text Editors</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#vim-r-tmux-essentials">Vim-R-Tmux</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#nvim-r-tmux">Nvim-R-Tmux</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Introduction to R</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rbasics_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_02.html">2. R Package Repositories</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_03.html">3. Installation of R Packages</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_04.html">4. Getting Around</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_05.html">5. Basic Syntax</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_06.html">6. Data Types</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_07.html">7. Data Objects</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_08.html">8. Important Utilities</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_09.html">9. Operators and Calculations</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_10.html">10. Reading and Writing External Data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_11.html">11. Useful R Functions</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_12.html">12. dplyr Environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_13.html">13. SQLite Databases</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_14.html">14. Graphics in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_15.html">15. Analysis Routine</a></li>
            
            
            
            
            
            
            <li class="active"><a href="mydoc_Rbasics_16.html">16. R Markdown</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_17.html">17. Shiny Web Apps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_18.html">18. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_19.html">19. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Programming in R</a>
        <ul>
            
            
            
            <li><a href="mydoc_Programming_in_R_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_02.html">2. Control Structures</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_03.html">3. Loops</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_04.html">4. Functions</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_05.html">5. Useful Utilities</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_06.html">6. Running R Scripts</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_07.html">7. Building R Packages</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_08.html">8. Programming Exercises</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_09.html">9. Homework 5</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_10.html">10. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Analysis Basics</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rsequences_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_02.html">2. Package Requirements</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_03.html">3. Strings in R Base</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_04.html">4. Sequences in Bioconductor</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_05.html">5. NGS Sequences</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_06.html">6. Range Operations</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_07.html">7. Transcript Ranges</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_08.html">8. Homework 6</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_09.html">9. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_10.html">10. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Workflow Overview</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeR_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_2.html">2. Getting Started</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_3.html">3. Workflow overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_4.html">4. Workflow templates</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_5.html">5. Version information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">RNA-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_02.html">2. Samples and environment settings</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_05.html">5. Read quantification</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_06.html">6. Analysis of DEGs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_07.html">7. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_08.html">8. Clustering and heat maps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_09.html">9. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_10.html">10. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">ChIP-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_05.html">5. Utilities for coverage data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_06.html">6. Peak calling with MACS2</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_07.html">7. Annotate peaks with genomic context</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_08.html">8. Count reads overlapping peaks</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_09.html">9. Differential binding analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_10.html">10. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_11.html">11. Motif analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">VAR-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeVARseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_05.html">5. Variant calling</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_06.html">6. Filter variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_07.html">7. Annotate filtered variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_08.html">8. Combine annotation results among samples</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_09.html">9. Summary statistics of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_10.html">10. Venn diagram of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_11.html">11. Plot variants programmatically</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Graphics and Visualization</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rgraphics_1.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_2.html">2. Base Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_3.html">3. Grid Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_4.html">4. lattice Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_5.html">5. ggplot2 Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_6.html">6. Specialty Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_7.html">7. Genome Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_8.html">8. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Cluster Analysis</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rclustering_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_2.html">2. Data Preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_3.html">3. Clustering Algorithms</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_4.html">4. Clustering Exercises</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_5.html">5. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Paper Presentations</a>
        <ul>
            
            
            
            <li><a href="mydoc_paper_presentations.html">Paper Presentations</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Course Projects</a>
        <ul>
            
            
            
            <li><a href="mydoc_project_intro.html">Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_chipseq1.html">ChIP-Seq1 - Motif Prediction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_chipseq2.html">ChIP-Seq2 - Peak Callers</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_rnaseq1.html">RNA-Seq1 - Aligners</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_rnaseq2.html">RNA-Seq2 - DEG Methods</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_varseq1.html">VAR-Seq1 - Functional Prediction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_data.html">Project Data</a></li>
            
            
            
            
        </ul>
        
        
        
        <!-- if you aren't using the accordion, uncomment this block:
           <p class="external">
               <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
           </p>
           -->
    </li>
</ul>
</div>

<!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
<script>$("li.active").parents('li').toggleClass("active");</script>

    <!-- Content Column -->
    <div class="col-md-9" id="tg-sb-content">
        <div class="post-header">
   <h1 class="post-title-main">16. R Markdown</h1>
</div>



<div class="post-content">

   

    
    
<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>

<div id="toc"></div>

    


<!--
    

    <a target="_blank" href="https://github.com/tgirke/systemPipeR/blob/gh-pages/pages//mydoc_Rbasics_16.md" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
-->
    

  <h2 id="overview">Overview</h2>

<p>R Markdown combines markdown (an easy to write plain text format) with embedded
R code chunks. When compiling R Markdown documents, the code components can be
evaluated so that both the code and its output can be included in the final
document. This makes analysis reports highly reproducible by allowing to automatically
regenerate them when the underlying R code or data changes. R Markdown
documents (<code class="highlighter-rouge">.Rmd</code> files) can be rendered to various formats including HTML and
PDF. The R code in an <code class="highlighter-rouge">.Rmd</code> document is processed by <code class="highlighter-rouge">knitr</code>, while the
resulting <code class="highlighter-rouge">.md</code> file is rendered by <code class="highlighter-rouge">pandoc</code> to the final output formats
(<em>e.g.</em> HTML or PDF). Historically, R Markdown is an extension of the older
<code class="highlighter-rouge">Sweave/Latex</code> environment. Rendering of mathematical expressions and reference
management is also supported by R Markdown using embedded Latex syntax and
Bibtex, respectively.</p>

<h2 id="quick-start">Quick Start</h2>

<h3 id="install-r-markdown">Install R Markdown</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">install.packages</span><span class="p">(</span><span class="s2">"rmarkdown"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<h3 id="initialize-a-new-r-markdown-rmd-script">Initialize a new R Markdown (<code class="highlighter-rouge">Rmd</code>) script</h3>

<p>To minimize typing, it can be helful to start with an R Markdown template and
then modify it as needed. Note the file name of an R Markdown scirpt needs to
have the extension <code class="highlighter-rouge">.Rmd</code>. Template files for the following examples are available 
here:</p>

<ul>
  <li>R Markdown sample script: <a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/_vignettes/07_Rbasics/sample.Rmd"><code class="highlighter-rouge">sample.Rmd</code></a></li>
  <li>Bibtex file for handling citations and reference section: <a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/_vignettes/07_Rbasics/bibtex.bib"><code class="highlighter-rouge">bibtex.bib</code></a></li>
</ul>

<p>Users want to download these files, open the <code class="highlighter-rouge">sample.Rmd</code> file with their preferred R IDE 
(<em>e.g.</em> RStudio, vim or emacs), initilize an R session and then direct their R session to 
the location of these two files.</p>

<h3 id="metadata-section">Metadata section</h3>

<p>The metadata section (YAML header) in an R Markdown script defines how it will be processed and 
rendered. The metadata section also includes both title, author, and date information as well as 
options for customizing the output format. For instance, PDF and HTML output can be defined 
with <code class="highlighter-rouge">pdf_document</code> and <code class="highlighter-rouge">html_document</code>, respectively. The <code class="highlighter-rouge">BiocStyle::</code> prefix will use the
formatting style of the <a href="http://bioconductor.org/packages/release/bioc/html/BiocStyle.html"><code class="highlighter-rouge">BiocStyle</code></a> 
package from Bioconductor.</p>

<div class="highlighter-rouge"><pre class="highlight"><code> ---
title: "My First R Markdown Document"
author: "Author: First Last"
date: "Last update: 25 June, 2017"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    fig_caption: yes

fontsize: 14pt
bibliography: bibtex.bib
 ---
</code></pre>
</div>

<h3 id="render-rmd-script">Render <code class="highlighter-rouge">Rmd</code> script</h3>

<p>An R Markdown script can be evaluated and rendered with the following <code class="highlighter-rouge">render</code> command or by pressing the <code class="highlighter-rouge">knit</code> button in RStudio.
The <code class="highlighter-rouge">output_format</code> argument defines the format of the output (<em>e.g.</em> <code class="highlighter-rouge">html_document</code>). The setting <code class="highlighter-rouge">output_format="all"</code> will generate 
all supported output formats. Alternatively, one can specify several output formats in the metadata section as shown in the above example.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">rmarkdown</span><span class="o">::</span><span class="n">render</span><span class="p">(</span><span class="s2">"sample.Rmd"</span><span class="p">,</span><span class="w"> </span><span class="n">clean</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">output_format</span><span class="o">=</span><span class="s2">"html_document"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p>The following shows two options how to run the rendering from the command-line.</p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="gp">$ </span><span class="nb">echo</span> <span class="s2">"rmarkdown::render('sample.Rmd', clean=TRUE)"</span> | R --slave
<span class="gp">$ </span>Rscript -e <span class="s2">"rmarkdown::render('sample.Rmd', clean=TRUE)"</span>
</code></pre>
</div>

<p>Alternatively, one can use a Makefile to evaluate and render an R Markdown
script. A sample Makefile for rendering the above <code class="highlighter-rouge">sample.Rmd</code> can be
downloaded <a href="https://raw.githubusercontent.com/tgirke/GEN242/master/vignettes/07_Rbasics/Makefile"><code class="highlighter-rouge">here</code></a>.
To apply it to a custom <code class="highlighter-rouge">Rmd</code> file, one needs open the Makefile in a text
editor and change the value assigned to <code class="highlighter-rouge">MAIN</code> (line 13) to the base name of
the corresponding <code class="highlighter-rouge">.Rmd</code> file (<em>e.g.</em> assign <code class="highlighter-rouge">systemPipeRNAseq</code> if the file
name is <code class="highlighter-rouge">systemPipeRNAseq.Rmd</code>).  To execute the <code class="highlighter-rouge">Makefile</code>, run the following
command from the command-line.</p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="gp">$ </span>make -B
</code></pre>
</div>

<h3 id="r-code-chunks">R code chunks</h3>

<p>R Code Chunks can be embedded in an R Markdown script by using three backticks
at the beginning of a new line along with arguments enclosed in curly braces
controlling the behavior of the code. The following lines contain the
plain R code. A code chunk is terminated by a new line starting with three backticks.
The following shows an example of such a code chunk. Note the backslashes are
not part of it. They have been added to print the code chunk syntax in this document.</p>

<div class="highlighter-rouge"><pre class="highlight"><code>	```\{r code_chunk_name, eval=FALSE\}
	x &lt;- 1:10
	```
</code></pre>
</div>

<p>The following lists the most important arguments to control the behavior of R code chunks:</p>

<ul>
  <li><code class="highlighter-rouge">r</code>: specifies language for code chunk, here R</li>
  <li><code class="highlighter-rouge">chode_chunk_name</code>: name of code chunk; this name needs to be unique</li>
  <li><code class="highlighter-rouge">eval</code>: if assigned <code class="highlighter-rouge">TRUE</code> the code will be evaluated</li>
  <li><code class="highlighter-rouge">warning</code>: if assigned <code class="highlighter-rouge">FALSE</code> warnings will not be shown</li>
  <li><code class="highlighter-rouge">message</code>: if assigned <code class="highlighter-rouge">FALSE</code> messages will not be shown</li>
  <li><code class="highlighter-rouge">cache</code>: if assigned <code class="highlighter-rouge">TRUE</code> results will be cached to reuse in future rendering instances</li>
  <li><code class="highlighter-rouge">fig.height</code>: allows to specify height of figures in inches</li>
  <li><code class="highlighter-rouge">fig.width</code>: allows to specify width of figures in inches</li>
</ul>

<p>For more details on code chunk options see <a href="https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf">here</a>.</p>

<h3 id="learning-markdown">Learning Markdown</h3>

<p>The basic syntax of Markdown and derivatives like kramdown is extremely easy to learn. Rather
than providing another introduction on this topic, here are some useful sites for learning Markdown:</p>

<ul>
  <li><a href="https://guides.github.com/features/mastering-markdown/">Markdown Intro on GitHub</a></li>
  <li><a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet">Markdown Cheet Sheet</a></li>
  <li><a href="http://rmarkdown.rstudio.com/authoring_basics.html">Markdown Basics from RStudio</a></li>
  <li><a href="http://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf">R Markdown Cheat Sheet</a></li>
  <li><a href="http://kramdown.gettalong.org/syntax.html">kramdown Syntax</a></li>
</ul>

<h3 id="tables">Tables</h3>

<p>There are several ways to render tables. First, they can be printed within the R code chunks. Second, 
much nicer formatted tables can be generated with the functions <code class="highlighter-rouge">kable</code>, <code class="highlighter-rouge">pander</code> or <code class="highlighter-rouge">xtable</code>. The following
example uses <code class="highlighter-rouge">kable</code> from the <code class="highlighter-rouge">knitr</code> package.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">knitr</span><span class="p">)</span><span class="w">
</span><span class="n">kable</span><span class="p">(</span><span class="n">iris</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">12</span><span class="p">,])</span><span class="w">
</span></code></pre>
</div>

<h2 id="sepallength---sepalwidth---petallength---petalwidth--species">Sepal.Length   Sepal.Width   Petal.Length   Petal.Width  Species</h2>
<div class="highlighter-rouge"><pre class="highlight"><code>      5.1           3.5            1.4           0.2  setosa  
      4.9           3.0            1.4           0.2  setosa  
      4.7           3.2            1.3           0.2  setosa  
      4.6           3.1            1.5           0.2  setosa  
      5.0           3.6            1.4           0.2  setosa  
      5.4           3.9            1.7           0.4  setosa  
      4.6           3.4            1.4           0.3  setosa  
      5.0           3.4            1.5           0.2  setosa  
      4.4           2.9            1.4           0.2  setosa  
      4.9           3.1            1.5           0.1  setosa  
      5.4           3.7            1.5           0.2  setosa  
      4.8           3.4            1.6           0.2  setosa  
</code></pre>
</div>

<h3 id="figures">Figures</h3>

<p>Plots generated by the R code chunks in an R Markdown document can be automatically 
inserted in the output file. The size of the figure can be controlled with the <code class="highlighter-rouge">fig.height</code>
and <code class="highlighter-rouge">fig.width</code> arguments.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span><span class="w">
</span><span class="n">dsmall</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">diamonds</span><span class="p">[</span><span class="n">sample</span><span class="p">(</span><span class="n">nrow</span><span class="p">(</span><span class="n">diamonds</span><span class="p">),</span><span class="w"> </span><span class="m">1000</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">ggplot</span><span class="p">(</span><span class="n">dsmall</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="p">,</span><span class="w"> </span><span class="n">price</span><span class="o">/</span><span class="n">carat</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_jitter</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">I</span><span class="p">(</span><span class="m">1</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<p><img src="./pages/mydoc/Rbasics_files/some_jitter_plot-1.png" width="672" /></p>

<p>Sometimes it can be useful to explicitly write an image to a file and then insert that 
image into the final document by referencing its file name in the R Markdown source. For 
instance, this can be useful for time consuming analyses. The following code will generate a 
file named <code class="highlighter-rouge">myplot.png</code>. To insert the file  in the final document, one can use standard 
Markdown or HTML syntax, <em>e.g.</em>: <code class="highlighter-rouge">&lt;img src="./pages/mydoc/Rbasics_files/myplot.png"/&gt;</code>.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">png</span><span class="p">(</span><span class="s2">"myplot.png"</span><span class="p">)</span><span class="w">
</span><span class="n">ggplot</span><span class="p">(</span><span class="n">dsmall</span><span class="p">,</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="p">,</span><span class="w"> </span><span class="n">price</span><span class="o">/</span><span class="n">carat</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">geom_jitter</span><span class="p">(</span><span class="n">alpha</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">I</span><span class="p">(</span><span class="m">1</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="m">2</span><span class="p">),</span><span class="w"> </span><span class="n">aes</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">))</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## png 
##   2
</code></pre>
</div>
<center><img title="some_title" src="./pages/mydoc/Rbasics_files/myplot.png" /></center>

<h3 id="inline-r-code">Inline R code</h3>

<p>To evaluate R code inline, one can enclose an R expression with a single back-tick
followed by <code class="highlighter-rouge">r</code> and then the actual expression.  For instance, the back-ticked version 
of r 1 + 1 evaluates to 2 and r pi evaluates to 3.1415927.</p>

<h3 id="mathematical-equations">Mathematical equations</h3>

<p>To render mathematical equations, one can use standard Latex syntax. When expressions are 
enclosed with single <code class="highlighter-rouge">$</code> signs then they will be shown inline, while 
enclosing them with double <code class="highlighter-rouge">$$</code> signs will show them in display mode. For instance, the following 
Latex syntax <code class="highlighter-rouge">d(X,Y) = \sqrt[]{ \sum_{i=1}^{n}{(x_{i}-y_{i})^2} }</code> renders in display mode as follows:</p>

<script type="math/tex; mode=display">d(X,Y) = \sqrt[]{ \sum_{i=1}^{n}{(x_{i}-y_{i})^2} }</script>

<h3 id="citations-and-bibliographies">Citations and bibliographies</h3>

<p>Citations and bibliographies can be autogenerated in R Markdown in a similar
way as in Latex/Bibtex. Reference collections should be stored in a separate
file in Bibtex or other supported formats. To cite a publication in an R Markdown 
script, one uses the syntax <code class="highlighter-rouge">(@&lt;id1&gt;)</code> where <code class="highlighter-rouge">&lt;id1&gt;</code> needs to be replaced with a 
reference identifier present in the Bibtex database listed in the metadata section 
of the R Markdown script  (<em>e.g.</em> <code class="highlighter-rouge">bibtex.bib</code>). For instance, to cite Lawrence et al. 
(2013), one  uses its reference identifier (<em>e.g.</em> <code class="highlighter-rouge">Lawrence2013-kt</code>) as <code class="highlighter-rouge">&lt;id1&gt;</code> (Lawrence et al., 2013). 
This will place the citation inline in the text and add the corresponding
reference to a reference list at the end of the output document. For the latter a 
special section called <code class="highlighter-rouge">References</code> needs to be specified at the end of the R Markdown script.
To fine control the formatting of citations and reference lists, users want to consult this 
the corresponding <a href="http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html">R Markdown page</a>.
Also, for general reference management and outputting references in Bibtex format <a href="https://paperpile.com/features">Paperpile</a> 
can be very helpful.</p>

<h3 id="viewing-r-markdown-report-on-hpcc-cluster">Viewing R Markdown report on HPCC cluster</h3>

<p>R Markdown reports located on UCRs HPCC Cluster can be viewed locally in a web browser (without moving 
the source HTML) by creating a symbolic link from a users <code class="highlighter-rouge">.html</code> directory. This way any updates to 
the report can be viewed immediately without creating another copy of the HTML file. For instance, if user 
<code class="highlighter-rouge">ttest</code> has generated an R Markdown report under <code class="highlighter-rouge">~/bigdata/today/rmarkdown/sample.html</code>, then the 
proper symbolic link to this file can be created as follows:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">cd</span><span class="w"> </span><span class="o">~/</span><span class="n">.html</span><span class="w">
</span><span class="n">ln</span><span class="w"> </span><span class="o">-</span><span class="n">s</span><span class="w"> </span><span class="o">~/</span><span class="n">bigdata</span><span class="o">/</span><span class="n">today</span><span class="o">/</span><span class="n">rmarkdown</span><span class="o">/</span><span class="n">sample.html</span><span class="w"> </span><span class="n">sample.html</span><span class="w">
</span></code></pre>
</div>

<p>After this one can view the report in a web browser using this URL <a href="http://biocluster.ucr.edu/~ttest/rmarkdown/sample.html">http://biocluster.ucr.edu/~ttest/rmarkdown/sample.html</a>.
If necessary access to the URL can be restricted with a password following the instructions <a href="http://hpcc.ucr.edu/manuals_linux-cluster_sharing.html#sharing-files-on-the-web">here</a>.</p>

<p><br /><br /><center><a href="mydoc_Rbasics_15.html"><img src="images/left_arrow.png" alt="Previous page." /></a>Previous Page &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Next Page
<a href="mydoc_Rbasics_17.html"><img src="images/right_arrow.png" alt="Next page." /></a></center></p>


    <div class="tags">
        
    </div>

    

</div>

<hr class="shaded"/>

<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2017 UC Riverside. All rights reserved. <br />
<span>Page last updated:</span> Sun Jun 25 17:41:27 2017<br/> Site last generated: Oct 8, 2017 <br />
<!--
<p><img src="images/company_logo.png" alt="Company logo"/></p>
-->
</div>
            </div>
</footer>


    </div>
    <!-- /.row -->
</div>
<!-- /.container -->
    </div>

</body>

</html>
