<!DOCTYPE html>
<head>
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>RIBO-Seq Workflow <br> <br> 1. Introduction | systemPipeR Website</title>
<link rel="stylesheet" href="css/syntax.css">


<link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css">-->
<link rel="stylesheet" href="css/modern-business.css">
<link rel="stylesheet" href="css/lavish-bootstrap.css">
<link rel="stylesheet" href="css/customstyles.css">
<link rel="stylesheet" href="css/theme-red.css">

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="js/jquery.navgoco.min.js"></script>


<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="js/toc.js"></script>
<script src="js/customscripts.js"></script>

<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

<link rel="shortcut icon" href="images/favicon.ico">

<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->

<link rel="alternate" type="application/rss+xml" title="systemPipeR" href="http://localhost:4000feed.xml">

    <script>
        $(document).ready(function() {
            // Initialize navgoco with default options
            $("#mysidebar").navgoco({
                caretHtml: '',
                accordion: true,
                openClass: 'active', // open
                save: false, // leave false or nav highlighting doesn't work right
                cookie: {
                    name: 'navgoco',
                    expires: false,
                    path: '/'
                },
                slide: {
                    duration: 400,
                    easing: 'swing'
                }
            });

            $("#collapseAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', false);
            });

            $("#expandAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', true);
            });

        });

    </script>
    <script>
        $(function () {
            $('[data-toggle="tooltip"]').tooltip()
        })
    </script>
    <script>
        $(document).ready(function() {
            $("#tg-sb-link").click(function() {
                $("#tg-sb-sidebar").toggle();
                $("#tg-sb-content").toggleClass('col-md-9');
                $("#tg-sb-content").toggleClass('col-md-12');
                $("#tg-sb-icon").toggleClass('fa-toggle-on');
                $("#tg-sb-icon").toggleClass('fa-toggle-off');
            });
        });
    </script>
    

</head>
<body>
<!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="fa fa-home fa-lg navbar-brand" href="index.html">&nbsp;<span class="projectTitle"> Data Analysis Workflow Environment</span></a>
        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- toggle sidebar button -->
                <li><a id="tg-sb-link" href="#"><i id="tg-sb-icon" class="fa fa-toggle-on"></i> Nav</a></li>
                <!-- entries without drop-downs appear here -->
                
                
                
                <li><a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html" target="_blank">Bioc</a></li>
                
                
                
                <li><a href="https://github.com/tgirke/systemPipeR" target="_blank">GitHub Repo</a></li>
                
                
                
                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        <li><a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual" target="_blank">R & Bioconductor Manual</a></li>
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        <li><a href="http://hpcc.ucr.edu/" target="_blank">HPC Center (HPCC)</a></li>
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                    </ul>
                </li>
                
                
                
			<li>



  <a class="email" title="Submit feedback" href="#" onclick="javascript:window.location='mailto:thomas.girke@ucr.edu?subject=systemPipeR Website feedback&body=I have some feedback about the RIBO-Seq Workflow <br> <br> 1. Introduction page: ' + window.location.href;"><i class="fa fa-envelope-o"></i> Feedback</a>

</li>

		
                <!--comment out this block if you want to hide search-->
                <li>
                    <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                            SimpleJekyllSearch.init({
                                searchInput: document.getElementById('search-input'),
                                resultsContainer: document.getElementById('results-container'),
                                dataSource: 'search.json',
                                searchResultTemplate: '<li><a href="{url}" title="RIBO-Seq Workflow <br> <br> 1. Introduction">{title}</a></li>',
                    noResultsText: 'No results found.',
                            limit: 10,
                            fuzzy: true,
                    })
                    </script>
                    <!--end search-->
                </li>
            </ul>
        </div>
        </div>
        <!-- /.container -->
</nav>

<!-- Page Content -->
<div class="container">
    <div class="col-lg-12">&nbsp;</div>
    <!-- Content Row -->
    <div class="row">
        <!-- Sidebar Column -->
        <div class="col-md-3" id="tg-sb-sidebar">

          












<ul id="mysidebar" class="nav">
    <li class="sidebarTitle"><center><font color="#7a1f1f" size="6"><i>systemPipeR</i></font></center> <center><font color="#8b9299" size="3">Project Site</font></center></li>
    
    
    
        
    
    <li>
        <a href="#">Home</a>
        <ul>
            
            
            
            <li><a href="index.html">Introduction</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Workflow Overview</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeR_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_2.html">2. Getting Started</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_3.html">3. Workflow overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_4.html">4. Workflow templates</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_5.html">5. Version information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">RNA-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_02.html">2. Samples and environment settings</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_05.html">5. Read quantification</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_06.html">6. Analysis of DEGs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_07.html">7. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_08.html">8. Clustering and heat maps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_09.html">9. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_10.html">10. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">ChIP-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_05.html">5. Utilities for coverage data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_06.html">6. Peak calling with MACS2</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_07.html">7. Annotate peaks with genomic context</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_08.html">8. Count reads overlapping peaks</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_09.html">9. Differential binding analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_10.html">10. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_11.html">11. Motif analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">VAR-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeVARseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_05.html">5. Variant calling</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_06.html">6. Filter variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_07.html">7. Annotate filtered variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_08.html">8. Combine annotation results among samples</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_09.html">9. Summary statistics of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_10.html">10. Venn diagram of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_11.html">11. Plot variants programmatically</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">RIBO-Seq Workflow</a>
        <ul>
            
            
            
            <li class="active"><a href="mydoc_systemPipeRIBOseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_02.html">2. Load workflow template</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_05.html">5. Read distribution across genomic features</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_06.html">6. Adding custom features to workflow</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_07.html">7. Genomic read coverage along transripts or CDSs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_08.html">8. Read quantification per annotation range</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_09.html">9. Analysis of differentially expressed genes with `edgeR`</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_10.html">10. GO term enrichment analysis of DEGs}</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_11.html">11. Differential ribosome loading analysis (translational efficiency)</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_12.html">12. Clustering and heat maps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_13.html">13. Render report in HTML and PDF format</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_14.html">14. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_15.html">15. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_16.html">16. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Training Events</a>
        <ul>
            
            
            
            <li><a href="mydoc_training.html">Training Events</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Publications</a>
        <ul>
            
            
            
            <li><a href="mydoc_publications.html">Publications</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Acknowledgement</a>
        <ul>
            
            
            
            <li><a href="mydoc_acknowledgement.html">Acknowledgement</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Contacts</a>
        <ul>
            
            
            
            <li><a href="mydoc_contacts.html">Contacts</a></li>
            
            
            
            
        </ul>
        
        
        
        <!-- if you aren't using the accordion, uncomment this block:
           <p class="external">
               <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
           </p>
           -->
    </li>
</ul>
</div>

<!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
<script>$("li.active").parents('li').toggleClass("active");</script>

    <!-- Content Column -->
    <div class="col-md-9" id="tg-sb-content">
        <div class="post-header">
   <h1 class="post-title-main">RIBO-Seq Workflow <br> <br> 1. Introduction</h1>
</div>



<div class="post-content">

   

    
    
<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>

<div id="toc"></div>

    


<!--
    

    <a target="_blank" href="https://github.com/tgirke/systemPipeR/blob/gh-pages/pages//mydoc_systemPipeRIBOseq_01.md" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
-->
    

  <p>First/last name (first.last@ucr.edu)</p>

<p>Last update: 15 October, 2017</p>

<p>Alternative formats of this tutorial:
[ <a href="http://girke.bioinformatics.ucr.edu/systemPipeR/pages/mydoc/systemPipeRIBOseq.html" target="_blank">HTML</a> ],
[ <a href="http://girke.bioinformatics.ucr.edu/systemPipeR/pages/mydoc/systemPipeRIBOseq.pdf" target="_blank">PDF</a> ],
[ <a href="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/_vignettes/14_RIBOseqWorkflow/systemPipeRIBOseq.Rmd" target="_blank">.Rmd</a> ],
[ <a href="https://raw.githubusercontent.com/tgirke/systemPipeR/gh-pages/_vignettes/14_RIBOseqWorkflow/systemPipeRIBOseq.R" target="_blank">.R</a> ]</p>

<p>Ribo-Seq and polyRibo-Seq are a specific form of RNA-Seq gene expression
experiments utilizing mRNA subpopulations directly bound to ribosomes.
Compared to standard RNA-Seq, their readout of gene expression provides a
better approximation of downstream protein abundance profiles due to their
close association with translational processes. The most important difference
among the two is that polyRibo-Seq utilizes polyribosomal RNA for sequencing,
whereas Ribo-Seq is a footprinting approach restricted to sequencing RNA
fragments protected by ribosomes (Ingolia et al., 2009; Aspden et al., 2014; @Juntawong2015-ru}.</p>

<p>The workflow presented in this vignette contains most of the data analysis
steps described by (Juntawong et al., 2014) including functionalities useful for
processing both polyRibo-Seq and Ribo-Seq experiments. To improve re-usability
and adapt to recent changes of software versions (<em>e.g.</em> R, Bioconductor and
short read aligners), the code has been optimized accordingly. Thus, the
results obtained with the updated workflow are expected to be similar but not
necessarily identical with the published results described in the original
paper.</p>

<p>Relevant analysis steps of this workflow include read preprocessing, read
alignments against a reference genome, counting of reads overlapping with a
wide range of genomic features (<em>e.g.</em> CDSs, UTRs, uORFs, rRNAs, etc.),
differential gene expression and differential ribosome binding analyses, as
well as a variety of genome-wide summary plots for visualizing RNA expression
trends. Functions are provided for evaluating the quality of Ribo-seq data,
for identifying novel expressed regions in the genomes, and for gaining
insights into gene regulation at the post-transcriptional and translational
levels. For example, the functions <code class="highlighter-rouge">genFeatures</code> and
<code class="highlighter-rouge">featuretypeCounts</code> can be used to quantify the expression output for
all feature types included in a genome annotation (<code class="highlighter-rouge">e.g.</code> genes,
introns, exons, miRNAs, intergenic regions, etc.). To determine the approximate
read length of ribosome footprints in Ribo-Seq experiments, these feature type
counts can be obtained and plotted for specific read lengths separately.
Typically, the most abundant read length obtained for translated features
corresponds to the approximate footprint length occupied by the ribosomes of a
given organism group. Based on the results from several Ribo-Seq studies, these
ribosome footprints are typically ~30 nucleotides long
(Ingolia et al., 2011; Ingolia et al., 2009; Juntawong et al., 2014).  However, their
length can vary by several nucleotides depending upon the optimization of the
RNA digestion step and various factors associated with translational
regulation.  For quality control purposes of Ribo-Seq experiments it is also
useful to monitor the abundance of reads mapping to rRNA genes due to the high
rRNA content of ribosomes. This information can be generated with the 
<code class="highlighter-rouge">featuretypeCounts</code> function described above.</p>

<p>Coverage trends along transcripts summarized for any number of transcripts can
be obtained and plotted with the functions <code class="highlighter-rouge">featureCoverage</code> and
<code class="highlighter-rouge">plotfeatureCoverage</code>, respectively. Their results allow monitoring
of the phasing of ribosome movements along triplets of coding sequences.
Commonly, high quality data will display here for the first nucleotide of each
codon the highest depth of coverage computed for the 5’ ends of the aligned
reads.</p>

<p>Ribo-seq data can also be used to evaluate various aspects of translational
control due to ribosome occupancy in upstream open reading frames (uORFs). The
latter are frequently present in (or near) 5’ UTRs of transcripts. For this,
the function <code class="highlighter-rouge">predORFs</code> can be used to identify ORFs in the
nucleotide sequences of transcripts or their subcomponents such as UTR regions.
After scaling the resulting ORF coordinates back to the corresponding genome
locations using <code class="highlighter-rouge">scaleRanges</code>, one can use these novel features
(<em>e.g.</em> uORFs) for expression analysis routines similar to those
employed for pre-existing annotations, such as the exonic regions of genes. For
instance, in Ribo-Seq experiments one can use this approach to systematically identify all
transcripts occupied by ribosomes in their uORF regions. The binding of
ribosomes to uORF regions may indicate a regulatory role in the translation of
the downstream main ORFs and/or translation of the uORFs into functionally
relevant peptides.</p>

<h2 id="experimental-design">Experimental design</h2>
<p>Typically, users want to specify here all information relevant for the analysis
of their NGS study. This includes detailed descriptions of FASTQ files,
experimental design, reference genome, gene annotations, etc.</p>

<p><br /><br /><center><a href="mydoc_systemPipeRIBOseq_01.html"><img src="images/left_arrow.png" alt="Previous page." /></a>Previous Page &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Next Page
<a href="mydoc_systemPipeRIBOseq_02.html"><img src="images/right_arrow.png" alt="Next page." /></a></center></p>


    <div class="tags">
        
    </div>

    

</div>

<hr class="shaded"/>

<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2017 UC Riverside. All rights reserved. <br />
<span>Page last updated:</span> Sun Oct 15 13:21:42 2017<br/> Site last generated: Nov 12, 2017 <br />
<!--
<p><img src="images/company_logo.png" alt="Company logo"/></p>
-->
</div>
            </div>
</footer>


    </div>
    <!-- /.row -->
</div>
<!-- /.container -->
    </div>

</body>

</html>
