<!DOCTYPE html>
<head>
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>6. Range Operations | systemPipeR Website</title>
<link rel="stylesheet" href="css/syntax.css">


<link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css">-->
<link rel="stylesheet" href="css/modern-business.css">
<link rel="stylesheet" href="css/lavish-bootstrap.css">
<link rel="stylesheet" href="css/customstyles.css">
<link rel="stylesheet" href="css/theme-blue.css">

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="js/jquery.navgoco.min.js"></script>


<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="js/toc.js"></script>
<script src="js/customscripts.js"></script>

<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

<link rel="shortcut icon" href="images/favicon.ico">

<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->

<link rel="alternate" type="application/rss+xml" title="" href="http://localhost:4000feed.xml">

    <script>
        $(document).ready(function() {
            // Initialize navgoco with default options
            $("#mysidebar").navgoco({
                caretHtml: '',
                accordion: true,
                openClass: 'active', // open
                save: false, // leave false or nav highlighting doesn't work right
                cookie: {
                    name: 'navgoco',
                    expires: false,
                    path: '/'
                },
                slide: {
                    duration: 400,
                    easing: 'swing'
                }
            });

            $("#collapseAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', false);
            });

            $("#expandAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', true);
            });

        });

    </script>
    <script>
        $(function () {
            $('[data-toggle="tooltip"]').tooltip()
        })
    </script>
    <script>
        $(document).ready(function() {
            $("#tg-sb-link").click(function() {
                $("#tg-sb-sidebar").toggle();
                $("#tg-sb-content").toggleClass('col-md-9');
                $("#tg-sb-content").toggleClass('col-md-12');
                $("#tg-sb-icon").toggleClass('fa-toggle-on');
                $("#tg-sb-icon").toggleClass('fa-toggle-off');
            });
        });
    </script>
    

</head>
<body>
<!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="fa fa-home fa-lg navbar-brand" href="index.html">&nbsp;<span class="projectTitle"> systemPipeR Project Website</span></a>
        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- toggle sidebar button -->
                <li><a id="tg-sb-link" href="#"><i id="tg-sb-icon" class="fa fa-toggle-on"></i> Nav</a></li>
                <!-- entries without drop-downs appear here -->
                
                
                
                <li><a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html" target="_blank">Bioc</a></li>
                
                
                
                <li><a href="https://github.com/tgirke/systemPipeR" target="_blank">GitHub Repo</a></li>
                
                
                
                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        <li><a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual" target="_blank">R & Bioconductor Manual</a></li>
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        <li><a href="http://hpcc.ucr.edu/" target="_blank">HPC Center (HPCC)</a></li>
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                    </ul>
                </li>
                
                
                
			<li>



  <a class="email" title="Submit feedback" href="#" onclick="javascript:window.location='mailto:thomas.girke@ucr.edu?subject=systemPipeR Website feedback&body=I have some feedback about the 6. Range Operations page: ' + window.location.href;"><i class="fa fa-envelope-o"></i> Feedback</a>

</li>

		
                <!--comment out this block if you want to hide search-->
                <li>
                    <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                            SimpleJekyllSearch.init({
                                searchInput: document.getElementById('search-input'),
                                resultsContainer: document.getElementById('results-container'),
                                dataSource: 'search.json',
                                searchResultTemplate: '<li><a href="{url}" title="6. Range Operations">{title}</a></li>',
                    noResultsText: 'No results found.',
                            limit: 10,
                            fuzzy: true,
                    })
                    </script>
                    <!--end search-->
                </li>
            </ul>
        </div>
        </div>
        <!-- /.container -->
</nav>

<!-- Page Content -->
<div class="container">
    <div class="col-lg-12">&nbsp;</div>
    <!-- Content Row -->
    <div class="row">
        <!-- Sidebar Column -->
        <div class="col-md-3" id="tg-sb-sidebar">

          












<ul id="mysidebar" class="nav">
    <li class="sidebarTitle"><center><font color="#26466D" size="6">systemPipeR</font></center> <center><font color="#8b9299" size="3">...</font></center></li>
    
    
    
        
    
    <li>
        <a href="#">Home</a>
        <ul>
            
            
            
            <li><a href="index.html">Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_syllabus.html">Syllabus</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Internal</a>
        <ul>
            
            
            
            <li><a href="mydoc_internal.html">Internal Resources</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Schedule</a>
        <ul>
            
            
            
            <li><a href="mydoc_schedule.html">Course Schedule</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Lecture Slides</a>
        <ul>
            
            
            
            <li><a href="mydoc_slides_01.html">Course Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_02.html">Molbiol Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_03.html">Databases and Software</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_04.html">DNA Sequencing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_05.html">Introduction to R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_06.html">Sequence Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_07.html">Programming in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_08.html">Multiple Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_09.html">Short Read Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_10.html">NGS Analysis Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_11.html">Gene Expression Analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_12.html">NGS Workflows</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_13.html">ChIP-Seq Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_14.html">VAR-Seq Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_15.html">Sequence Assembly</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_16.html">Gene Annotation and Ontologies</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_17.html">Cluster Analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_18.html">Profile HMMs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_19.html">Introduction to Phylogenetics</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Homework</a>
        <ul>
            
            
            
            <li><a href="mydoc_homework_01.html">HW1 - Online Tools</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_02.html">HW2 - Biocluster/Linux</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_03.html">HW3 - Introduction to R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_04.html">HW4 - Pairwise Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_09.html">HW5 - Programming in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_08.html">HW6 - Sequence Analysis</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">GitHub</a>
        <ul>
            
            
            
            <li><a href="mydoc_tutorial_01.html">Table of Content</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#github-in-gen242">GitHub in GEN242</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#installing-git">Installation</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#git-basics-from-command-line">Git from Command-line</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#github-basics-from-command-line">GitHub from Command-line</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#using-github-from-rstudio">GitHub from RStudio</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Linux & HPCC Cluster</a>
        <ul>
            
            
            
            <li><a href="mydoc_tutorial_02.html">Table of Content</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#hpcc-cluster-overview">HPCC Cluster</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#linux-basics">Linux Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#software-and-module-system">Software & Module System</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#big-data-storage">Big Data Storage</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#queuing-system-slurm">Queueing System Slurm</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#textcode-editors">Text Editors</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#vim-r-tmux-essentials">Vim-R-Tmux</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#nvim-r-tmux">Nvim-R-Tmux</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Introduction to R</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rbasics_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_02.html">2. R Package Repositories</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_03.html">3. Installation of R Packages</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_04.html">4. Getting Around</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_05.html">5. Basic Syntax</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_06.html">6. Data Types</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_07.html">7. Data Objects</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_08.html">8. Important Utilities</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_09.html">9. Operators and Calculations</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_10.html">10. Reading and Writing External Data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_11.html">11. Useful R Functions</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_12.html">12. dplyr Environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_13.html">13. SQLite Databases</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_14.html">14. Graphics in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_15.html">15. Analysis Routine</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_16.html">16. R Markdown</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_17.html">17. Shiny Web Apps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_18.html">18. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_19.html">19. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Programming in R</a>
        <ul>
            
            
            
            <li><a href="mydoc_Programming_in_R_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_02.html">2. Control Structures</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_03.html">3. Loops</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_04.html">4. Functions</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_05.html">5. Useful Utilities</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_06.html">6. Running R Scripts</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_07.html">7. Building R Packages</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_08.html">8. Programming Exercises</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_09.html">9. Homework 5</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_10.html">10. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Analysis Basics</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rsequences_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_02.html">2. Package Requirements</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_03.html">3. Strings in R Base</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_04.html">4. Sequences in Bioconductor</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_05.html">5. NGS Sequences</a></li>
            
            
            
            
            
            
            <li class="active"><a href="mydoc_Rsequences_06.html">6. Range Operations</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_07.html">7. Transcript Ranges</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_08.html">8. Homework 6</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_09.html">9. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_10.html">10. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Workflow Overview</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeR_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_2.html">2. Getting Started</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_3.html">3. Workflow overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_4.html">4. Workflow templates</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_5.html">5. Version information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">RNA-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_02.html">2. Samples and environment settings</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_05.html">5. Read quantification</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_06.html">6. Analysis of DEGs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_07.html">7. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_08.html">8. Clustering and heat maps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_09.html">9. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_10.html">10. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">ChIP-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_05.html">5. Utilities for coverage data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_06.html">6. Peak calling with MACS2</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_07.html">7. Annotate peaks with genomic context</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_08.html">8. Count reads overlapping peaks</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_09.html">9. Differential binding analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_10.html">10. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_11.html">11. Motif analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">VAR-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeVARseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_05.html">5. Variant calling</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_06.html">6. Filter variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_07.html">7. Annotate filtered variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_08.html">8. Combine annotation results among samples</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_09.html">9. Summary statistics of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_10.html">10. Venn diagram of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_11.html">11. Plot variants programmatically</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Graphics and Visualization</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rgraphics_1.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_2.html">2. Base Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_3.html">3. Grid Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_4.html">4. lattice Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_5.html">5. ggplot2 Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_6.html">6. Specialty Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_7.html">7. Genome Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_8.html">8. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Cluster Analysis</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rclustering_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_2.html">2. Data Preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_3.html">3. Clustering Algorithms</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_4.html">4. Clustering Exercises</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_5.html">5. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Paper Presentations</a>
        <ul>
            
            
            
            <li><a href="mydoc_paper_presentations.html">Paper Presentations</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Course Projects</a>
        <ul>
            
            
            
            <li><a href="mydoc_project_intro.html">Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_chipseq1.html">ChIP-Seq1 - Motif Prediction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_chipseq2.html">ChIP-Seq2 - Peak Callers</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_rnaseq1.html">RNA-Seq1 - Aligners</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_rnaseq2.html">RNA-Seq2 - DEG Methods</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_varseq1.html">VAR-Seq1 - Functional Prediction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_data.html">Project Data</a></li>
            
            
            
            
        </ul>
        
        
        
        <!-- if you aren't using the accordion, uncomment this block:
           <p class="external">
               <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
           </p>
           -->
    </li>
</ul>
</div>

<!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
<script>$("li.active").parents('li').toggleClass("active");</script>

    <!-- Content Column -->
    <div class="col-md-9" id="tg-sb-content">
        <div class="post-header">
   <h1 class="post-title-main">6. Range Operations</h1>
</div>



<div class="post-content">

   

    
    
<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>

<div id="toc"></div>

    


<!--
    

    <a target="_blank" href="https://github.com/tgirke/systemPipeR/blob/gh-pages/pages//mydoc_Rsequences_06.md" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
-->
    

  <h2 id="important-data-objects-for-range-operations">Important Data Objects for Range Operations</h2>

<ul>
  <li><code class="highlighter-rouge">IRanges</code>: stores range data only (IRanges library)</li>
  <li><code class="highlighter-rouge">GRanges</code>: stores ranges and annotations (GenomicRanges library)</li>
  <li><code class="highlighter-rouge">GRangesList</code>: list version of GRanges container (GenomicRanges library)</li>
</ul>

<h2 id="range-data-are-stored-in-iranges-and-granges-containers">Range Data Are Stored in <code class="highlighter-rouge">IRanges</code> and <code class="highlighter-rouge">GRanges</code> Containers</h2>

<h3 id="construct-granges-object">Construct <code class="highlighter-rouge">GRanges</code> Object</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">GenomicRanges</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">rtracklayer</span><span class="p">)</span><span class="w">
</span><span class="n">gr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GRanges</span><span class="p">(</span><span class="n">seqnames</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Rle</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"chr1"</span><span class="p">,</span><span class="w"> </span><span class="s2">"chr2"</span><span class="p">,</span><span class="w"> </span><span class="s2">"chr1"</span><span class="p">,</span><span class="w"> </span><span class="s2">"chr3"</span><span class="p">),</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">)),</span><span class="w"> </span><span class="n">ranges</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">IRanges</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">end</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">7</span><span class="o">:</span><span class="m">16</span><span class="p">,</span><span class="w"> </span><span class="n">names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">head</span><span class="p">(</span><span class="nb">letters</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">)),</span><span class="w"> </span><span class="n">strand</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Rle</span><span class="p">(</span><span class="n">strand</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"-"</span><span class="p">,</span><span class="w"> </span><span class="s2">"+"</span><span class="p">,</span><span class="w"> </span><span class="s2">"*"</span><span class="p">,</span><span class="w"> </span><span class="s2">"+"</span><span class="p">,</span><span class="w"> </span><span class="s2">"-"</span><span class="p">)),</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">)),</span><span class="w"> </span><span class="n">score</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">GC</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">seq</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="m">0</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="m">10</span><span class="p">))</span><span class="w"> </span><span class="c1"># Example of creating a GRanges object with its constructor function.
</span></code></pre>
</div>

<h3 id="import-gff-into-granges-object">Import GFF into <code class="highlighter-rouge">GRanges</code> Object</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">import.gff</span><span class="p">(</span><span class="s2">"http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/gff3.gff"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Imports a simplified GFF3 genome annotation file.
</span><span class="n">seqlengths</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">end</span><span class="p">(</span><span class="n">ranges</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="n">which</span><span class="p">(</span><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)[,</span><span class="s2">"type"</span><span class="p">]</span><span class="o">==</span><span class="s2">"chromosome"</span><span class="p">),]))</span><span class="w"> 
</span><span class="nf">names</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w"> </span><span class="c1"># Assigns names to corresponding slot
</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 4 ranges and 10 metadata columns:
##     seqnames           ranges strand |   source       type     score     phase                  ID
##        &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt;
##   1     Chr1 [   1, 30427671]      + |   TAIR10 chromosome      &lt;NA&gt;      &lt;NA&gt;                Chr1
##   2     Chr1 [3631,     5899]      + |   TAIR10       gene      &lt;NA&gt;      &lt;NA&gt;           AT1G01010
##   3     Chr1 [3631,     5899]      + |   TAIR10       mRNA      &lt;NA&gt;      &lt;NA&gt;         AT1G01010.1
##   4     Chr1 [3760,     5630]      + |   TAIR10    protein      &lt;NA&gt;      &lt;NA&gt; AT1G01010.1-Protein
##            Name                Note          Parent       Index Derives_from
##     &lt;character&gt;     &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##   1        Chr1                                            &lt;NA&gt;         &lt;NA&gt;
##   2   AT1G01010 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   3 AT1G01010.1                           AT1G01010           1         &lt;NA&gt;
##   4 AT1G01010.1                                            &lt;NA&gt;  AT1G01010.1
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<h3 id="coerce-granges-object-to-dataframe">Coerce <code class="highlighter-rouge">GRanges</code> object to <code class="highlighter-rouge">data.frame</code></h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">as.data.frame</span><span class="p">(</span><span class="n">gff</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">7</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##   seqnames start      end    width strand source       type
## 1     Chr1     1 30427671 30427671      + TAIR10 chromosome
## 2     Chr1  3631     5899     2269      + TAIR10       gene
## 3     Chr1  3631     5899     2269      + TAIR10       mRNA
## 4     Chr1  3760     5630     1871      + TAIR10    protein
</code></pre>
</div>

<h3 id="coerce-granges-to-rangeddata-object-and-vice-versa">Coerce <code class="highlighter-rouge">GRanges</code> to <code class="highlighter-rouge">RangedData</code> object and vice versa</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gff_rd</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="s2">"RangedData"</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning in cntxt$tailcall: closing unused connection 7 (/tmp/Rtmp7jsa8n/file78fc7c51c885)
</code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Warning in cntxt$tailcall: closing unused connection 5 (/tmp/Rtmp7jsa8n/file78fc7c51c885)
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gff_gr</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as</span><span class="p">(</span><span class="n">gff_rd</span><span class="p">,</span><span class="w"> </span><span class="s2">"GRanges"</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<h2 id="utilities-for-range-containers">Utilities for Range Containers</h2>

<h3 id="accessor-and-subsetting-methods-for-granges-objects">Accessor and subsetting methods for GRanges objects</h3>

<p>Subsetting and replacement</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 4 ranges and 10 metadata columns:
##     seqnames           ranges strand |   source       type     score     phase                  ID
##        &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt;
##   1     Chr1 [   1, 30427671]      + |   TAIR10 chromosome      &lt;NA&gt;      &lt;NA&gt;                Chr1
##   2     Chr1 [3631,     5899]      + |   TAIR10       gene      &lt;NA&gt;      &lt;NA&gt;           AT1G01010
##   3     Chr1 [3631,     5899]      + |   TAIR10       mRNA      &lt;NA&gt;      &lt;NA&gt;         AT1G01010.1
##   4     Chr1 [3760,     5630]      + |   TAIR10    protein      &lt;NA&gt;      &lt;NA&gt; AT1G01010.1-Protein
##            Name                Note          Parent       Index Derives_from
##     &lt;character&gt;     &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##   1        Chr1                                            &lt;NA&gt;         &lt;NA&gt;
##   2   AT1G01010 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   3 AT1G01010.1                           AT1G01010           1         &lt;NA&gt;
##   4 AT1G01010.1                                            &lt;NA&gt;  AT1G01010.1
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="s2">"type"</span><span class="p">,</span><span class="w"> </span><span class="s2">"ID"</span><span class="p">)]</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 4 ranges and 2 metadata columns:
##     seqnames           ranges strand |       type                  ID
##        &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;factor&gt;         &lt;character&gt;
##   1     Chr1 [   1, 30427671]      + | chromosome                Chr1
##   2     Chr1 [3631,     5899]      + |       gene           AT1G01010
##   3     Chr1 [3631,     5899]      + |       mRNA         AT1G01010.1
##   4     Chr1 [3760,     5630]      + |    protein AT1G01010.1-Protein
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="p">[</span><span class="m">2</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gff</span><span class="p">[</span><span class="m">3</span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<p>GRanges objects can be concatenated with the <code class="highlighter-rouge">c</code> function</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="nf">c</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">],</span><span class="w"> </span><span class="n">gff</span><span class="p">[</span><span class="m">401</span><span class="o">:</span><span class="m">402</span><span class="p">])</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 4 ranges and 10 metadata columns:
##       seqnames           ranges strand |   source           type     score     phase
##          &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##     1     Chr1 [   1, 30427671]      + |   TAIR10     chromosome      &lt;NA&gt;      &lt;NA&gt;
##     2     Chr1 [3631,     5899]      + |   TAIR10           mRNA      &lt;NA&gt;      &lt;NA&gt;
##   401     Chr5 [5516,     5769]      - |   TAIR10        protein      &lt;NA&gt;      &lt;NA&gt;
##   402     Chr5 [5770,     5801]      - |   TAIR10 five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##                        ID        Name            Note          Parent       Index Derives_from
##               &lt;character&gt; &lt;character&gt; &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##     1                Chr1        Chr1                                        &lt;NA&gt;         &lt;NA&gt;
##     2         AT1G01010.1 AT1G01010.1                       AT1G01010           1         &lt;NA&gt;
##   401 AT5G01015.2-Protein AT5G01015.2                                        &lt;NA&gt;  AT5G01015.2
##   402                &lt;NA&gt;        &lt;NA&gt;                     AT5G01015.2        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<p>Acessor functions</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">seqnames</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## factor-Rle of length 449 with 7 runs
##   Lengths:   72   22   38  118  172   13   14
##   Values : Chr1 Chr2 Chr3 Chr4 Chr5 ChrC ChrM
## Levels(7): Chr1 Chr2 Chr3 Chr4 Chr5 ChrC ChrM
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">ranges</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## IRanges object with 449 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##     1         1  30427671  30427671
##     2      3631      5899      2269
##     3      3631      5899      2269
##     4      3760      5630      1871
##     5      3631      3913       283
##   ...       ...       ...       ...
##   445     11918     12241       324
##   446     11918     12241       324
##   447     11918     12241       324
##   448     11918     12241       324
##   449     11918     12241       324
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">strand</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## factor-Rle of length 449 with 13 runs
##   Lengths:  18  54  28  21  12 117   1 171   1  12   1   8   5
##   Values :   +   -   +   -   +   -   +   -   +   -   +   -   +
## Levels(3): + - *
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">seqlengths</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##     Chr1     Chr2     Chr3     Chr4     Chr5     ChrC     ChrM 
## 30427671 19698289 23459830 18585056 26975502   154478   366924
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">start</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1]    1 3631 3631 3760
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">end</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 30427671     5899     5899     5630
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">width</span><span class="p">(</span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## [1] 30427671     2269     2269     1871
</code></pre>
</div>

<p>Accessing metadata component</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w"> </span><span class="c1"># or elementMetadata(gff)
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## DataFrame with 449 rows and 10 columns
##       source       type     score     phase                  ID        Name                Note
##     &lt;factor&gt;   &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;         &lt;character&gt; &lt;character&gt;     &lt;CharacterList&gt;
## 1     TAIR10 chromosome        NA        NA                Chr1        Chr1                    
## 2     TAIR10       mRNA        NA        NA         AT1G01010.1 AT1G01010.1                    
## 3     TAIR10       mRNA        NA        NA         AT1G01010.1 AT1G01010.1                    
## 4     TAIR10    protein        NA        NA AT1G01010.1-Protein AT1G01010.1                    
## 5     TAIR10       exon        NA        NA                  NA          NA                    
## ...      ...        ...       ...       ...                 ...         ...                 ...
## 445   TAIR10       gene        NA        NA           ATMG00030   ATMG00030 protein_coding_gene
## 446   TAIR10       mRNA        NA        NA         ATMG00030.1 ATMG00030.1                    
## 447   TAIR10    protein        NA        NA ATMG00030.1-Protein ATMG00030.1                    
## 448   TAIR10       exon        NA        NA                  NA          NA                    
## 449   TAIR10        CDS        NA         0                  NA          NA                    
##                              Parent       Index Derives_from
##                     &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
## 1                                            NA           NA
## 2                         AT1G01010           1           NA
## 3                         AT1G01010           1           NA
## 4                                            NA  AT1G01010.1
## 5                       AT1G01010.1          NA           NA
## ...                             ...         ...          ...
## 445                                          NA           NA
## 446                       ATMG00030           1           NA
## 447                                          NA  ATMG00030.1
## 448                     ATMG00030.1          NA           NA
## 449 ATMG00030.1,ATMG00030.1-Protein          NA           NA
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)[,</span><span class="w"> </span><span class="s2">"type"</span><span class="p">][</span><span class="m">1</span><span class="o">:</span><span class="m">20</span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] chromosome      mRNA            mRNA            protein         exon            five_prime_UTR 
##  [7] CDS             exon            CDS             exon            CDS             exon           
## [13] CDS             exon            CDS             exon            CDS             three_prime_UTR
## [19] gene            mRNA           
## Levels: chromosome gene mRNA protein exon five_prime_UTR CDS three_prime_UTR rRNA tRNA
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="p">[</span><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)[</span><span class="w"> </span><span class="p">,</span><span class="s2">"type"</span><span class="p">]</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="s2">"gene"</span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 21 ranges and 10 metadata columns:
##       seqnames         ranges strand |   source     type     score     phase          ID
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt; &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt; &lt;character&gt;
##    19     Chr1 [ 5928,  8737]      - |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT1G01020
##    64     Chr1 [11649, 13714]      - |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT1G01030
##    74     Chr2 [ 1025,  2810]      + |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT2G01008
##    84     Chr2 [ 3706,  5513]      + |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT2G01010
##    87     Chr2 [ 5782,  5945]      + |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   AT2G01020
##   ...      ...            ...    ... .      ...      ...       ...       ...         ...
##   427     ChrC [  383,  1444]      - |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATCG00020
##   432     ChrC [ 1717,  4347]      - |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATCG00030
##   437     ChrM [  273,   734]      - |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATMG00010
##   442     ChrM [ 8848, 11415]      - |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATMG00020
##   445     ChrM [11918, 12241]      + |   TAIR10     gene      &lt;NA&gt;      &lt;NA&gt;   ATMG00030
##              Name                Note          Parent       Index Derives_from
##       &lt;character&gt;     &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;character&gt;  &lt;character&gt;
##    19   AT1G01020 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##    64   AT1G01030 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##    74   AT2G01008 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##    84   AT2G01010                rRNA                        &lt;NA&gt;         &lt;NA&gt;
##    87   AT2G01020                rRNA                        &lt;NA&gt;         &lt;NA&gt;
##   ...         ...                 ...             ...         ...          ...
##   427   ATCG00020 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   432   ATCG00030                tRNA                        &lt;NA&gt;         &lt;NA&gt;
##   437   ATMG00010 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   442   ATMG00020                rRNA                        &lt;NA&gt;         &lt;NA&gt;
##   445   ATMG00030 protein_coding_gene                        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<h3 id="useful-utilities-for-granges-objects">Useful utilities for GRanges objects</h3>

<p>Remove chromosome ranges</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gff</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gff</span><span class="p">[</span><span class="n">values</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="o">$</span><span class="n">type</span><span class="w"> </span><span class="o">!=</span><span class="w"> </span><span class="s2">"chromosome"</span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<p>Erase the strand information</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">strand</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"*"</span><span class="w"> 
</span></code></pre>
</div>

<p>Collapses overlapping ranges to continuous ranges.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">reduce</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 22 ranges and 0 metadata columns:
##        seqnames         ranges strand
##           &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     Chr1 [ 3631,  5899]      *
##    [2]     Chr1 [ 5928,  8737]      *
##    [3]     Chr1 [11649, 13714]      *
##    [4]     Chr2 [ 1025,  2810]      *
##    [5]     Chr2 [ 3706,  5513]      *
##    ...      ...            ...    ...
##   [18]     ChrC [  383,  1444]      *
##   [19]     ChrC [ 1717,  4347]      *
##   [20]     ChrM [  273,   734]      *
##   [21]     ChrM [ 8848, 11415]      *
##   [22]     ChrM [11918, 12241]      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<p>Return uncovered regions</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">gaps</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 43 ranges and 0 metadata columns:
##        seqnames           ranges strand
##           &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     Chr1 [   1, 30427671]      +
##    [2]     Chr1 [   1, 30427671]      -
##    [3]     Chr1 [   1,     3630]      *
##    [4]     Chr1 [5900,     5927]      *
##    [5]     Chr1 [8738,    11648]      *
##    ...      ...              ...    ...
##   [39]     ChrM  [    1, 366924]      -
##   [40]     ChrM  [    1,    272]      *
##   [41]     ChrM  [  735,   8847]      *
##   [42]     ChrM  [11416,  11917]      *
##   [43]     ChrM  [12242, 366924]      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<p>More intuitive way to get uncovered regions</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">setdiff</span><span class="p">(</span><span class="n">as</span><span class="p">(</span><span class="n">seqinfo</span><span class="p">(</span><span class="n">gff</span><span class="p">),</span><span class="w"> </span><span class="s2">"GRanges"</span><span class="p">),</span><span class="w"> </span><span class="n">gff</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 29 ranges and 0 metadata columns:
##        seqnames            ranges strand
##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;
##    [1]     Chr1 [    1,     3630]      *
##    [2]     Chr1 [ 5900,     5927]      *
##    [3]     Chr1 [ 8738,    11648]      *
##    [4]     Chr1 [13715, 30427671]      *
##    [5]     Chr2 [    1,     1024]      *
##    ...      ...               ...    ...
##   [25]     ChrC   [ 4348, 154478]      *
##   [26]     ChrM   [    1,    272]      *
##   [27]     ChrM   [  735,   8847]      *
##   [28]     ChrM   [11416,  11917]      *
##   [29]     ChrM   [12242, 366924]      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<p>Return disjoint ranges</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">disjoin</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 211 ranges and 0 metadata columns:
##         seqnames         ranges strand
##            &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt;
##     [1]     Chr1   [3631, 3759]      *
##     [2]     Chr1   [3760, 3913]      *
##     [3]     Chr1   [3914, 3995]      *
##     [4]     Chr1   [3996, 4276]      *
##     [5]     Chr1   [4277, 4485]      *
##     ...      ...            ...    ...
##   [207]     ChrC [ 1752,  4310]      *
##   [208]     ChrC [ 4311,  4347]      *
##   [209]     ChrM [  273,   734]      *
##   [210]     ChrM [ 8848, 11415]      *
##   [211]     ChrM [11918, 12241]      *
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<p>Returns coverage of ranges</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">coverage</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## RleList of length 7
## $Chr1
## integer-Rle of length 30427671 with 45 runs
##   Lengths:     3630      129      154       82      281 ...      233      161      380 30413957
##   Values :        0        4        5        3        5 ...        4        2        4        0
## 
## $Chr2
## integer-Rle of length 19698289 with 14 runs
##   Lengths:     1024      248      185       53      362 ...      164      625      102 19691617
##   Values :        0        5        3        5        3 ...        3        0        5        0
## 
## $Chr3
## integer-Rle of length 23459830 with 29 runs
##   Lengths:     1652      145      139      111       95 ...      155      148      156 23453781
##   Values :        0        4        5        3        5 ...        3        5        4        0
## 
## $Chr4
## integer-Rle of length 18585056 with 72 runs
##   Lengths:     1179      357     1358      128      872 ...      212      114       74 18571697
##   Values :        0        5        0        5        3 ...        3        5        4        0
## 
## $Chr5
## integer-Rle of length 26975502 with 64 runs
##   Lengths:     1222       28       28      109       72 ...       76       55      174 26967058
##   Values :        0        4        7       13       16 ...        3        5        4        0
## 
## ...
## &lt;2 more elements&gt;
</code></pre>
</div>

<p>Return the index pairings for overlapping ranges</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">findOverlaps</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## Hits object with 55 hits and 0 metadata columns:
##        queryHits subjectHits
##        &lt;integer&gt;   &lt;integer&gt;
##    [1]         1           1
##    [2]         1           2
##    [3]         1           4
##    [4]         1           3
##    [5]         2           1
##    ...       ...         ...
##   [51]        16           1
##   [52]        16           2
##   [53]        16           3
##   [54]        17           1
##   [55]        17           2
##   -------
##   queryLength: 442 / subjectLength: 4
</code></pre>
</div>

<p>Counts overlapping ranges</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">countOverlaps</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])[</span><span class="m">1</span><span class="o">:</span><span class="m">40</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 
##  4  4  4  4  3  4  3  3  3  3  3  3  3  3  3  3  2  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0 
## 35 36 37 38 39 40 41 
##  0  0  0  0  0  0  0
</code></pre>
</div>

<p>Return only overlapping ranges</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">subsetByOverlaps</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="n">gff</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">])</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 17 ranges and 10 metadata columns:
##      seqnames       ranges strand |   source            type     score     phase
##         &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;        &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##    2     Chr1 [3631, 5899]      * |   TAIR10            mRNA      &lt;NA&gt;      &lt;NA&gt;
##    3     Chr1 [3631, 5899]      * |   TAIR10            mRNA      &lt;NA&gt;      &lt;NA&gt;
##    4     Chr1 [3760, 5630]      * |   TAIR10         protein      &lt;NA&gt;      &lt;NA&gt;
##    5     Chr1 [3631, 3913]      * |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##    6     Chr1 [3631, 3759]      * |   TAIR10  five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##   ..      ...          ...    ... .      ...             ...       ...       ...
##   14     Chr1 [5174, 5326]      * |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##   15     Chr1 [5174, 5326]      * |   TAIR10             CDS      &lt;NA&gt;         0
##   16     Chr1 [5439, 5899]      * |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##   17     Chr1 [5439, 5630]      * |   TAIR10             CDS      &lt;NA&gt;         0
##   18     Chr1 [5631, 5899]      * |   TAIR10 three_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##                       ID        Name            Note                          Parent       Index
##              &lt;character&gt; &lt;character&gt; &lt;CharacterList&gt;                 &lt;CharacterList&gt; &lt;character&gt;
##    2         AT1G01010.1 AT1G01010.1                                       AT1G01010           1
##    3         AT1G01010.1 AT1G01010.1                                       AT1G01010           1
##    4 AT1G01010.1-Protein AT1G01010.1                                                        &lt;NA&gt;
##    5                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##    6                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##   ..                 ...         ...             ...                             ...         ...
##   14                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##   15                &lt;NA&gt;        &lt;NA&gt;                 AT1G01010.1,AT1G01010.1-Protein        &lt;NA&gt;
##   16                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##   17                &lt;NA&gt;        &lt;NA&gt;                 AT1G01010.1,AT1G01010.1-Protein        &lt;NA&gt;
##   18                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##      Derives_from
##       &lt;character&gt;
##    2         &lt;NA&gt;
##    3         &lt;NA&gt;
##    4  AT1G01010.1
##    5         &lt;NA&gt;
##    6         &lt;NA&gt;
##   ..          ...
##   14         &lt;NA&gt;
##   15         &lt;NA&gt;
##   16         &lt;NA&gt;
##   17         &lt;NA&gt;
##   18         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<h2 id="grangeslist-objects">GRangesList Objects</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">split</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">gff</span><span class="p">))</span><span class="w"> </span><span class="c1"># Stores every range in separate component of a GRangesList object
</span><span class="n">split</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="n">seqnames</span><span class="p">(</span><span class="n">gff</span><span class="p">))</span><span class="w"> </span><span class="c1"># Stores ranges of each chromosome in separate component.
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRangesList object of length 7:
## $Chr1 
## GRanges object with 71 ranges and 10 metadata columns:
##      seqnames         ranges strand |   source            type     score     phase
##         &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;        &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##    2     Chr1   [3631, 5899]      * |   TAIR10            mRNA      &lt;NA&gt;      &lt;NA&gt;
##    3     Chr1   [3631, 5899]      * |   TAIR10            mRNA      &lt;NA&gt;      &lt;NA&gt;
##    4     Chr1   [3760, 5630]      * |   TAIR10         protein      &lt;NA&gt;      &lt;NA&gt;
##    5     Chr1   [3631, 3913]      * |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##    6     Chr1   [3631, 3759]      * |   TAIR10  five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##   ..      ...            ...    ... .      ...             ...       ...       ...
##   68     Chr1 [13335, 13714]      * |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##   69     Chr1 [12941, 13173]      * |   TAIR10  five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##   70     Chr1 [11864, 12940]      * |   TAIR10             CDS      &lt;NA&gt;         0
##   71     Chr1 [11649, 11863]      * |   TAIR10 three_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##   72     Chr1 [11649, 13173]      * |   TAIR10            exon      &lt;NA&gt;      &lt;NA&gt;
##                       ID        Name            Note                          Parent       Index
##              &lt;character&gt; &lt;character&gt; &lt;CharacterList&gt;                 &lt;CharacterList&gt; &lt;character&gt;
##    2         AT1G01010.1 AT1G01010.1                                       AT1G01010           1
##    3         AT1G01010.1 AT1G01010.1                                       AT1G01010           1
##    4 AT1G01010.1-Protein AT1G01010.1                                                        &lt;NA&gt;
##    5                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##    6                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01010.1        &lt;NA&gt;
##   ..                 ...         ...             ...                             ...         ...
##   68                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;
##   69                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;
##   70                &lt;NA&gt;        &lt;NA&gt;                 AT1G01030.1,AT1G01030.1-Protein        &lt;NA&gt;
##   71                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;
##   72                &lt;NA&gt;        &lt;NA&gt;                                     AT1G01030.1        &lt;NA&gt;
##      Derives_from
##       &lt;character&gt;
##    2         &lt;NA&gt;
##    3         &lt;NA&gt;
##    4  AT1G01010.1
##    5         &lt;NA&gt;
##    6         &lt;NA&gt;
##   ..          ...
##   68         &lt;NA&gt;
##   69         &lt;NA&gt;
##   70         &lt;NA&gt;
##   71         &lt;NA&gt;
##   72         &lt;NA&gt;
## 
## ...
## &lt;6 more elements&gt;
## -------
## seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">unlist</span><span class="p">(</span><span class="n">sp</span><span class="p">)</span><span class="w"> </span><span class="c1"># Returns data as GRanges object
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRanges object with 442 ranges and 10 metadata columns:
##           seqnames         ranges strand |   source           type     score     phase
##              &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;       &lt;factor&gt; &lt;numeric&gt; &lt;integer&gt;
##       1.2     Chr1   [3631, 5899]      * |   TAIR10           mRNA      &lt;NA&gt;      &lt;NA&gt;
##       2.3     Chr1   [3631, 5899]      * |   TAIR10           mRNA      &lt;NA&gt;      &lt;NA&gt;
##       3.4     Chr1   [3760, 5630]      * |   TAIR10        protein      &lt;NA&gt;      &lt;NA&gt;
##       4.5     Chr1   [3631, 3913]      * |   TAIR10           exon      &lt;NA&gt;      &lt;NA&gt;
##       5.6     Chr1   [3631, 3759]      * |   TAIR10 five_prime_UTR      &lt;NA&gt;      &lt;NA&gt;
##       ...      ...            ...    ... .      ...            ...       ...       ...
##   438.445     ChrM [11918, 12241]      * |   TAIR10           gene      &lt;NA&gt;      &lt;NA&gt;
##   439.446     ChrM [11918, 12241]      * |   TAIR10           mRNA      &lt;NA&gt;      &lt;NA&gt;
##   440.447     ChrM [11918, 12241]      * |   TAIR10        protein      &lt;NA&gt;      &lt;NA&gt;
##   441.448     ChrM [11918, 12241]      * |   TAIR10           exon      &lt;NA&gt;      &lt;NA&gt;
##   442.449     ChrM [11918, 12241]      * |   TAIR10            CDS      &lt;NA&gt;         0
##                            ID        Name                Note                          Parent
##                   &lt;character&gt; &lt;character&gt;     &lt;CharacterList&gt;                 &lt;CharacterList&gt;
##       1.2         AT1G01010.1 AT1G01010.1                                           AT1G01010
##       2.3         AT1G01010.1 AT1G01010.1                                           AT1G01010
##       3.4 AT1G01010.1-Protein AT1G01010.1                                                    
##       4.5                &lt;NA&gt;        &lt;NA&gt;                                         AT1G01010.1
##       5.6                &lt;NA&gt;        &lt;NA&gt;                                         AT1G01010.1
##       ...                 ...         ...                 ...                             ...
##   438.445           ATMG00030   ATMG00030 protein_coding_gene                                
##   439.446         ATMG00030.1 ATMG00030.1                                           ATMG00030
##   440.447 ATMG00030.1-Protein ATMG00030.1                                                    
##   441.448                &lt;NA&gt;        &lt;NA&gt;                                         ATMG00030.1
##   442.449                &lt;NA&gt;        &lt;NA&gt;                     ATMG00030.1,ATMG00030.1-Protein
##                 Index Derives_from
##           &lt;character&gt;  &lt;character&gt;
##       1.2           1         &lt;NA&gt;
##       2.3           1         &lt;NA&gt;
##       3.4        &lt;NA&gt;  AT1G01010.1
##       4.5        &lt;NA&gt;         &lt;NA&gt;
##       5.6        &lt;NA&gt;         &lt;NA&gt;
##       ...         ...          ...
##   438.445        &lt;NA&gt;         &lt;NA&gt;
##   439.446           1         &lt;NA&gt;
##   440.447        &lt;NA&gt;  ATMG00030.1
##   441.448        &lt;NA&gt;         &lt;NA&gt;
##   442.449        &lt;NA&gt;         &lt;NA&gt;
##   -------
##   seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">sp</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="s2">"type"</span><span class="p">]</span><span class="w"> </span><span class="c1"># Subsetting of GRangesList objects is similar to GRanges objects.
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## GRangesList object of length 4:
## $1 
## GRanges object with 1 range and 1 metadata column:
##     seqnames       ranges strand |     type
##        &lt;Rle&gt;    &lt;IRanges&gt;  &lt;Rle&gt; | &lt;factor&gt;
##   2     Chr1 [3631, 5899]      * |     mRNA
## 
## $2 
## GRanges object with 1 range and 1 metadata column:
##     seqnames       ranges strand | type
##   3     Chr1 [3631, 5899]      * | mRNA
## 
## $3 
## GRanges object with 1 range and 1 metadata column:
##     seqnames       ranges strand |    type
##   4     Chr1 [3760, 5630]      * | protein
## 
## ...
## &lt;1 more element&gt;
## -------
## seqinfo: 7 sequences from an unspecified genome
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">lapply</span><span class="p">(</span><span class="n">sp</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span><span class="w"> </span><span class="n">length</span><span class="p">)</span><span class="w"> </span><span class="c1"># Looping over GRangesList objects similar to lists
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## $`1`
## [1] 1
## 
## $`2`
## [1] 1
## 
## $`3`
## [1] 1
## 
## $`4`
## [1] 1
</code></pre>
</div>

<p><br /><br /><center><a href="mydoc_Rsequences_05.html"><img src="images/left_arrow.png" alt="Previous page." /></a>Previous Page &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Next Page
<a href="mydoc_Rsequences_07.html"><img src="images/right_arrow.png" alt="Next page." /></a></center></p>


    <div class="tags">
        
    </div>

    

</div>

<hr class="shaded"/>

<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2017 UC Riverside. All rights reserved. <br />
<span>Page last updated:</span> Wed Jun  7 19:39:39 2017<br/> Site last generated: Oct 12, 2017 <br />
<!--
<p><img src="images/company_logo.png" alt="Company logo"/></p>
-->
</div>
            </div>
</footer>


    </div>
    <!-- /.row -->
</div>
<!-- /.container -->
    </div>

</body>

</html>
