<!DOCTYPE html>
<head>
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>Introduction to HPCC Cluster and Linux | systemPipeR Website</title>
<link rel="stylesheet" href="css/syntax.css">


<link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css">-->
<link rel="stylesheet" href="css/modern-business.css">
<link rel="stylesheet" href="css/lavish-bootstrap.css">
<link rel="stylesheet" href="css/customstyles.css">
<link rel="stylesheet" href="css/theme-blue.css">

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="js/jquery.navgoco.min.js"></script>


<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="js/toc.js"></script>
<script src="js/customscripts.js"></script>

<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

<link rel="shortcut icon" href="images/favicon.ico">

<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->

<link rel="alternate" type="application/rss+xml" title="" href="http://localhost:4000feed.xml">

    <script>
        $(document).ready(function() {
            // Initialize navgoco with default options
            $("#mysidebar").navgoco({
                caretHtml: '',
                accordion: true,
                openClass: 'active', // open
                save: false, // leave false or nav highlighting doesn't work right
                cookie: {
                    name: 'navgoco',
                    expires: false,
                    path: '/'
                },
                slide: {
                    duration: 400,
                    easing: 'swing'
                }
            });

            $("#collapseAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', false);
            });

            $("#expandAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', true);
            });

        });

    </script>
    <script>
        $(function () {
            $('[data-toggle="tooltip"]').tooltip()
        })
    </script>
    <script>
        $(document).ready(function() {
            $("#tg-sb-link").click(function() {
                $("#tg-sb-sidebar").toggle();
                $("#tg-sb-content").toggleClass('col-md-9');
                $("#tg-sb-content").toggleClass('col-md-12');
                $("#tg-sb-icon").toggleClass('fa-toggle-on');
                $("#tg-sb-icon").toggleClass('fa-toggle-off');
            });
        });
    </script>
    

</head>
<body>
<!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="fa fa-home fa-lg navbar-brand" href="index.html">&nbsp;<span class="projectTitle"> systemPipeR Project Website</span></a>
        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- toggle sidebar button -->
                <li><a id="tg-sb-link" href="#"><i id="tg-sb-icon" class="fa fa-toggle-on"></i> Nav</a></li>
                <!-- entries without drop-downs appear here -->
                
                
                
                <li><a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html" target="_blank">Bioc</a></li>
                
                
                
                <li><a href="https://github.com/tgirke/systemPipeR" target="_blank">GitHub Repo</a></li>
                
                
                
                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        <li><a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual" target="_blank">R & Bioconductor Manual</a></li>
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        <li><a href="http://hpcc.ucr.edu/" target="_blank">HPC Center (HPCC)</a></li>
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                    </ul>
                </li>
                
                
                
			<li>



  <a class="email" title="Submit feedback" href="#" onclick="javascript:window.location='mailto:thomas.girke@ucr.edu?subject=systemPipeR Website feedback&body=I have some feedback about the Introduction to HPCC Cluster and Linux page: ' + window.location.href;"><i class="fa fa-envelope-o"></i> Feedback</a>

</li>

		
                <!--comment out this block if you want to hide search-->
                <li>
                    <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                            SimpleJekyllSearch.init({
                                searchInput: document.getElementById('search-input'),
                                resultsContainer: document.getElementById('results-container'),
                                dataSource: 'search.json',
                                searchResultTemplate: '<li><a href="{url}" title="Introduction to HPCC Cluster and Linux">{title}</a></li>',
                    noResultsText: 'No results found.',
                            limit: 10,
                            fuzzy: true,
                    })
                    </script>
                    <!--end search-->
                </li>
            </ul>
        </div>
        </div>
        <!-- /.container -->
</nav>

<!-- Page Content -->
<div class="container">
    <div class="col-lg-12">&nbsp;</div>
    <!-- Content Row -->
    <div class="row">
        <!-- Sidebar Column -->
        <div class="col-md-3" id="tg-sb-sidebar">

          












<ul id="mysidebar" class="nav">
    <li class="sidebarTitle"><center><font color="#26466D" size="6">systemPipeR</font></center> <center><font color="#8b9299" size="3">...</font></center></li>
    
    
    
        
    
    <li>
        <a href="#">Home</a>
        <ul>
            
            
            
            <li><a href="index.html">Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_syllabus.html">Syllabus</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Internal</a>
        <ul>
            
            
            
            <li><a href="mydoc_internal.html">Internal Resources</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Schedule</a>
        <ul>
            
            
            
            <li><a href="mydoc_schedule.html">Course Schedule</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Lecture Slides</a>
        <ul>
            
            
            
            <li><a href="mydoc_slides_01.html">Course Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_02.html">Molbiol Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_03.html">Databases and Software</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_04.html">DNA Sequencing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_05.html">Introduction to R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_06.html">Sequence Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_07.html">Programming in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_08.html">Multiple Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_09.html">Short Read Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_10.html">NGS Analysis Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_11.html">Gene Expression Analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_12.html">NGS Workflows</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_13.html">ChIP-Seq Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_14.html">VAR-Seq Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_15.html">Sequence Assembly</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_16.html">Gene Annotation and Ontologies</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_17.html">Cluster Analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_18.html">Profile HMMs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_19.html">Introduction to Phylogenetics</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Homework</a>
        <ul>
            
            
            
            <li><a href="mydoc_homework_01.html">HW1 - Online Tools</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_02.html">HW2 - Biocluster/Linux</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_03.html">HW3 - Introduction to R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_04.html">HW4 - Pairwise Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_09.html">HW5 - Programming in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_08.html">HW6 - Sequence Analysis</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">GitHub</a>
        <ul>
            
            
            
            <li><a href="mydoc_tutorial_01.html">Table of Content</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#github-in-gen242">GitHub in GEN242</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#installing-git">Installation</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#git-basics-from-command-line">Git from Command-line</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#github-basics-from-command-line">GitHub from Command-line</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#using-github-from-rstudio">GitHub from RStudio</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Linux & HPCC Cluster</a>
        <ul>
            
            
            
            <li class="active"><a href="mydoc_tutorial_02.html">Table of Content</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#hpcc-cluster-overview">HPCC Cluster</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#linux-basics">Linux Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#software-and-module-system">Software & Module System</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#big-data-storage">Big Data Storage</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#queuing-system-slurm">Queueing System Slurm</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#textcode-editors">Text Editors</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#vim-r-tmux-essentials">Vim-R-Tmux</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#nvim-r-tmux">Nvim-R-Tmux</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Introduction to R</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rbasics_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_02.html">2. R Package Repositories</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_03.html">3. Installation of R Packages</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_04.html">4. Getting Around</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_05.html">5. Basic Syntax</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_06.html">6. Data Types</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_07.html">7. Data Objects</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_08.html">8. Important Utilities</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_09.html">9. Operators and Calculations</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_10.html">10. Reading and Writing External Data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_11.html">11. Useful R Functions</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_12.html">12. dplyr Environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_13.html">13. SQLite Databases</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_14.html">14. Graphics in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_15.html">15. Analysis Routine</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_16.html">16. R Markdown</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_17.html">17. Shiny Web Apps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_18.html">18. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_19.html">19. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Programming in R</a>
        <ul>
            
            
            
            <li><a href="mydoc_Programming_in_R_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_02.html">2. Control Structures</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_03.html">3. Loops</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_04.html">4. Functions</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_05.html">5. Useful Utilities</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_06.html">6. Running R Scripts</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_07.html">7. Building R Packages</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_08.html">8. Programming Exercises</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_09.html">9. Homework 5</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_10.html">10. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Analysis Basics</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rsequences_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_02.html">2. Package Requirements</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_03.html">3. Strings in R Base</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_04.html">4. Sequences in Bioconductor</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_05.html">5. NGS Sequences</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_06.html">6. Range Operations</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_07.html">7. Transcript Ranges</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_08.html">8. Homework 6</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_09.html">9. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_10.html">10. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Workflow Overview</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeR_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_2.html">2. Getting Started</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_3.html">3. Workflow overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_4.html">4. Workflow templates</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_5.html">5. Version information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">RNA-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_02.html">2. Samples and environment settings</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_05.html">5. Read quantification</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_06.html">6. Analysis of DEGs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_07.html">7. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_08.html">8. Clustering and heat maps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_09.html">9. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_10.html">10. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">ChIP-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_05.html">5. Utilities for coverage data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_06.html">6. Peak calling with MACS2</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_07.html">7. Annotate peaks with genomic context</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_08.html">8. Count reads overlapping peaks</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_09.html">9. Differential binding analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_10.html">10. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_11.html">11. Motif analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">VAR-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeVARseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_05.html">5. Variant calling</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_06.html">6. Filter variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_07.html">7. Annotate filtered variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_08.html">8. Combine annotation results among samples</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_09.html">9. Summary statistics of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_10.html">10. Venn diagram of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_11.html">11. Plot variants programmatically</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Graphics and Visualization</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rgraphics_1.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_2.html">2. Base Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_3.html">3. Grid Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_4.html">4. lattice Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_5.html">5. ggplot2 Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_6.html">6. Specialty Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_7.html">7. Genome Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_8.html">8. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Cluster Analysis</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rclustering_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_2.html">2. Data Preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_3.html">3. Clustering Algorithms</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_4.html">4. Clustering Exercises</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_5.html">5. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Paper Presentations</a>
        <ul>
            
            
            
            <li><a href="mydoc_paper_presentations.html">Paper Presentations</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Course Projects</a>
        <ul>
            
            
            
            <li><a href="mydoc_project_intro.html">Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_chipseq1.html">ChIP-Seq1 - Motif Prediction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_chipseq2.html">ChIP-Seq2 - Peak Callers</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_rnaseq1.html">RNA-Seq1 - Aligners</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_rnaseq2.html">RNA-Seq2 - DEG Methods</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_varseq1.html">VAR-Seq1 - Functional Prediction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_data.html">Project Data</a></li>
            
            
            
            
        </ul>
        
        
        
        <!-- if you aren't using the accordion, uncomment this block:
           <p class="external">
               <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
           </p>
           -->
    </li>
</ul>
</div>

<!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
<script>$("li.active").parents('li').toggleClass("active");</script>

    <!-- Content Column -->
    <div class="col-md-9" id="tg-sb-content">
        <div class="post-header">
   <h1 class="post-title-main">Introduction to HPCC Cluster and Linux</h1>
</div>



<div class="post-content">

   

    
    
<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>

<div id="toc"></div>

    


<!--
    

    <a target="_blank" href="https://github.com/tgirke/systemPipeR/blob/gh-pages/pages//mydoc_tutorial_02.md" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
-->
    

  <h2 id="hpcc-cluster-overview">HPCC Cluster Overview</h2>

<p>The HPCC Cluster (formerly called biocluster) is a shared research computing system available at UCR. The HPCC website is available <a href="http://hpcc.ucr.edu/index.html">here</a>.</p>

<h3 id="what-is-a-computer-cluster">What Is a Computer Cluster?</h3>

<ul>
  <li>
    <p>A computer cluster is an assembly of CPU units, so called computer nodes that work together to perform many computations in parallel. To achieve this, an internal network (e.g. Infiniband interconnect) connects the nodes to a larger unit, while a head node controls the load and traffic across the entire system.</p>
  </li>
  <li>
    <p>Usually, users log into the head node to submit their computer requests via <code class="highlighter-rouge">srun</code> to a queuing system provided by resource management and scheduling software, such as SGE, Slurm or TORQUE/MAUI. The queuing system distributes the processes to the computer nodes in a controlled fashion.</p>
  </li>
  <li>
    <p>Because the head node controls the entire system, users should never run computing jobs on the head node directly!</p>
  </li>
  <li>
    <p>For code testing purposes, one can log into one of the nodes with <code class="highlighter-rouge">srun --pty bash -l</code> and run jobs interactively. Alternatively, one can log into the test node owl via ssh.</p>
  </li>
</ul>

<h3 id="hardware-infrastructure">Hardware Infrastructure</h3>

<h3 id="computer-nodes">Computer nodes</h3>

<ul>
  <li>Over 4,500 CPU cores</li>
  <li>48 AMD computer nodes, each with 64 CPU cores and 512GB RAM</li>
  <li>40 Intel computer nodes, each with 32 CPU cores and 512GB RAM</li>
  <li>6 high-memory nodes, each 32 CPU cores and 1024GB RAM</li>
  <li>12 GPU nodes, each with 5,000 cuda cores</li>
</ul>

<h3 id="interconnect">Interconnect</h3>
<ul>
  <li>FDR IB @56Gbs</li>
</ul>

<h3 id="storage">Storage</h3>

<ul>
  <li>Parallel GPFS storage system with 2.1 PB usable space</li>
  <li>Backup of same architecture and similar amount</li>
</ul>

<h3 id="user-traffic">User traffic</h3>

<ul>
  <li>Computing tasks need to be submitted via <code class="highlighter-rouge">srun</code></li>
  <li>HPCC Cluster headnode only for login, not for computing tasks!</li>
  <li>Monitor cluster activity: <code class="highlighter-rouge">squeue</code> or <code class="highlighter-rouge">jobMonitor</code> (<code class="highlighter-rouge">qstatMonitor</code>)</li>
</ul>

<h3 id="manuals">Manuals</h3>

<ul>
  <li><a href="http://hpcc.ucr.edu/manuals_linux-cluster_intro.html">HPCC Cluster Manual</a></li>
  <li><a href="http://hpcc.ucr.edu/manuals_linux-basics.html">Linux Manual</a></li>
</ul>

<h2 id="linux-basics">Linux Basics</h2>

<h3 id="log-into-hpcc-cluster">Log into HPCC Cluster</h3>

<ul>
  <li>Login command on OS X or Linux</li>
</ul>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code>ssh -XY user@biocluster.ucr.edu
</code></pre>
</div>

<p>Type password</p>

<ul>
  <li>
    <p>Windows: provide same information in a terminal application like <a href="http://www.chiark.greenend.org.uk/~sgtatham/putty/download.html">Putty</a> or <a href="http://mobaxterm.mobatek.net/">MobaXterm</a>.</p>

    <ul>
      <li>Host name: <code class="highlighter-rouge">biocluster.ucr.edu</code></li>
      <li>User name: …</li>
      <li>Password: …</li>
    </ul>
  </li>
</ul>

<h3 id="important-linux-commands">Important Linux Commands</h3>

<p>Finding help</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>man &lt;program_name&gt;
</code></pre>
</div>

<p>List content of current directory</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>ls
</code></pre>
</div>

<p>Print current working directory</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="nb">pwd</span>
</code></pre>
</div>

<p>Search in files and directories</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>grep
</code></pre>
</div>

<p>Word count</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>wc
</code></pre>
</div>

<p>Create directory</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>mkdir
</code></pre>
</div>

<p>Delete files and directories</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>rm
</code></pre>
</div>

<p>Move and rename files</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>mv
</code></pre>
</div>

<p>Copy files from internet to <code class="highlighter-rouge">pwd</code></p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>wget
</code></pre>
</div>

<p>Viewing files</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>less
</code></pre>
</div>

<h3 id="file-exchange">File Exchange</h3>

<p><strong>GUI applications</strong></p>

<ul>
  <li>Windows: <a href="http://winscp.net/eng/index.php">WinSCP</a></li>
  <li>Mac OS X: <a href="http://cyberduck.en.softonic.com/mac">CyberDuck</a></li>
  <li>Win/OS X/Linux: <a href="https://filezilla-project.org/">FileZilla</a></li>
</ul>

<p><strong>SCP command-line tool</strong></p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code>scp file user@remotehost:/home/user/ <span class="c"># From local to remote </span>
scp user@remotehost:/home/user/file . <span class="c"># From remote to local </span>
</code></pre>
</div>

<h3 id="std-inouterr-redirect--wildcards">STD IN/OUT/ERR, Redirect &amp; Wildcards</h3>

<p>Wildcard <code class="highlighter-rouge">*</code> to specify many files</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>file.<span class="k">*</span>                        
</code></pre>
</div>
<p>Redirect <code class="highlighter-rouge">ls</code> output to file</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>ls &gt; file                     
</code></pre>
</div>
<p>Specify file as input to command</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="nb">command</span> &lt; myfile              
</code></pre>
</div>
<p>Append output of command to file</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="nb">command</span> &gt;&gt; myfile             
</code></pre>
</div>
<p>Pipe <code class="highlighter-rouge">STDOUT</code> of one command to another command</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>command1 | command2     
</code></pre>
</div>
<p>Turn off progress info</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="nb">command</span> &gt; /dev/null 
</code></pre>
</div>
<p>Pipe output of <code class="highlighter-rouge">grep</code> to <code class="highlighter-rouge">wc</code></p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>grep pattern file | wc        
</code></pre>
</div>
<p>Print <code class="highlighter-rouge">STDERR</code> to file</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>grep pattern nonexistingfile 2 &gt; mystderr 
</code></pre>
</div>

<h3 id="homework-assignment-hw2">Homework Assignment (HW2)</h3>

<p>See HW2 page <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc_homework_02.html">here</a>.</p>

<h3 id="permissions-and-ownership">Permissions and ownership</h3>

<p>List directories and files</p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code>ls -al 
</code></pre>
</div>

<p>The previous command shows something like this for each file/dir: <code class="highlighter-rouge">drwxrwxrwx</code>. The 
meaning of this syntax is as follows:</p>

<ul>
  <li><code class="highlighter-rouge">d</code>: directory</li>
  <li><code class="highlighter-rouge">rwx</code>: read, write and execute permissions, respectively
    <ul>
      <li>first triplet: user permissions (u)</li>
      <li>second triplet: group permissions (g)</li>
      <li>third triplet: world permissions (o)</li>
    </ul>
  </li>
</ul>

<p>Example for assigning write and execute permissions to user, group and world</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>chmod ugo+rx my_file
</code></pre>
</div>

<ul>
  <li><code class="highlighter-rouge">+</code> causes the permissions selected to be added</li>
  <li><code class="highlighter-rouge">-</code> causes them to be removed</li>
  <li><code class="highlighter-rouge">=</code> causes them to be the only permissions that the file has.</li>
</ul>

<p>Syntax for changing user &amp; group ownership</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>chown &lt;user&gt;:&lt;group&gt; &lt;file or dir&gt; 
</code></pre>
</div>

<h3 id="symbolic-links">Symbolic Links</h3>
<p>Symbolic links are short nicknames to files and directories that save typing of their full paths.</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>ln -s original_filename new_nickname
</code></pre>
</div>

<h2 id="software-and-module-system">Software and module system</h2>

<ul>
  <li>Over 750 software tools are currently installed on HPCC Cluster</li>
  <li>Most common research databases used in bioinformatics are available</li>
  <li>Support of most common programming languages used in research computing</li>
  <li>A module system is used to facilitate the management of software tools. This includes any number of versions of each software.</li>
  <li>New software install requests can be sent to support@biocluster.ucr.edu.</li>
  <li>To use software manged under the module system, users need to learn using some basic commands. The most common commands are listed below.</li>
</ul>

<p>Print available modules</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>module avail
</code></pre>
</div>

<p>Print available modules starting with R</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>module avail R
</code></pre>
</div>

<p>Load default module R</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>module load R
</code></pre>
</div>

<p>Load specific module R version</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>module load R/3.2.2
</code></pre>
</div>

<p>List loaded modules</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>module list
</code></pre>
</div>

<p>Unload module R</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>module unload R
</code></pre>
</div>

<p>Unload specific module R</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>module unload R/3.2.3-dev
</code></pre>
</div>

<h2 id="big-data-storage">Big data storage</h2>

<p>Each user account on HPCC Cluster comes only with 20GB of disk space. Much more disk space is 
available in a dedicated <code class="highlighter-rouge">bigdata</code> directory. How much space depends on the subscription 
of each user group. The path of <code class="highlighter-rouge">bigdata</code> and <code class="highlighter-rouge">bigdata-shared</code> is as follows:</p>

<ul>
  <li><code class="highlighter-rouge">/bigdata/labname/username</code></li>
  <li><code class="highlighter-rouge">/bigdata/labname/shared</code></li>
</ul>

<p>All lab members share the same bigdata pool. The course number <code class="highlighter-rouge">gen242</code> is used as <code class="highlighter-rouge">labname</code>
for user accounts adminstered under GEN242.</p>

<p>The disk usage of <code class="highlighter-rouge">home</code> and <code class="highlighter-rouge">bigdata</code> can be monitored on the <a href="https://dashboard.bioinfo.ucr.edu/">HPCC Cluster Dashboard</a>.</p>

<h2 id="queuing-system-slurm">Queuing system: <code class="highlighter-rouge">Slurm</code></h2>

<p>HPCC Cluster uses <code class="highlighter-rouge">Slurm</code> as queuing and load balancing system. To control user traffic, any 
type of compute intensive jobs need to be submitted via the <code class="highlighter-rouge">sbatch</code> or <code class="highlighter-rouge">srun</code> (see below) to the computer
nodes. Much more detailed information on this topic can be found on these sites:</p>

<ul>
  <li><a href="http://hpcc.ucr.edu/manuals_linux-cluster_jobs.html">UCR HPCC Manual</a></li>
  <li><a href="https://slurm.schedmd.com/documentation.html">Slurm Documentation</a></li>
  <li><a href="http://www.nersc.gov/users/computational-systems/cori/running-jobs/for-edison-users/torque-moab-vs-slurm-comparisons/">Torque/Slurm Comparison</a></li>
  <li><a href="https://sites.google.com/a/case.edu/hpc-upgraded-cluster/slurm-cluster-commands">Switching from Torque to Slurm</a></li>
  <li><a href="http://www.ceci-hpc.be/slurm_tutorial.html">Slurm Quick Start Tutorial</a></li>
</ul>

<h3 id="job-submission-with-sbatch">Job submission with <code class="highlighter-rouge">sbatch</code></h3>

<p>Print information about queues/partitions available on a cluster.</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>sinfo
</code></pre>
</div>

<p>Compute jobs are submitted <code class="highlighter-rouge">sbatch</code> via a submission script (here <code class="highlighter-rouge">script_name.sh</code>).</p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code>sbatch script_name.sh
</code></pre>
</div>

<p>The following sample submission script (<code class="highlighter-rouge">script_name.sh</code>) executes an R script named <code class="highlighter-rouge">my_script.R</code>.</p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code><span class="c">#!/bin/bash -l</span>

<span class="c">#SBATCH --nodes=1</span>
<span class="c">#SBATCH --ntasks=1</span>
<span class="c">#SBATCH --cpus-per-task=1</span>
<span class="c">#SBATCH --mem-per-cpu=1G</span>
<span class="c">#SBATCH --time=1-00:15:00 # 1 day and 15 minutes</span>
<span class="c">#SBATCH --mail-user=useremail@address.com</span>
<span class="c">#SBATCH --mail-type=ALL</span>
<span class="c">#SBATCH --job-name="some_test"</span>
<span class="c">#SBATCH -p batch # Choose queue/parition from: intel, batch, highmem, gpu, short</span>

Rscript my_script.R
</code></pre>
</div>

<p>Interactive session: logs user into node</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>srun --pty bash -l
</code></pre>
</div>

<p>Interactive session with specific resource requests</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>srun --x11 --partition<span class="o">=</span>short --mem<span class="o">=</span>2gb --cpus-per-task 4 --ntasks 1 --time 1:00:00 --pty bash -l
</code></pre>
</div>

<p><code class="highlighter-rouge">STDOUT</code> and <code class="highlighter-rouge">STDERROR</code> of jobs will be written to files named <code class="highlighter-rouge">slurm-&lt;jobid&gt;.out</code> or to custom a file specified under <code class="highlighter-rouge">#SBATCH --output</code> in the submission script.</p>

<h3 id="monitoring-jobs-with-squeue">Monitoring jobs with <code class="highlighter-rouge">squeue</code></h3>

<p>List all jobs in queue</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>squeue
</code></pre>
</div>

<p>List jobs of a specific user</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>squeue -u &lt;user&gt;
</code></pre>
</div>

<p>Print more detailed information about a job</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>scontrol show job &lt;JOBID&gt;
</code></pre>
</div>

<p>Custom command to summarize and visualize cluster activity</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>jobMonitor
</code></pre>
</div>

<h3 id="deleting-and-altering-jobs">Deleting and altering jobs</h3>

<p>Delete a single job</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>scancel -i &lt;JOBID&gt;
</code></pre>
</div>

<p>Delete all jobs of a user</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>scancel -u &lt;username&gt; 
</code></pre>
</div>

<p>Delete all jobs of a certain name</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>scancel --name &lt;myJobName&gt;
</code></pre>
</div>

<p>Altering jobs with <code class="highlighter-rouge">scontrol update</code>. The below example changes the walltime (<code class="highlighter-rouge">&lt;NEW_TIME&gt;</code>) of a specific job (<code class="highlighter-rouge">&lt;JOBID&gt;</code>).</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>scontrol update <span class="nv">jobid</span><span class="o">=</span>&lt;JOBID&gt; <span class="nv">TimeLimit</span><span class="o">=</span>&lt;NEW_TIME&gt;
</code></pre>
</div>

<h3 id="resource-limits">Resource limits</h3>

<p>Resourse limits for users can be viewed as follows.</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>sacctmgr show account <span class="nv">$GROUP</span> <span class="nv">format</span><span class="o">=</span>Account,User,Partition,GrpCPUs,GrpMem,GrpNodes --ass | grep <span class="nv">$USER</span>
</code></pre>
</div>

<p>Similarly, one can view the limits of the group a user belongs to.</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>sacctmgr show account <span class="nv">$GROUP</span> <span class="nv">format</span><span class="o">=</span>Account,User,Partition,GrpCPUs,GrpMem,GrpNodes,GrpTRES%30 --ass | head -3
</code></pre>
</div>

<h2 id="textcode-editors">Text/code editors</h2>

<p>The following list includes examples of several widely used code editors.</p>

<ul>
  <li><strong>Vi/Vim/Neovim</strong>: Non-graphical (terminal-based) editor. Vi is guaranteed to be available on any system. Vim is the improved version of vi.</li>
  <li><strong>Emacs</strong>: Non-graphical or window-based editor. You still need to know keystroke commands to use it. Installed on all Linux distributions and on most other Unix systems.</li>
  <li><strong>Pico</strong>: Simple terminal-based editor available on most versions of Unix. Uses keystroke commands, but they are listed in logical fashion at bottom of screen.</li>
  <li><strong>Nano</strong>: A simple terminal-based editor which is default on modern Debian systems.</li>
  <li><strong>Atom</strong>: Modern text editor developed by GitHub project.</li>
</ul>

<h3 id="why-does-it-matter">Why does it matter?</h3>

<p>To work efficiently on remote systems like a computer cluster, it is essential
to learn how to work in a pure command-line interface. GUI environments like
RStudio and similar coding environments are not suitable for this. In addition,
there is a lot of value of knowing how to work in an environment that is not
restricted to a specific programming language. Therefore, this class embraces
RStudio where it is useful, but for working on remote systems like HPCC Cluster, it 
uses Vim and Tmux. Both are useful for many programming languages.
Combinded with the <code class="highlighter-rouge">vim-r</code> plugin they also provide a powerful command-line working
environment for R. The following provides a brief introduction to this environment.</p>

<h3 id="vim-overview">Vim overview</h3>

<p>The following opens a file (here <code class="highlighter-rouge">myfile</code>) with vim</p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>vim myfile.txt
</code></pre>
</div>

<p>Once you are in Vim the most important commands are:</p>

<ul>
  <li><code class="highlighter-rouge">i</code>: The <code class="highlighter-rouge">i</code> key brings you into the insert mode for typing.</li>
  <li><code class="highlighter-rouge">Esc</code>: The <code class="highlighter-rouge">Esc</code> key brings you out of the insert mode.</li>
  <li><code class="highlighter-rouge">:</code>: The <code class="highlighter-rouge">:</code> key starts the command mode at the bottom of the screen.</li>
</ul>

<p>Use the arrow keys to move your cursor in the text. Using <code class="highlighter-rouge">Fn Up/Down key</code> allows to page through
the text quicker. In the following command overview, all commands starting with <code class="highlighter-rouge">:</code> need to be typed in the command mode. 
All other commands are typed in the normal mode after pushing the <code class="highlighter-rouge">Esc</code> key.</p>

<p>Important modifier keys to control vim</p>

<ul>
  <li><code class="highlighter-rouge">:w</code>: save changes to file. If you are in editing mode you have to hit <code class="highlighter-rouge">Esc</code> first.</li>
  <li><code class="highlighter-rouge">:q</code>: quit file that has not been changed</li>
  <li><code class="highlighter-rouge">:wq</code>: save and quit file</li>
  <li><code class="highlighter-rouge">:!q</code>: quit file without saving any changes</li>
</ul>

<h3 id="useful-resources-for-learning-vim">Useful resources for learning vim</h3>

<ul>
  <li><a href="http://www.openvim.com">Interactive Vim Tutorial</a></li>
  <li><a href="http://vimdoc.sourceforge.net/">Official Vim Documentation</a></li>
  <li><a href="http://manuals.bioinformatics.ucr.edu/home/linux-basics#TOC-Vim-Manual">UCR Linux Manual</a></li>
</ul>

<h2 id="vim-r-tmux-essentials">Vim-R-Tmux essentials</h2>
<p>Terminal-based Working Environment for R: <a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r/vim-r">Vim-R-Tmux</a></p>

<center><img title="Vim-R-Tmux" src="pages/mydoc/Rbasics_files/screenshot.png" /></center>
<center>Vim-R-Tmux IDE for R</center>

<h3 id="basics">Basics</h3>

<p>Tmux is a terminal multiplexer that allows to split terminal windows and to detach/reattach to
existing terminal sessions. Combinded with the <code class="highlighter-rouge">vim-r</code> plugin it provides a powerful command-line working 
environment for R where users can send code from a script to the R console or command-line.
Both tmux and the <code class="highlighter-rouge">vim-r</code> plugin need to be installed on a system. On HPCC Cluster both are configured
in each user account. A detailed user manual is available <a href="http://manuals.bioinformatics.ucr.edu/home/programming-in-r/vim-r">here</a>.
The following gives a short introduction into the basic usage:</p>

<p><strong>1. Start tmux session</strong></p>
<div class="language-sh highlighter-rouge"><pre class="highlight"><code>tmux <span class="c"># starts a new tmux session </span>
tmux a <span class="c">#  attaches to an existing session </span>
</code></pre>
</div>

<p><strong>2. Open R script in vim</strong></p>

<p>This can be any of these file types: <code class="highlighter-rouge">*.R</code>, <code class="highlighter-rouge">*.Rnw</code> or <code class="highlighter-rouge">*.Rmd</code>.</p>

<div class="language-sh highlighter-rouge"><pre class="highlight"><code>vim myscript.R
</code></pre>
</div>

<p><strong>3. Open vim-connected R session by pressing the <code class="highlighter-rouge">F2</code> key</strong></p>

<p>This will open an R session in a separate tmux pane. Note, in the provided <code class="highlighter-rouge">.tmux.conf</code> file 
the command key binding has been reassigned from the tmux default <code class="highlighter-rouge">Ctrl-b</code> to <code class="highlighter-rouge">Ctrl-a</code>, 
and the shortcut for starting R from vim has been reassigned from <code class="highlighter-rouge">\rf</code> to <code class="highlighter-rouge">F2</code> in the <code class="highlighter-rouge">.vimrc</code> 
file. The command key binding <code class="highlighter-rouge">Ctrl-a</code> is the most important key sequence in order to move 
around in tmux. For instance, the key sequence <code class="highlighter-rouge">Ctrl-a o</code> will switch between the vim and R 
panes, and <code class="highlighter-rouge">Ctrl-a Ctrl-o</code> will swap the two panes.</p>

<p><strong>4. Send R code vim to the R pane</strong></p>

<p>Single lines of code can be sent from vim to the R console by pressing the space bar. To send 
several lines at once, one can select them in vim’s visual mode and then hit the space bar. 
Please note, the default command for sending code lines in the vim-r-plugin is <code class="highlighter-rouge">\l</code>. This key 
binding has been remapped in the provided <code class="highlighter-rouge">.vimrc</code> file to the space bar. Most other key 
bindings (shortcuts) still start with the <code class="highlighter-rouge">\</code> as LocalLeader, <em>e.g.</em> <code class="highlighter-rouge">\rh</code> opens the help for 
a function/object where the curser is located in vim. More details on this are given below.</p>

<h3 id="important-keybindings-for-vim">Important keybindings for vim</h3>

<ul>
  <li><code class="highlighter-rouge">&lt;F2&gt;</code>: opens vim-connected R session; remapped in <code class="highlighter-rouge">.vimrc</code> from default <code class="highlighter-rouge">\rf</code></li>
  <li><code class="highlighter-rouge">&lt;spacebar&gt;</code>: sends code from vim to R; here remapped in <code class="highlighter-rouge">.vimrc</code> from default <code class="highlighter-rouge">\l</code></li>
  <li><code class="highlighter-rouge">:split</code> or <code class="highlighter-rouge">:vsplit</code>: splits viewport (similar to pane split in tmux)</li>
  <li><code class="highlighter-rouge">Ctrl-w-w</code>: jumps cursor to next viewport</li>
  <li><code class="highlighter-rouge">Ctrl-w-r</code>: swaps viewports</li>
  <li><code class="highlighter-rouge">Ctrl-s</code> and <code class="highlighter-rouge">Ctrl-x</code>: freezes/unfreezes vim (some systems)</li>
</ul>

<h3 id="important-keybindings-for-tmux">Important keybindings for tmux</h3>

<p><strong>Pane-level commands</strong></p>

<ul>
  <li><code class="highlighter-rouge">Ctrl-a %</code>: splits pane vertically</li>
  <li><code class="highlighter-rouge">Ctrl-a "</code>: splits pane horizontally</li>
  <li><code class="highlighter-rouge">Ctrl-a o</code>: jumps cursor to next pane</li>
  <li><code class="highlighter-rouge">Ctrl-a Ctrl-o</code>: swaps panes</li>
  <li><code class="highlighter-rouge">Ctrl-a &lt;space bar&gt;</code>: rotates pane arrangement</li>
</ul>

<p><strong>Window-level comands</strong></p>

<ul>
  <li><code class="highlighter-rouge">Ctrl-a n</code>: switches to next tmux window</li>
  <li><code class="highlighter-rouge">Ctrl-a Ctrl-a</code>: switches to previous tmux window</li>
  <li><code class="highlighter-rouge">Ctrl-a c</code>: creates a new tmux window</li>
  <li><code class="highlighter-rouge">Ctrl-a 1</code>: switches to specific tmux window selected by number</li>
</ul>

<p><strong>Session-level comands</strong></p>

<ul>
  <li><code class="highlighter-rouge">Ctrl-a d</code>: detaches from current session</li>
  <li><code class="highlighter-rouge">Ctrl-a s</code>: switch between available tmux sesssions</li>
  <li><code class="highlighter-rouge">$ tmux new -s &lt;name&gt;</code>: starts new session with a specific name</li>
  <li><code class="highlighter-rouge">$ tmux ls</code>: lists available tmux session(s)</li>
  <li><code class="highlighter-rouge">$ tmux attach -t &lt;id&gt;</code>: attaches to specific tmux session</li>
  <li><code class="highlighter-rouge">$ tmux attach</code>: reattaches to session</li>
  <li><code class="highlighter-rouge">$ tmux kill-session -t &lt;id&gt;</code>: kills a specific tmux session</li>
  <li><code class="highlighter-rouge">Ctrl-a : kill-session</code>: kills a session from tmux command mode that can be initiated with <code class="highlighter-rouge">Ctrl-a :</code></li>
</ul>

<h2 id="nvim-r-tmux">Nvim-R-Tmux</h2>

<p>The new Nvim-R plugin also supports the latest extension of Vim called Neovim.
The main advantages of Neovim compared to Vim are its better performance and
its built-in terminal emulator facilitating the communication among Neovim and
interactive programming environments such as R. Since the Vim and Neovim
environments are managed independently, one can run them in parallel on the
same system without interfering with each other.</p>

<p>A short manual for Nvim-R-Tmux is available <a href="https://gist.github.com/tgirke/7a7c197b443243937f68c422e5471899">here</a>.</p>


    <div class="tags">
        
    </div>

    

</div>

<hr class="shaded"/>

<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2017 UC Riverside. All rights reserved. <br />
 Site last generated: Oct 12, 2017 <br />
<!--
<p><img src="images/company_logo.png" alt="Company logo"/></p>
-->
</div>
            </div>
</footer>


    </div>
    <!-- /.row -->
</div>
<!-- /.container -->
    </div>

</body>

</html>
