<!DOCTYPE html>
<head>
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>3. Workflow overview | systemPipeR Website</title>
<link rel="stylesheet" href="css/syntax.css">


<link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css">-->
<link rel="stylesheet" href="css/modern-business.css">
<link rel="stylesheet" href="css/lavish-bootstrap.css">
<link rel="stylesheet" href="css/customstyles.css">
<link rel="stylesheet" href="css/theme-red.css">

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="js/jquery.navgoco.min.js"></script>


<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="js/toc.js"></script>
<script src="js/customscripts.js"></script>

<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

<link rel="shortcut icon" href="images/favicon.ico">

<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->

<link rel="alternate" type="application/rss+xml" title="systemPipeR" href="http://localhost:4000feed.xml">

    <script>
        $(document).ready(function() {
            // Initialize navgoco with default options
            $("#mysidebar").navgoco({
                caretHtml: '',
                accordion: true,
                openClass: 'active', // open
                save: false, // leave false or nav highlighting doesn't work right
                cookie: {
                    name: 'navgoco',
                    expires: false,
                    path: '/'
                },
                slide: {
                    duration: 400,
                    easing: 'swing'
                }
            });

            $("#collapseAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', false);
            });

            $("#expandAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', true);
            });

        });

    </script>
    <script>
        $(function () {
            $('[data-toggle="tooltip"]').tooltip()
        })
    </script>
    <script>
        $(document).ready(function() {
            $("#tg-sb-link").click(function() {
                $("#tg-sb-sidebar").toggle();
                $("#tg-sb-content").toggleClass('col-md-9');
                $("#tg-sb-content").toggleClass('col-md-12');
                $("#tg-sb-icon").toggleClass('fa-toggle-on');
                $("#tg-sb-icon").toggleClass('fa-toggle-off');
            });
        });
    </script>
    

</head>
<body>
<!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="fa fa-home fa-lg navbar-brand" href="index.html">&nbsp;<span class="projectTitle"> Data Analysis Workflow Environment</span></a>
        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- toggle sidebar button -->
                <li><a id="tg-sb-link" href="#"><i id="tg-sb-icon" class="fa fa-toggle-on"></i> Nav</a></li>
                <!-- entries without drop-downs appear here -->
                
                
                
                <li><a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html" target="_blank">Bioc</a></li>
                
                
                
                <li><a href="https://github.com/tgirke/systemPipeR" target="_blank">GitHub Repo</a></li>
                
                
                
                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        <li><a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual" target="_blank">R & Bioconductor Manual</a></li>
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        <li><a href="http://hpcc.ucr.edu/" target="_blank">HPC Center (HPCC)</a></li>
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                    </ul>
                </li>
                
                
                
			<li>



  <a class="email" title="Submit feedback" href="#" onclick="javascript:window.location='mailto:thomas.girke@ucr.edu?subject=systemPipeR Website feedback&body=I have some feedback about the 3. Workflow overview page: ' + window.location.href;"><i class="fa fa-envelope-o"></i> Feedback</a>

</li>

		
                <!--comment out this block if you want to hide search-->
                <li>
                    <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                            SimpleJekyllSearch.init({
                                searchInput: document.getElementById('search-input'),
                                resultsContainer: document.getElementById('results-container'),
                                dataSource: 'search.json',
                                searchResultTemplate: '<li><a href="{url}" title="3. Workflow overview">{title}</a></li>',
                    noResultsText: 'No results found.',
                            limit: 10,
                            fuzzy: true,
                    })
                    </script>
                    <!--end search-->
                </li>
            </ul>
        </div>
        </div>
        <!-- /.container -->
</nav>

<!-- Page Content -->
<div class="container">
    <div class="col-lg-12">&nbsp;</div>
    <!-- Content Row -->
    <div class="row">
        <!-- Sidebar Column -->
        <div class="col-md-3" id="tg-sb-sidebar">

          












<ul id="mysidebar" class="nav">
    <li class="sidebarTitle"><center><font color="#7a1f1f" size="6"><i>systemPipeR</i></font></center> <center><font color="#8b9299" size="3">Project Site</font></center></li>
    
    
    
        
    
    <li>
        <a href="#">Home</a>
        <ul>
            
            
            
            <li><a href="index.html">Introduction</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Workflow Overview</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeR_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_2.html">2. Getting Started</a></li>
            
            
            
            
            
            
            <li class="active"><a href="mydoc_systemPipeR_3.html">3. Workflow overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_4.html">4. Workflow templates</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_5.html">5. Version information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">RNA-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_02.html">2. Samples and environment settings</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_05.html">5. Read quantification</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_06.html">6. Analysis of DEGs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_07.html">7. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_08.html">8. Clustering and heat maps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_09.html">9. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_10.html">10. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">ChIP-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_05.html">5. Utilities for coverage data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_06.html">6. Peak calling with MACS2</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_07.html">7. Annotate peaks with genomic context</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_08.html">8. Count reads overlapping peaks</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_09.html">9. Differential binding analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_10.html">10. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_11.html">11. Motif analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">RIBO-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_02.html">2. Load workflow template</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_05.html">5. Read distribution across genomic features</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_06.html">6. Adding custom features to workflow</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_07.html">7. Genomic read coverage along transripts or CDSs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_08.html">8. Read quantification per annotation range</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_09.html">9. Analysis of differentially expressed genes with `edgeR`</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_10.html">10. GO term enrichment analysis of DEGs}</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_11.html">11. Differential ribosome loading analysis (translational efficiency)</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_12.html">12. Clustering and heat maps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_13.html">13. Render report in HTML and PDF format</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_14.html">14. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_15.html">15. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRIBOseq_16.html">16. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">VAR-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeVARseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_05.html">5. Variant calling</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_06.html">6. Filter variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_07.html">7. Annotate filtered variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_08.html">8. Combine annotation results among samples</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_09.html">9. Summary statistics of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_10.html">10. Venn diagram of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_11.html">11. Plot variants programmatically</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Training Events</a>
        <ul>
            
            
            
            <li><a href="mydoc_training.html">Training Events</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Publications</a>
        <ul>
            
            
            
            <li><a href="mydoc_publications.html">Publications</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Acknowledgement</a>
        <ul>
            
            
            
            <li><a href="mydoc_acknowledgement.html">Acknowledgement</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Contacts</a>
        <ul>
            
            
            
            <li><a href="mydoc_contacts.html">Contacts</a></li>
            
            
            
            
        </ul>
        
        
        
        <!-- if you aren't using the accordion, uncomment this block:
           <p class="external">
               <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
           </p>
           -->
    </li>
</ul>
</div>

<!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
<script>$("li.active").parents('li').toggleClass("active");</script>

    <!-- Content Column -->
    <div class="col-md-9" id="tg-sb-content">
        <div class="post-header">
   <h1 class="post-title-main">3. Workflow overview</h1>
</div>



<div class="post-content">

   

    
    
<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>

<div id="toc"></div>

    


<!--
    

    <a target="_blank" href="https://github.com/tgirke/systemPipeR/blob/gh-pages/pages//mydoc_systemPipeR_3.md" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
-->
    

  <h2 id="define-environment-settings-and-samples">Define environment settings and samples</h2>

<p>A typical workflow starts with generating the expected working environment
containing the proper directory structure, input files and parameter settings.
To simplify this task, one can load one of the existing NGS workflows templates
provided by <em><code class="highlighter-rouge">systemPipeRdata</code></em> into the current working directory. The
following does this for the <em><code class="highlighter-rouge">rnaseq</code></em> template. The name of the resulting
workflow directory can be specified under the <em><code class="highlighter-rouge">mydirname</code></em> argument. The
default <em><code class="highlighter-rouge">NULL</code></em> uses the name of the chosen workflow. An error is issued if a
directory of the same name and path exists already. On Linux and OS X systems
one can also create new workflow instances from the command-line of a terminal as shown
<a href="http://bioconductor.org/packages/devel/data/experiment/vignettes/systemPipeRdata/inst/doc/systemPipeRdata.html#generate-workflow-template">here</a>.
To apply workflows to custom data, the user needs to modify the <em><code class="highlighter-rouge">targets</code></em> file and if
necessary update the corresponding <em><code class="highlighter-rouge">param</code></em> file(s). A collection of pre-generated <em><code class="highlighter-rouge">param</code></em>
files is provided in the <em><code class="highlighter-rouge">param</code></em> subdirectory of each workflow template. They
are also viewable in the GitHub repository of <em><code class="highlighter-rouge">systemPipeRdata</code></em> (<a href="https://github.com/tgirke/systemPipeRdata/tree/master/inst/extdata/param">see
here</a>).</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeRdata</span><span class="p">)</span><span class="w">
</span><span class="n">genWorkenvir</span><span class="p">(</span><span class="n">workflow</span><span class="o">=</span><span class="s2">"rnaseq"</span><span class="p">,</span><span class="w"> </span><span class="n">mydirname</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span><span class="n">setwd</span><span class="p">(</span><span class="s2">"rnaseq"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Construct <em><code class="highlighter-rouge">SYSargs</code></em> object from <em><code class="highlighter-rouge">param</code></em> and <em><code class="highlighter-rouge">targets</code></em> files.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/trim.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h2 id="read-preprocessing">Read Preprocessing</h2>
<p>The function <em><code class="highlighter-rouge">preprocessReads</code></em> allows to apply predefined or custom
read preprocessing functions to all FASTQ files referenced in a
<em><code class="highlighter-rouge">SYSargs</code></em> container, such as quality filtering or adaptor trimming
routines. The paths to the resulting output FASTQ files are stored in the
<em><code class="highlighter-rouge">outpaths</code></em> slot of the <em><code class="highlighter-rouge">SYSargs</code></em> object. Internally,
<em><code class="highlighter-rouge">preprocessReads</code></em> uses the <em><code class="highlighter-rouge">FastqStreamer</code></em> function from
the <em><code class="highlighter-rouge">ShortRead</code></em> package to stream through large FASTQ files in a
memory-efficient manner. The following example performs adaptor trimming with
the <em><code class="highlighter-rouge">trimLRPatterns</code></em> function from the <em><code class="highlighter-rouge">Biostrings</code></em> package.
After the trimming step a new targets file is generated (here
<em><code class="highlighter-rouge">targets_trim.txt</code></em>) containing the paths to the trimmed FASTQ files.
The new targets file can be used for the next workflow step with an updated
<em><code class="highlighter-rouge">SYSargs</code></em> instance, <em>e.g.</em> running the NGS alignments with the
trimmed FASTQ files.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">preprocessReads</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">Fct</span><span class="o">=</span><span class="s2">"trimLRPatterns(Rpattern='GCCCGGGTAA', subject=fq)"</span><span class="p">,</span><span class="w"> 
                </span><span class="n">batchsize</span><span class="o">=</span><span class="m">100000</span><span class="p">,</span><span class="w"> </span><span class="n">overwrite</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">compress</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">writeTargetsout</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"targets_trim.txt"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>The following example shows how one can design a custom read preprocessing function 
using utilities provided by the <em><code class="highlighter-rouge">ShortRead</code></em> package, and then run it
in batch mode with the <em>‘preprocessReads’</em> function (here on paired-end reads).</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/trimPE.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targetsPE.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">filterFct</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">fq</span><span class="p">,</span><span class="w"> </span><span class="n">cutoff</span><span class="o">=</span><span class="m">20</span><span class="p">,</span><span class="w"> </span><span class="n">Nexceptions</span><span class="o">=</span><span class="m">0</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">qcount</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rowSums</span><span class="p">(</span><span class="n">as</span><span class="p">(</span><span class="n">quality</span><span class="p">(</span><span class="n">fq</span><span class="p">),</span><span class="w"> </span><span class="s2">"matrix"</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">cutoff</span><span class="p">)</span><span class="w">
    </span><span class="n">fq</span><span class="p">[</span><span class="n">qcount</span><span class="w"> </span><span class="o">&lt;=</span><span class="w"> </span><span class="n">Nexceptions</span><span class="p">]</span><span class="w"> </span><span class="c1"># Retains reads where Phred scores are &gt;= cutoff with N exceptions</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">preprocessReads</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">Fct</span><span class="o">=</span><span class="s2">"filterFct(fq, cutoff=20, Nexceptions=0)"</span><span class="p">,</span><span class="w"> </span><span class="n">batchsize</span><span class="o">=</span><span class="m">100000</span><span class="p">)</span><span class="w">
</span><span class="n">writeTargetsout</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"targets_PEtrim.txt"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h2 id="fastq-quality-report">FASTQ quality report</h2>
<p>The following <em><code class="highlighter-rouge">seeFastq</code></em> and <em><code class="highlighter-rouge">seeFastqPlot</code></em> functions generate and plot a series of
useful quality statistics for a set of FASTQ files including per cycle quality
box plots, base proportions, base-level quality trends, relative k-mer
diversity, length and occurrence distribution of reads, number of reads above
quality cutoffs and mean quality distribution.<br />
The function <em><code class="highlighter-rouge">seeFastq</code></em> computes the quality statistics and stores the results in a
relatively small list object that can be saved to disk with <em><code class="highlighter-rouge">save()</code></em> and
reloaded with <em><code class="highlighter-rouge">load()</code></em> for later plotting. The argument <em><code class="highlighter-rouge">klength</code></em> specifies the
k-mer length and <em><code class="highlighter-rouge">batchsize</code></em> the number of reads to random sample from each
FASTQ file.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">fqlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">seeFastq</span><span class="p">(</span><span class="n">fastq</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">batchsize</span><span class="o">=</span><span class="m">10000</span><span class="p">,</span><span class="w"> </span><span class="n">klength</span><span class="o">=</span><span class="m">8</span><span class="p">)</span><span class="w">
</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"./results/fastqReport.pdf"</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="o">=</span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="o">=</span><span class="m">4</span><span class="o">*</span><span class="nf">length</span><span class="p">(</span><span class="n">fqlist</span><span class="p">))</span><span class="w">
</span><span class="n">seeFastqPlot</span><span class="p">(</span><span class="n">fqlist</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>

<p><img src="./pages/mydoc/systemPipeR_files/fastqReport.png" alt="" /></p>
<div align="center"><b>Figure 2:</b> FASTQ quality report. To zoom in, rigth click image and open it in a separate browser tab. </div>

<p>Parallelization of QC report on single machine with multiple cores</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">seeFastq</span><span class="p">(</span><span class="n">fastq</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">batchsize</span><span class="o">=</span><span class="m">100000</span><span class="p">,</span><span class="w"> </span><span class="n">klength</span><span class="o">=</span><span class="m">8</span><span class="p">)</span><span class="w">
</span><span class="n">fqlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">,</span><span class="w"> </span><span class="n">BPPARAM</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MulticoreParam</span><span class="p">(</span><span class="n">workers</span><span class="o">=</span><span class="m">8</span><span class="p">))</span><span class="w">
</span><span class="n">seeFastqPlot</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">fqlist</span><span class="p">,</span><span class="w"> </span><span class="n">recursive</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<p>Parallelization of QC report via scheduler (<em>e.g.</em> Torque) across several compute nodes</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BatchJobs</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span><span class="w">
    </span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
    </span><span class="n">seeFastq</span><span class="p">(</span><span class="n">fastq</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">batchsize</span><span class="o">=</span><span class="m">100000</span><span class="p">,</span><span class="w"> </span><span class="n">klength</span><span class="o">=</span><span class="m">8</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">funs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeClusterFunctionsSLURM</span><span class="p">(</span><span class="s2">"slurm.tmpl"</span><span class="p">)</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchJobsParam</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">resources</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">ntasks</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">ncpus</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"6gb"</span><span class="p">),</span><span class="w"> </span><span class="n">cluster.functions</span><span class="o">=</span><span class="n">funs</span><span class="p">)</span><span class="w">
</span><span class="n">register</span><span class="p">(</span><span class="n">param</span><span class="p">)</span><span class="w">
</span><span class="n">fqlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w">
</span><span class="n">seeFastqPlot</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">fqlist</span><span class="p">,</span><span class="w"> </span><span class="n">recursive</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<h2 id="alignment-with-tophat2">Alignment with <em><code class="highlighter-rouge">Tophat2</code></em></h2>
<p>Build <em><code class="highlighter-rouge">Bowtie2</code></em> index.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">moduleload</span><span class="p">(</span><span class="n">modules</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w"> </span><span class="c1"># Skip if module system is not available</span><span class="w">
</span><span class="n">system</span><span class="p">(</span><span class="s2">"bowtie2-build ./data/tair10.fasta ./data/tair10.fasta"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Execute <em><code class="highlighter-rouge">SYSargs</code></em> on a single machine without submitting to a queuing system of a compute cluster. This way the input FASTQ files will be processed sequentially. If available, multiple CPU cores can be used for processing each file. The number of CPU cores (here 4) to use for each process is defined in the <em><code class="highlighter-rouge">*.param</code></em> file. With <em><code class="highlighter-rouge">cores(args)</code></em> one can return this value from the <em><code class="highlighter-rouge">SYSargs</code></em> object. Note, if a module system is not installed or used, then the corresponding <em><code class="highlighter-rouge">*.param</code></em> file needs to be edited accordingly by either providing an empty field in the line(s) starting with <em><code class="highlighter-rouge">module</code></em> or by deleting these lines.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">bampaths</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">runCommandline</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Alternatively, the computation can be greatly accelerated by processing many files in parallel using several compute nodes of a cluster, where a scheduling/queuing system is used for load balancing. To avoid over-subscription of CPU cores on the compute nodes, the value from <em><code class="highlighter-rouge">cores(args)</code></em> is passed on to the submission command, here <em><code class="highlighter-rouge">nodes</code></em> in the <em><code class="highlighter-rouge">resources</code></em> list object. The number of independent parallel cluster processes is defined under the <em><code class="highlighter-rouge">Njobs</code></em> argument. The following example will run 18 processes in parallel using for each 4 CPU cores. If the resources available on a cluster allow to run all 18 processes at the same time then the shown sample submission will utilize in total 72 CPU cores. Note, <em><code class="highlighter-rouge">clusterRun</code></em> can be used with most queueing systems as it is based on utilities from the <em><code class="highlighter-rouge">BatchJobs</code></em> package which supports the use of template files (<em><code class="highlighter-rouge">*.tmpl</code></em>) for defining the run parameters of different schedulers. To run the following code, one needs to have both a conf file (see <em><code class="highlighter-rouge">.BatchJob</code></em> samples <a href="https://code.google.com/p/batchjobs/wiki/DortmundUsage">here</a>) and a template file (see <em><code class="highlighter-rouge">*.tmpl</code></em> samples <a href="https://github.com/tudo-r/BatchJobs/tree/master/examples">here</a>) for the queueing available on a system. The following example uses the sample conf and template files for the Torque scheduler provided by this package.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">resources</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">ntasks</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">ncpus</span><span class="o">=</span><span class="n">cores</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"10G"</span><span class="p">)</span><span class="w">
</span><span class="n">reg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterRun</span><span class="p">(</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">conffile</span><span class="o">=</span><span class="s2">".BatchJobs.R"</span><span class="p">,</span><span class="w"> </span><span class="n">template</span><span class="o">=</span><span class="s2">"slurm.tmpl"</span><span class="p">,</span><span class="w"> </span><span class="n">Njobs</span><span class="o">=</span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">runid</span><span class="o">=</span><span class="s2">"01"</span><span class="p">,</span><span class="w"> 
		  </span><span class="n">resourceList</span><span class="o">=</span><span class="n">resources</span><span class="p">)</span><span class="w">
</span><span class="n">waitForJobs</span><span class="p">(</span><span class="n">reg</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Useful commands for monitoring progress of submitted jobs</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">showStatus</span><span class="p">(</span><span class="n">reg</span><span class="p">)</span><span class="w">
</span><span class="n">file.exists</span><span class="p">(</span><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w">
</span><span class="n">sapply</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">loadResult</span><span class="p">(</span><span class="n">reg</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="p">))</span><span class="w"> </span><span class="c1"># Works after job completion</span><span class="w">
</span></code></pre></div></div>

<h2 id="read-and-alignment-count-stats">Read and alignment count stats</h2>
<p>Generate table of read and alignment counts for all samples.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">read_statsDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">alignStats</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="w"> 
</span><span class="n">write.table</span><span class="p">(</span><span class="n">read_statsDF</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/alignStats.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>The following shows the first four lines of the sample alignment stats file provided by the <em><code class="highlighter-rouge">systemPipeR</code></em> package. For simplicity the number of PE reads is multiplied here by 2 to approximate proper alignment frequencies where each read in a pair is counted.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">read.table</span><span class="p">(</span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"alignStats.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">),</span><span class="w"> </span><span class="n">header</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##   FileName Nreads2x Nalign Perc_Aligned Nalign_Primary Perc_Aligned_Primary
## 1      M1A   192918 177961     92.24697         177961             92.24697
## 2      M1B   197484 159378     80.70426         159378             80.70426
## 3      A1A   189870 176055     92.72397         176055             92.72397
## 4      A1B   188854 147768     78.24457         147768             78.24457
</code></pre></div></div>

<p>Parallelization of read/alignment stats on single machine with multiple cores</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">alignStats</span><span class="p">(</span><span class="n">args</span><span class="p">[</span><span class="n">x</span><span class="p">])</span><span class="w">
</span><span class="n">read_statsList</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">,</span><span class="w"> </span><span class="n">BPPARAM</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">MulticoreParam</span><span class="p">(</span><span class="n">workers</span><span class="o">=</span><span class="m">8</span><span class="p">))</span><span class="w">
</span><span class="n">read_statsDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">do.call</span><span class="p">(</span><span class="s2">"rbind"</span><span class="p">,</span><span class="w"> </span><span class="n">read_statsList</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Parallelization of read/alignment stats via scheduler (<em>e.g.</em> Torque) across several compute nodes</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BatchJobs</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span><span class="w">
    </span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
    </span><span class="n">alignStats</span><span class="p">(</span><span class="n">args</span><span class="p">[</span><span class="n">x</span><span class="p">])</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">funs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeClusterFunctionsSLURM</span><span class="p">(</span><span class="s2">"slurm.tmpl"</span><span class="p">)</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchJobsParam</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">resources</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">ntasks</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">ncpus</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"6gb"</span><span class="p">),</span><span class="w"> </span><span class="n">cluster.functions</span><span class="o">=</span><span class="n">funs</span><span class="p">)</span><span class="w">
</span><span class="n">register</span><span class="p">(</span><span class="n">param</span><span class="p">)</span><span class="w">
</span><span class="n">read_statsList</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w">
</span><span class="n">read_statsDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">do.call</span><span class="p">(</span><span class="s2">"rbind"</span><span class="p">,</span><span class="w"> </span><span class="n">read_statsList</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h2 id="create-symbolic-links-for-viewing-bam-files-in-igv">Create symbolic links for viewing BAM files in IGV</h2>
<p>The genome browser IGV supports reading of indexed/sorted BAM files via web URLs. This way it can be avoided to create unnecessary copies of these large files. To enable this approach, an HTML directory with http access needs to be available in the user account (<em>e.g.</em> <em><code class="highlighter-rouge">home/publichtml</code></em>) of a system. If this is not the case then the BAM files need to be moved or copied to the system where IGV runs. In the following, <em><code class="highlighter-rouge">htmldir</code></em> defines the path to the HTML directory with http access where the symbolic links to the BAM files will be stored. The corresponding URLs will be written to a text file specified under the <code class="highlighter-rouge">_urlfile</code>_ argument.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">symLink2bam</span><span class="p">(</span><span class="n">sysargs</span><span class="o">=</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">htmldir</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"~/.html/"</span><span class="p">,</span><span class="w"> </span><span class="s2">"somedir/"</span><span class="p">),</span><span class="w"> 
            </span><span class="n">urlbase</span><span class="o">=</span><span class="s2">"http://myserver.edu/~username/"</span><span class="p">,</span><span class="w"> 
        </span><span class="n">urlfile</span><span class="o">=</span><span class="s2">"IGVurl.txt"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h2 id="alternative-ngs-aligners">Alternative NGS Aligners</h2>

<h3 id="alignment-with-bowtie2-eg-for-mirna-profiling">Alignment with <em><code class="highlighter-rouge">Bowtie2</code></em> (<em>e.g.</em> for miRNA profiling)</h3>
<p>The following example runs <em><code class="highlighter-rouge">Bowtie2</code></em> as a single process without submitting it to a cluster.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"bowtieSE.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">moduleload</span><span class="p">(</span><span class="n">modules</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w"> </span><span class="c1"># Skip if module system is not available</span><span class="w">
</span><span class="n">bampaths</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">runCommandline</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Alternatively, submit the job to compute nodes.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">resources</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">ntasks</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">ncpus</span><span class="o">=</span><span class="n">cores</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"10G"</span><span class="p">)</span><span class="w">
</span><span class="n">reg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">clusterRun</span><span class="p">(</span><span class="n">args</span><span class="p">,</span><span class="w"> </span><span class="n">conffile</span><span class="o">=</span><span class="s2">".BatchJobs.R"</span><span class="p">,</span><span class="w"> </span><span class="n">template</span><span class="o">=</span><span class="s2">"slurm.tmpl"</span><span class="p">,</span><span class="w"> </span><span class="n">Njobs</span><span class="o">=</span><span class="m">18</span><span class="p">,</span><span class="w"> </span><span class="n">runid</span><span class="o">=</span><span class="s2">"01"</span><span class="p">,</span><span class="w"> 
		  </span><span class="n">resourceList</span><span class="o">=</span><span class="n">resources</span><span class="p">)</span><span class="w">
</span><span class="n">waitForJobs</span><span class="p">(</span><span class="n">reg</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h3 id="alignment-with-bwa-mem-eg-for-var-seq">Alignment with <em><code class="highlighter-rouge">BWA-MEM</code></em> (<em>e.g.</em> for VAR-Seq)</h3>
<p>The following example runs BWA-MEM as a single process without submitting it to a cluster.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/bwa.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">moduleload</span><span class="p">(</span><span class="n">modules</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w"> </span><span class="c1"># Skip if module system is not available</span><span class="w">
</span><span class="n">system</span><span class="p">(</span><span class="s2">"bwa index -a bwtsw ./data/tair10.fasta"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Indexes reference genome</span><span class="w">
</span><span class="n">bampaths</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">runCommandline</span><span class="p">(</span><span class="n">args</span><span class="o">=</span><span class="n">args</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">])</span><span class="w">
</span></code></pre></div></div>

<h3 id="alignment-with-rsubread-eg-for-rna-seq">Alignment with <em><code class="highlighter-rouge">Rsubread</code></em> (<em>e.g.</em> for RNA-Seq)</h3>
<p>The following example shows how one can use within the <code class="highlighter-rouge">systemPipeR</code> environment the R-based 
aligner <code class="highlighter-rouge">Rsubread</code> or other R-based functions that read from input files and write to output files.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">Rsubread</span><span class="p">)</span><span class="w">
</span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/rsubread.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">buildindex</span><span class="p">(</span><span class="n">basename</span><span class="o">=</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">reference</span><span class="o">=</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w"> </span><span class="c1"># Build indexed reference genome</span><span class="w">
</span><span class="n">align</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">readfile1</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span><span class="w"> </span><span class="n">input_format</span><span class="o">=</span><span class="s2">"FASTQ"</span><span class="p">,</span><span class="w"> 
      </span><span class="n">output_file</span><span class="o">=</span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span><span class="w"> </span><span class="n">output_format</span><span class="o">=</span><span class="s2">"SAM"</span><span class="p">,</span><span class="w"> </span><span class="n">nthreads</span><span class="o">=</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">indels</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">TH1</span><span class="o">=</span><span class="m">2</span><span class="p">)</span><span class="w">
</span><span class="k">for</span><span class="p">(</span><span class="n">i</span><span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)))</span><span class="w"> </span><span class="n">asBam</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">i</span><span class="p">],</span><span class="w"> </span><span class="n">destination</span><span class="o">=</span><span class="n">gsub</span><span class="p">(</span><span class="s2">".sam"</span><span class="p">,</span><span class="w"> </span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">i</span><span class="p">]),</span><span class="w"> </span><span class="n">overwrite</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">indexDestination</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h3 id="alignment-with-gsnap-eg-for-var-seq-and-rna-seq">Alignment with <em><code class="highlighter-rouge">gsnap</code></em> (<em>e.g.</em> for VAR-Seq and RNA-Seq)</h3>
<p>Another R-based short read aligner is <em><code class="highlighter-rouge">gsnap</code></em> from the <em><code class="highlighter-rouge">gmapR</code></em> package (Wu et al., 2010). 
The code sample below introduces how to run this aligner on multiple nodes of a compute cluster.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">gmapR</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BatchJobs</span><span class="p">)</span><span class="w">
</span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"param/gsnap.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targetsPE.txt"</span><span class="p">)</span><span class="w">
</span><span class="n">gmapGenome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GmapGenome</span><span class="p">(</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">directory</span><span class="o">=</span><span class="s2">"data"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="o">=</span><span class="s2">"gmap_tair10chr/"</span><span class="p">,</span><span class="w"> </span><span class="n">create</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">library</span><span class="p">(</span><span class="n">gmapR</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">)</span><span class="w">
    </span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"gsnap.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targetsPE.txt"</span><span class="p">)</span><span class="w">
    </span><span class="n">gmapGenome</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GmapGenome</span><span class="p">(</span><span class="n">reference</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">directory</span><span class="o">=</span><span class="s2">"data"</span><span class="p">,</span><span class="w"> </span><span class="n">name</span><span class="o">=</span><span class="s2">"gmap_tair10chr/"</span><span class="p">,</span><span class="w"> </span><span class="n">create</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
    </span><span class="n">p</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GsnapParam</span><span class="p">(</span><span class="n">genome</span><span class="o">=</span><span class="n">gmapGenome</span><span class="p">,</span><span class="w"> </span><span class="n">unique_only</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">molecule</span><span class="o">=</span><span class="s2">"DNA"</span><span class="p">,</span><span class="w"> </span><span class="n">max_mismatches</span><span class="o">=</span><span class="m">3</span><span class="p">)</span><span class="w">
    </span><span class="n">o</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">gsnap</span><span class="p">(</span><span class="n">input_a</span><span class="o">=</span><span class="n">infile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">input_b</span><span class="o">=</span><span class="n">infile2</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">params</span><span class="o">=</span><span class="n">p</span><span class="p">,</span><span class="w"> </span><span class="n">output</span><span class="o">=</span><span class="n">outfile1</span><span class="p">(</span><span class="n">args</span><span class="p">)[</span><span class="n">x</span><span class="p">])</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">funs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeClusterFunctionsSLURM</span><span class="p">(</span><span class="s2">"slurm.tmpl"</span><span class="p">)</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchJobsParam</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">resources</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">ntasks</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">ncpus</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"6gb"</span><span class="p">),</span><span class="w"> </span><span class="n">cluster.functions</span><span class="o">=</span><span class="n">funs</span><span class="p">)</span><span class="w">
</span><span class="n">register</span><span class="p">(</span><span class="n">param</span><span class="p">)</span><span class="w">
</span><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h2 id="read-counting-for-mrna-profiling-experiments">Read counting for mRNA profiling experiments</h2>
<p>Create <em><code class="highlighter-rouge">txdb</code></em> (needs to be done only once)</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">GenomicFeatures</span><span class="p">)</span><span class="w">
</span><span class="n">txdb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeTxDbFromGFF</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="s2">"data/tair10.gff"</span><span class="p">,</span><span class="w"> </span><span class="n">format</span><span class="o">=</span><span class="s2">"gff"</span><span class="p">,</span><span class="w"> </span><span class="n">dataSource</span><span class="o">=</span><span class="s2">"TAIR"</span><span class="p">,</span><span class="w"> </span><span class="n">organism</span><span class="o">=</span><span class="s2">"A. thaliana"</span><span class="p">)</span><span class="w">
</span><span class="n">saveDb</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"./data/tair10.sqlite"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>The following performs read counting with <em><code class="highlighter-rouge">summarizeOverlaps</code></em> in parallel mode with multiple cores.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">)</span><span class="w">
</span><span class="n">txdb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">loadDb</span><span class="p">(</span><span class="s2">"./data/tair10.sqlite"</span><span class="p">)</span><span class="w">
</span><span class="n">eByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">exonsBy</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"gene"</span><span class="p">)</span><span class="w">
</span><span class="n">bfl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BamFileList</span><span class="p">(</span><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">yieldSize</span><span class="o">=</span><span class="m">50000</span><span class="p">,</span><span class="w"> </span><span class="n">index</span><span class="o">=</span><span class="n">character</span><span class="p">())</span><span class="w">
</span><span class="n">multicoreParam</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">MulticoreParam</span><span class="p">(</span><span class="n">workers</span><span class="o">=</span><span class="m">4</span><span class="p">);</span><span class="w"> </span><span class="n">register</span><span class="p">(</span><span class="n">multicoreParam</span><span class="p">);</span><span class="w"> </span><span class="n">registered</span><span class="p">()</span><span class="w">
</span><span class="n">counteByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">bfl</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">summarizeOverlaps</span><span class="p">(</span><span class="n">eByg</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">mode</span><span class="o">=</span><span class="s2">"Union"</span><span class="p">,</span><span class="w"> </span><span class="n">ignore.strand</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">inter.feature</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">singleEnd</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span><span class="w"> </span><span class="c1"># Note: for strand-specific RNA-Seq set 'ignore.strand=FALSE' and for PE data set 'singleEnd=FALSE'</span><span class="w">
</span><span class="n">countDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">counteByg</span><span class="p">),</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">assays</span><span class="p">(</span><span class="n">counteByg</span><span class="p">[[</span><span class="n">x</span><span class="p">]])</span><span class="o">$</span><span class="n">counts</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">rowRanges</span><span class="p">(</span><span class="n">counteByg</span><span class="p">[[</span><span class="m">1</span><span class="p">]]));</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">bfl</span><span class="p">)</span><span class="w">
</span><span class="n">rpkmDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">returnRPKM</span><span class="p">(</span><span class="n">counts</span><span class="o">=</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">ranges</span><span class="o">=</span><span class="n">eByg</span><span class="p">))</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/countDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">rpkmDFeByg</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/rpkmDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p>Please note, in addition to read counts this step generates RPKM normalized expression values. For most statistical differential expression or abundance analysis methods, such as <em><code class="highlighter-rouge">edgeR</code></em> or <em><code class="highlighter-rouge">DESeq2</code></em>, the raw count values should be used as input. The usage of RPKM values should be restricted to specialty applications required by some users, <em>e.g.</em> manually comparing the expression levels of different genes or features.</p>

<p>Read counting with <em><code class="highlighter-rouge">summarizeOverlaps</code></em> using multiple nodes of a cluster</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">)</span><span class="w">
</span><span class="n">f</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="p">{</span><span class="w">
    </span><span class="n">library</span><span class="p">(</span><span class="n">systemPipeR</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">BiocParallel</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">GenomicFeatures</span><span class="p">)</span><span class="w">
    </span><span class="n">txdb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">loadDb</span><span class="p">(</span><span class="s2">"./data/tair10.sqlite"</span><span class="p">)</span><span class="w">
    </span><span class="n">eByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">exonsBy</span><span class="p">(</span><span class="n">txdb</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="s2">"gene"</span><span class="p">)</span><span class="w">
    </span><span class="n">args</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">systemArgs</span><span class="p">(</span><span class="n">sysma</span><span class="o">=</span><span class="s2">"tophat.param"</span><span class="p">,</span><span class="w"> </span><span class="n">mytargets</span><span class="o">=</span><span class="s2">"targets.txt"</span><span class="p">)</span><span class="w">
    </span><span class="n">bfl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BamFileList</span><span class="p">(</span><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">yieldSize</span><span class="o">=</span><span class="m">50000</span><span class="p">,</span><span class="w"> </span><span class="n">index</span><span class="o">=</span><span class="n">character</span><span class="p">())</span><span class="w">
    </span><span class="n">summarizeOverlaps</span><span class="p">(</span><span class="n">eByg</span><span class="p">,</span><span class="w"> </span><span class="n">bfl</span><span class="p">[</span><span class="n">x</span><span class="p">],</span><span class="w"> </span><span class="n">mode</span><span class="o">=</span><span class="s2">"Union"</span><span class="p">,</span><span class="w"> </span><span class="n">ignore.strand</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">inter.feature</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">singleEnd</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span><span class="p">}</span><span class="w">
</span><span class="n">funs</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeClusterFunctionsSLURM</span><span class="p">(</span><span class="s2">"slurm.tmpl"</span><span class="p">)</span><span class="w">
</span><span class="n">param</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchJobsParam</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">resources</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="n">walltime</span><span class="o">=</span><span class="s2">"20:00:00"</span><span class="p">,</span><span class="w"> </span><span class="n">ntasks</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">ncpus</span><span class="o">=</span><span class="m">1</span><span class="p">,</span><span class="w"> </span><span class="n">memory</span><span class="o">=</span><span class="s2">"6gb"</span><span class="p">),</span><span class="w"> </span><span class="n">cluster.functions</span><span class="o">=</span><span class="n">funs</span><span class="p">)</span><span class="w">
</span><span class="n">register</span><span class="p">(</span><span class="n">param</span><span class="p">)</span><span class="w">
</span><span class="n">counteByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bplapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">args</span><span class="p">),</span><span class="w"> </span><span class="n">f</span><span class="p">)</span><span class="w"> 
</span><span class="n">countDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">sapply</span><span class="p">(</span><span class="n">seq</span><span class="p">(</span><span class="n">along</span><span class="o">=</span><span class="n">counteByg</span><span class="p">),</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">assays</span><span class="p">(</span><span class="n">counteByg</span><span class="p">[[</span><span class="n">x</span><span class="p">]])</span><span class="o">$</span><span class="n">counts</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">rowRanges</span><span class="p">(</span><span class="n">counteByg</span><span class="p">[[</span><span class="m">1</span><span class="p">]]));</span><span class="w"> </span><span class="n">colnames</span><span class="p">(</span><span class="n">countDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">outpaths</span><span class="p">(</span><span class="n">args</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<h2 id="read-counting-for-mirna-profiling-experiments">Read counting for miRNA profiling experiments</h2>
<p>Download miRNA genes from miRBase</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">system</span><span class="p">(</span><span class="s2">"wget ftp://mirbase.org/pub/mirbase/19/genomes/My_species.gff3 -P ./data/"</span><span class="p">)</span><span class="w">
</span><span class="n">gff</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">import.gff</span><span class="p">(</span><span class="s2">"./data/My_species.gff3"</span><span class="p">)</span><span class="w">
</span><span class="n">gff</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">split</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="n">elementMetadata</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="o">$</span><span class="n">ID</span><span class="p">)</span><span class="w">
</span><span class="n">bams</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">bampaths</span><span class="p">);</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">bams</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">targets</span><span class="o">$</span><span class="n">SampleName</span><span class="w">
</span><span class="n">bfl</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BamFileList</span><span class="p">(</span><span class="n">bams</span><span class="p">,</span><span class="w"> </span><span class="n">yieldSize</span><span class="o">=</span><span class="m">50000</span><span class="p">,</span><span class="w"> </span><span class="n">index</span><span class="o">=</span><span class="n">character</span><span class="p">())</span><span class="w">
</span><span class="n">countDFmiR</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">summarizeOverlaps</span><span class="p">(</span><span class="n">gff</span><span class="p">,</span><span class="w"> </span><span class="n">bfl</span><span class="p">,</span><span class="w"> </span><span class="n">mode</span><span class="o">=</span><span class="s2">"Union"</span><span class="p">,</span><span class="w"> </span><span class="n">ignore.strand</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">inter.feature</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w"> </span><span class="c1"># Note: inter.feature=FALSE important since pre and mature miRNA ranges overlap</span><span class="w">
</span><span class="n">rpkmDFmiR</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">apply</span><span class="p">(</span><span class="n">countDFmiR</span><span class="p">,</span><span class="w"> </span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="k">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="w"> </span><span class="n">returnRPKM</span><span class="p">(</span><span class="n">counts</span><span class="o">=</span><span class="n">x</span><span class="p">,</span><span class="w"> </span><span class="n">gffsub</span><span class="o">=</span><span class="n">gff</span><span class="p">))</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">assays</span><span class="p">(</span><span class="n">countDFmiR</span><span class="p">)</span><span class="o">$</span><span class="n">counts</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/countDFmiR.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">rpkmDFmiR</span><span class="p">,</span><span class="w"> </span><span class="s2">"results/rpkmDFmiR.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">NA</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h2 id="correlation-analysis-of-samples">Correlation analysis of samples</h2>
<p>The following computes the sample-wise Spearman correlation coefficients from the <em><code class="highlighter-rouge">rlog</code></em> (regularized-logarithm) transformed expression values generated with the <em><code class="highlighter-rouge">DESeq2</code></em> package. After transformation to a distance matrix, hierarchical clustering is performed with the <em><code class="highlighter-rouge">hclust</code></em> function and the result is plotted as a dendrogram (<a href="./results/sample_tree.pdf">sample_tree.pdf</a>).</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">,</span><span class="w"> </span><span class="n">warn.conflicts</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">quietly</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">);</span><span class="w"> </span><span class="n">library</span><span class="p">(</span><span class="n">ape</span><span class="p">,</span><span class="w"> </span><span class="n">warn.conflicts</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">countDFpath</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"countDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w">
</span><span class="n">countDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.matrix</span><span class="p">(</span><span class="n">read.table</span><span class="p">(</span><span class="n">countDFpath</span><span class="p">))</span><span class="w">
</span><span class="n">colData</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">data.frame</span><span class="p">(</span><span class="n">row.names</span><span class="o">=</span><span class="n">targetsin</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="o">$</span><span class="n">SampleName</span><span class="p">,</span><span class="w"> </span><span class="n">condition</span><span class="o">=</span><span class="n">targetsin</span><span class="p">(</span><span class="n">args</span><span class="p">)</span><span class="o">$</span><span class="n">Factor</span><span class="p">)</span><span class="w">
</span><span class="n">dds</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DESeqDataSetFromMatrix</span><span class="p">(</span><span class="n">countData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">countDF</span><span class="p">,</span><span class="w"> </span><span class="n">colData</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">colData</span><span class="p">,</span><span class="w"> </span><span class="n">design</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="o">~</span><span class="w"> </span><span class="n">condition</span><span class="p">)</span><span class="w">
</span><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cor</span><span class="p">(</span><span class="n">assay</span><span class="p">(</span><span class="n">rlog</span><span class="p">(</span><span class="n">dds</span><span class="p">)),</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"spearman"</span><span class="p">)</span><span class="w">
</span><span class="n">hc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">hclust</span><span class="p">(</span><span class="n">dist</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">d</span><span class="p">))</span><span class="w">
</span><span class="n">plot.phylo</span><span class="p">(</span><span class="n">as.phylo</span><span class="p">(</span><span class="n">hc</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"p"</span><span class="p">,</span><span class="w"> </span><span class="n">edge.col</span><span class="o">=</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">edge.width</span><span class="o">=</span><span class="m">3</span><span class="p">,</span><span class="w"> </span><span class="n">show.node.label</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">no.margin</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p><img src="./pages/mydoc/systemPipeR_files/sample_tree_rlog-1.png" width="100%" /></p>

<div align="center"><b>Figure 3:</b> Correlation dendrogram of samples for <i>rlog</i> values. </div>

<p>Alternatively, the clustering can be performed with <em><code class="highlighter-rouge">RPKM</code></em> normalized expression values. In combination with Spearman correlation the results of the two clustering methods are often relatively similar.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">rpkmDFeBygpath</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"rpkmDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w">
</span><span class="n">rpkmDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.table</span><span class="p">(</span><span class="n">rpkmDFeBygpath</span><span class="p">,</span><span class="w"> </span><span class="n">check.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">rpkmDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">rpkmDFeByg</span><span class="p">[</span><span class="n">rowMeans</span><span class="p">(</span><span class="n">rpkmDFeByg</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">50</span><span class="p">,]</span><span class="w">
</span><span class="n">d</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">cor</span><span class="p">(</span><span class="n">rpkmDFeByg</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"spearman"</span><span class="p">)</span><span class="w">
</span><span class="n">hc</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">hclust</span><span class="p">(</span><span class="n">as.dist</span><span class="p">(</span><span class="m">1</span><span class="o">-</span><span class="n">d</span><span class="p">))</span><span class="w">
</span><span class="n">plot.phylo</span><span class="p">(</span><span class="n">as.phylo</span><span class="p">(</span><span class="n">hc</span><span class="p">),</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"p"</span><span class="p">,</span><span class="w"> </span><span class="n">edge.col</span><span class="o">=</span><span class="s2">"blue"</span><span class="p">,</span><span class="w"> </span><span class="n">edge.width</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">show.node.label</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span><span class="w"> </span><span class="n">no.margin</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h2 id="deg-analysis-with-edger">DEG analysis with <em><code class="highlighter-rouge">edgeR</code></em></h2>
<p>The following <em><code class="highlighter-rouge">run_edgeR</code></em> function is a convenience wrapper for
identifying differentially expressed genes (DEGs) in batch mode with
<em><code class="highlighter-rouge">edgeR</code></em>’s GML method (Robinson et al., 2010) for any number of
pairwise sample comparisons specified under the <em><code class="highlighter-rouge">cmp</code></em> argument. Users
are strongly encouraged to consult the 
<a href="\href{http://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf"><em><code class="highlighter-rouge">edgeR</code></em></a> vignette 
for more detailed information on this topic and how to properly run <em><code class="highlighter-rouge">edgeR</code></em> 
on data sets with more complex experimental designs.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">targets</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="n">targetspath</span><span class="p">,</span><span class="w"> </span><span class="n">comment</span><span class="o">=</span><span class="s2">"#"</span><span class="p">)</span><span class="w">
</span><span class="n">cmp</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">readComp</span><span class="p">(</span><span class="n">file</span><span class="o">=</span><span class="n">targetspath</span><span class="p">,</span><span class="w"> </span><span class="n">format</span><span class="o">=</span><span class="s2">"matrix"</span><span class="p">,</span><span class="w"> </span><span class="n">delim</span><span class="o">=</span><span class="s2">"-"</span><span class="p">)</span><span class="w">
</span><span class="n">cmp</span><span class="p">[[</span><span class="m">1</span><span class="p">]]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##       [,1]  [,2] 
##  [1,] "M1"  "A1" 
##  [2,] "M1"  "V1" 
##  [3,] "A1"  "V1" 
##  [4,] "M6"  "A6" 
##  [5,] "M6"  "V6" 
##  [6,] "A6"  "V6" 
##  [7,] "M12" "A12"
##  [8,] "M12" "V12"
##  [9,] "A12" "V12"
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">countDFeBygpath</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">system.file</span><span class="p">(</span><span class="s2">"extdata"</span><span class="p">,</span><span class="w"> </span><span class="s2">"countDFeByg.xls"</span><span class="p">,</span><span class="w"> </span><span class="n">package</span><span class="o">=</span><span class="s2">"systemPipeR"</span><span class="p">)</span><span class="w">
</span><span class="n">countDFeByg</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read.delim</span><span class="p">(</span><span class="n">countDFeBygpath</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">)</span><span class="w">
</span><span class="n">edgeDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">run_edgeR</span><span class="p">(</span><span class="n">countDF</span><span class="o">=</span><span class="n">countDFeByg</span><span class="p">,</span><span class="w"> </span><span class="n">targets</span><span class="o">=</span><span class="n">targets</span><span class="p">,</span><span class="w"> </span><span class="n">cmp</span><span class="o">=</span><span class="n">cmp</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span><span class="w"> </span><span class="n">independent</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">mdsplot</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## Disp = 0.20653 , BCV = 0.4545
</code></pre></div></div>

<p>Filter and plot DEG results for up and down regulated genes. Because of the small size of the toy data set used by this vignette, the <em>FDR</em> value has been set to a relatively high threshold (here 10%). More commonly used <em>FDR</em> cutoffs are 1% or 5%. The definition of ‘<em>up</em>’ and ‘<em>down</em>’ is given in the corresponding help file. To open it, type <em><code class="highlighter-rouge">?filterDEGs</code></em> in the R console.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DEG_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filterDEGs</span><span class="p">(</span><span class="n">degDF</span><span class="o">=</span><span class="n">edgeDF</span><span class="p">,</span><span class="w"> </span><span class="n">filter</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="n">Fold</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">FDR</span><span class="o">=</span><span class="m">10</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<p><img src="./pages/mydoc/systemPipeR_files/edger_deg_counts-1.png" width="100%" class="widefigure" /></p>
<div align="center"><b>Figure 4:</b> Up and down regulated DEGs identified by <i>edgeR</i>. </div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nf">names</span><span class="p">(</span><span class="n">DEG_list</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>## [1] "UporDown" "Up"       "Down"     "Summary"
</code></pre></div></div>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DEG_list</span><span class="o">$</span><span class="n">Summary</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,]</span><span class="w">
</span></code></pre></div></div>

<div class="highlighter-rouge"><div class="highlight"><pre class="highlight"><code>##       Comparisons Counts_Up_or_Down Counts_Up Counts_Down
## M1-A1       M1-A1                 0         0           0
## M1-V1       M1-V1                 1         1           0
## A1-V1       A1-V1                 1         1           0
## M6-A6       M6-A6                 0         0           0
</code></pre></div></div>

<h2 id="deg-analysis-with-deseq2">DEG analysis with <em><code class="highlighter-rouge">DESeq2</code></em></h2>
<p>The following <em><code class="highlighter-rouge">run_DESeq2</code></em> function is a convenience wrapper for
identifying DEGs in batch mode with <em><code class="highlighter-rouge">DESeq2</code></em> (Love et al., 2014) for any number of
pairwise sample comparisons specified under the <em><code class="highlighter-rouge">cmp</code></em> argument. Users
are strongly encouraged to consult the 
<a href="http://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.pdf"><em><code class="highlighter-rouge">DESeq2</code></em></a> vignette
for more detailed information on this topic and how to properly run <em><code class="highlighter-rouge">DESeq2</code></em> 
on data sets with more complex experimental designs.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">degseqDF</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">run_DESeq2</span><span class="p">(</span><span class="n">countDF</span><span class="o">=</span><span class="n">countDFeByg</span><span class="p">,</span><span class="w"> </span><span class="n">targets</span><span class="o">=</span><span class="n">targets</span><span class="p">,</span><span class="w"> </span><span class="n">cmp</span><span class="o">=</span><span class="n">cmp</span><span class="p">[[</span><span class="m">1</span><span class="p">]],</span><span class="w"> </span><span class="n">independent</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<p>Filter and plot DEG results for up and down regulated genes.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">DEG_list2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filterDEGs</span><span class="p">(</span><span class="n">degDF</span><span class="o">=</span><span class="n">degseqDF</span><span class="p">,</span><span class="w"> </span><span class="n">filter</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="n">Fold</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">FDR</span><span class="o">=</span><span class="m">10</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<p><img src="./pages/mydoc/systemPipeR_files/deseq2_deg_counts-1.png" width="100%" class="widefigure" /></p>
<div align="center"><b>Figure 5:</b> Up and down regulated DEGs identified by <i>DESeq2</i>. </div>

<h2 id="venn-diagrams">Venn Diagrams</h2>
<p>The function <em><code class="highlighter-rouge">overLapper</code></em> can compute Venn intersects for large numbers of sample sets (up to 20 or more) and <em><code class="highlighter-rouge">vennPlot</code></em> can plot 2-5 way Venn diagrams. A useful feature is the possiblity to combine the counts from several Venn comparisons with the same number of sample sets in a single Venn diagram (here for 4 up and down DEG sets).</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">vennsetup</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">overLapper</span><span class="p">(</span><span class="n">DEG_list</span><span class="o">$</span><span class="n">Up</span><span class="p">[</span><span class="m">6</span><span class="o">:</span><span class="m">9</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"vennsets"</span><span class="p">)</span><span class="w">
</span><span class="n">vennsetdown</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">overLapper</span><span class="p">(</span><span class="n">DEG_list</span><span class="o">$</span><span class="n">Down</span><span class="p">[</span><span class="m">6</span><span class="o">:</span><span class="m">9</span><span class="p">],</span><span class="w"> </span><span class="n">type</span><span class="o">=</span><span class="s2">"vennsets"</span><span class="p">)</span><span class="w">
</span><span class="n">vennPlot</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">vennsetup</span><span class="p">,</span><span class="w"> </span><span class="n">vennsetdown</span><span class="p">),</span><span class="w"> </span><span class="n">mymain</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">mysub</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">colmode</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">ccol</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"blue"</span><span class="p">,</span><span class="w"> </span><span class="s2">"red"</span><span class="p">))</span><span class="w">
</span></code></pre></div></div>

<p><img src="./pages/mydoc/systemPipeR_files/vennplot-1.png" width="100%" /></p>
<div align="center"><b>Figure 6:</b> Venn Diagram for 4 Up and Down DEG Sets. </div>

<h2 id="go-term-enrichment-analysis-of-degs">GO term enrichment analysis of DEGs</h2>
<h3 id="obtain-gene-to-go-mappings">Obtain gene-to-GO mappings</h3>
<p>The following shows how to obtain gene-to-GO mappings from <em><code class="highlighter-rouge">biomaRt</code></em> (here for <em>A. thaliana</em>) and how to organize them for the downstream GO term enrichment analysis. Alternatively, the gene-to-GO mappings can be obtained for many organisms from Bioconductor’s  <em><code class="highlighter-rouge">*.db</code></em> genome annotation packages or GO annotation files provided by various genome databases. For each annotation this relatively slow preprocessing step needs to be performed only once. Subsequently, the preprocessed data can be loaded with the <em><code class="highlighter-rouge">load</code></em> function as shown in the next subsection.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="s2">"biomaRt"</span><span class="p">)</span><span class="w">
</span><span class="n">listMarts</span><span class="p">()</span><span class="w"> </span><span class="c1"># To choose BioMart database</span><span class="w">
</span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">useMart</span><span class="p">(</span><span class="s2">"ENSEMBL_MART_PLANT"</span><span class="p">);</span><span class="w"> </span><span class="n">listDatasets</span><span class="p">(</span><span class="n">m</span><span class="p">)</span><span class="w"> 
</span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">useMart</span><span class="p">(</span><span class="s2">"ENSEMBL_MART_PLANT"</span><span class="p">,</span><span class="w"> </span><span class="n">dataset</span><span class="o">=</span><span class="s2">"athaliana_eg_gene"</span><span class="p">)</span><span class="w">
</span><span class="n">listAttributes</span><span class="p">(</span><span class="n">m</span><span class="p">)</span><span class="w"> </span><span class="c1"># Choose data types you want to download</span><span class="w">
</span><span class="n">go</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"go_accession"</span><span class="p">,</span><span class="w"> </span><span class="s2">"tair_locus"</span><span class="p">,</span><span class="w"> </span><span class="s2">"go_namespace_1003"</span><span class="p">),</span><span class="w"> </span><span class="n">mart</span><span class="o">=</span><span class="n">m</span><span class="p">)</span><span class="w">
</span><span class="n">go</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">go</span><span class="p">[</span><span class="n">go</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span><span class="o">!=</span><span class="s2">""</span><span class="p">,];</span><span class="w"> </span><span class="n">go</span><span class="p">[,</span><span class="m">3</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">go</span><span class="p">[,</span><span class="m">3</span><span class="p">])</span><span class="w">
</span><span class="n">dir.create</span><span class="p">(</span><span class="s2">"./data/GO"</span><span class="p">)</span><span class="w">
</span><span class="n">write.table</span><span class="p">(</span><span class="n">go</span><span class="p">,</span><span class="w"> </span><span class="s2">"data/GO/GOannotationsBiomart_mod.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">quote</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">row.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">col.names</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">"\t"</span><span class="p">)</span><span class="w">
</span><span class="n">catdb</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">makeCATdb</span><span class="p">(</span><span class="n">myfile</span><span class="o">=</span><span class="s2">"data/GO/GOannotationsBiomart_mod.txt"</span><span class="p">,</span><span class="w"> </span><span class="n">lib</span><span class="o">=</span><span class="kc">NULL</span><span class="p">,</span><span class="w"> </span><span class="n">org</span><span class="o">=</span><span class="s2">""</span><span class="p">,</span><span class="w"> </span><span class="n">colno</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="m">3</span><span class="p">),</span><span class="w"> </span><span class="n">idconv</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span><span class="n">save</span><span class="p">(</span><span class="n">catdb</span><span class="p">,</span><span class="w"> </span><span class="n">file</span><span class="o">=</span><span class="s2">"data/GO/catdb.RData"</span><span class="p">)</span><span class="w"> 
</span></code></pre></div></div>

<h3 id="batch-go-term-enrichment-analysis">Batch GO term enrichment analysis</h3>
<p>Apply the enrichment analysis to the DEG sets obtained in the above differential expression analysis. Note, in the following example the <em>FDR</em> filter is set here to an unreasonably high value, simply because of the small size of the toy data set used in this vignette. Batch enrichment analysis of many gene sets is performed with the <em><code class="highlighter-rouge">GOCluster_Report</code></em> function. When <em><code class="highlighter-rouge">method="all"</code></em>, it returns all GO terms passing the p-value cutoff specified under the <em><code class="highlighter-rouge">cutoff</code></em> arguments. When <em><code class="highlighter-rouge">method="slim"</code></em>, it returns only the GO terms specified under the <em><code class="highlighter-rouge">myslimv</code></em> argument. The given example shows how one can obtain such a GO slim vector from BioMart for a specific organism.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">load</span><span class="p">(</span><span class="s2">"data/GO/catdb.RData"</span><span class="p">)</span><span class="w">
</span><span class="n">DEG_list</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">filterDEGs</span><span class="p">(</span><span class="n">degDF</span><span class="o">=</span><span class="n">edgeDF</span><span class="p">,</span><span class="w"> </span><span class="n">filter</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="n">Fold</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">FDR</span><span class="o">=</span><span class="m">50</span><span class="p">),</span><span class="w"> </span><span class="n">plot</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span><span class="w">
</span><span class="n">up_down</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DEG_list</span><span class="o">$</span><span class="n">UporDown</span><span class="p">;</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">up_down</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">up_down</span><span class="p">),</span><span class="w"> </span><span class="s2">"_up_down"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w">
</span><span class="n">up</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DEG_list</span><span class="o">$</span><span class="n">Up</span><span class="p">;</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">up</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">up</span><span class="p">),</span><span class="w"> </span><span class="s2">"_up"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w">
</span><span class="n">down</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DEG_list</span><span class="o">$</span><span class="n">Down</span><span class="p">;</span><span class="w"> </span><span class="nf">names</span><span class="p">(</span><span class="n">down</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">paste</span><span class="p">(</span><span class="nf">names</span><span class="p">(</span><span class="n">down</span><span class="p">),</span><span class="w"> </span><span class="s2">"_down"</span><span class="p">,</span><span class="w"> </span><span class="n">sep</span><span class="o">=</span><span class="s2">""</span><span class="p">)</span><span class="w">
</span><span class="n">DEGlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="n">up_down</span><span class="p">,</span><span class="w"> </span><span class="n">up</span><span class="p">,</span><span class="w"> </span><span class="n">down</span><span class="p">)</span><span class="w">
</span><span class="n">DEGlist</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">DEGlist</span><span class="p">[</span><span class="n">sapply</span><span class="p">(</span><span class="n">DEGlist</span><span class="p">,</span><span class="w"> </span><span class="n">length</span><span class="p">)</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">0</span><span class="p">]</span><span class="w">
</span><span class="n">BatchResult</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GOCluster_Report</span><span class="p">(</span><span class="n">catdb</span><span class="o">=</span><span class="n">catdb</span><span class="p">,</span><span class="w"> </span><span class="n">setlist</span><span class="o">=</span><span class="n">DEGlist</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"all"</span><span class="p">,</span><span class="w"> </span><span class="n">id_type</span><span class="o">=</span><span class="s2">"gene"</span><span class="p">,</span><span class="w"> </span><span class="n">CLSZ</span><span class="o">=</span><span class="m">2</span><span class="p">,</span><span class="w"> </span><span class="n">cutoff</span><span class="o">=</span><span class="m">0.9</span><span class="p">,</span><span class="w"> </span><span class="n">gocats</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"MF"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BP"</span><span class="p">,</span><span class="w"> </span><span class="s2">"CC"</span><span class="p">),</span><span class="w"> </span><span class="n">recordSpecGO</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span><span class="n">library</span><span class="p">(</span><span class="s2">"biomaRt"</span><span class="p">);</span><span class="w"> </span><span class="n">m</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">useMart</span><span class="p">(</span><span class="s2">"ENSEMBL_MART_PLANT"</span><span class="p">,</span><span class="w"> </span><span class="n">dataset</span><span class="o">=</span><span class="s2">"athaliana_eg_gene"</span><span class="p">)</span><span class="w">
</span><span class="n">goslimvec</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="nf">as.character</span><span class="p">(</span><span class="n">getBM</span><span class="p">(</span><span class="n">attributes</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"goslim_goa_accession"</span><span class="p">),</span><span class="w"> </span><span class="n">mart</span><span class="o">=</span><span class="n">m</span><span class="p">)[,</span><span class="m">1</span><span class="p">])</span><span class="w">
</span><span class="n">BatchResultslim</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">GOCluster_Report</span><span class="p">(</span><span class="n">catdb</span><span class="o">=</span><span class="n">catdb</span><span class="p">,</span><span class="w"> </span><span class="n">setlist</span><span class="o">=</span><span class="n">DEGlist</span><span class="p">,</span><span class="w"> </span><span class="n">method</span><span class="o">=</span><span class="s2">"slim"</span><span class="p">,</span><span class="w"> </span><span class="n">id_type</span><span class="o">=</span><span class="s2">"gene"</span><span class="p">,</span><span class="w"> </span><span class="n">myslimv</span><span class="o">=</span><span class="n">goslimvec</span><span class="p">,</span><span class="w"> </span><span class="n">CLSZ</span><span class="o">=</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">cutoff</span><span class="o">=</span><span class="m">0.01</span><span class="p">,</span><span class="w"> </span><span class="n">gocats</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"MF"</span><span class="p">,</span><span class="w"> </span><span class="s2">"BP"</span><span class="p">,</span><span class="w"> </span><span class="s2">"CC"</span><span class="p">),</span><span class="w"> </span><span class="n">recordSpecGO</span><span class="o">=</span><span class="kc">NULL</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>

<h3 id="plot-batch-go-term-results">Plot batch GO term results</h3>
<p>The <em><code class="highlighter-rouge">data.frame</code></em> generated by <em><code class="highlighter-rouge">GOCluster_Report</code></em> can be plotted with the <em><code class="highlighter-rouge">goBarplot</code></em> function. Because of the variable size of the sample sets, it may not always be desirable to show the results from different DEG sets in the same bar plot. Plotting single sample sets is achieved by subsetting the input data frame as shown in the first line of the following example.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">gos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchResultslim</span><span class="p">[</span><span class="n">grep</span><span class="p">(</span><span class="s2">"M6-V6_up_down"</span><span class="p">,</span><span class="w"> </span><span class="n">BatchResultslim</span><span class="o">$</span><span class="n">CLID</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">gos</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">BatchResultslim</span><span class="w">
</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"GOslimbarplotMF.pdf"</span><span class="p">,</span><span class="w"> </span><span class="n">height</span><span class="o">=</span><span class="m">8</span><span class="p">,</span><span class="w"> </span><span class="n">width</span><span class="o">=</span><span class="m">10</span><span class="p">);</span><span class="w"> </span><span class="n">goBarplot</span><span class="p">(</span><span class="n">gos</span><span class="p">,</span><span class="w"> </span><span class="n">gocat</span><span class="o">=</span><span class="s2">"MF"</span><span class="p">);</span><span class="w"> </span><span class="n">dev.off</span><span class="p">()</span><span class="w">
</span><span class="n">goBarplot</span><span class="p">(</span><span class="n">gos</span><span class="p">,</span><span class="w"> </span><span class="n">gocat</span><span class="o">=</span><span class="s2">"BP"</span><span class="p">)</span><span class="w">
</span><span class="n">goBarplot</span><span class="p">(</span><span class="n">gos</span><span class="p">,</span><span class="w"> </span><span class="n">gocat</span><span class="o">=</span><span class="s2">"CC"</span><span class="p">)</span><span class="w">
</span></code></pre></div></div>
<p><img src="./pages/mydoc/systemPipeR_files/GOslimbarplotMF.png" alt="" /></p>
<div align="center"><b>Figure 7:</b> GO Slim Barplot for MF Ontology.</div>

<h2 id="clustering-and-heat-maps">Clustering and heat maps</h2>
<p>The following example performs hierarchical clustering on the <em><code class="highlighter-rouge">rlog</code></em> transformed expression matrix subsetted by the DEGs identified in the 
above differential expression analysis. It uses a Pearson correlation-based distance measure and complete linkage for cluster joining.</p>

<div class="language-r highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">pheatmap</span><span class="p">)</span><span class="w">
</span><span class="n">geneids</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">unique</span><span class="p">(</span><span class="nf">as.character</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">DEG_list</span><span class="p">[[</span><span class="m">1</span><span class="p">]])))</span><span class="w">
</span><span class="n">y</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">assay</span><span class="p">(</span><span class="n">rlog</span><span class="p">(</span><span class="n">dds</span><span class="p">))[</span><span class="n">geneids</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span><span class="n">pdf</span><span class="p">(</span><span class="s2">"heatmap1.pdf"</span><span class="p">)</span><span class="w">
</span><span class="n">pheatmap</span><span class="p">(</span><span class="n">y</span><span class="p">,</span><span class="w"> </span><span class="n">scale</span><span class="o">=</span><span class="s2">"row"</span><span class="p">,</span><span class="w"> </span><span class="n">clustering_distance_rows</span><span class="o">=</span><span class="s2">"correlation"</span><span class="p">,</span><span class="w"> </span><span class="n">clustering_distance_cols</span><span class="o">=</span><span class="s2">"correlation"</span><span class="p">)</span><span class="w">
</span><span class="n">dev.off</span><span class="p">()</span><span class="w">
</span></code></pre></div></div>
<p><img src="./pages/mydoc/systemPipeR_files/heatmap1.png" alt="" /></p>
<div align="center"><b>Figure 8:</b> Heat map with hierarchical clustering dendrograms of DEGs.</div>

<p><br /><br /><center><a href="mydoc_systemPipeR_2.html"><img src="images/left_arrow.png" alt="Previous page." /></a>Previous Page &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Next Page
<a href="mydoc_systemPipeR_4.html"><img src="images/right_arrow.png" alt="Next page." /></a></center></p>


    <div class="tags">
        
    </div>

    

</div>

<hr class="shaded"/>

<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2017 UC Riverside. All rights reserved. <br />
<span>Page last updated:</span> Mon Nov 13 18:33:55 2017<br/> Site last generated: Nov 13, 2017 <br />
<!--
<p><img src="images/company_logo.png" alt="Company logo"/></p>
-->
</div>
            </div>
</footer>


    </div>
    <!-- /.row -->
</div>
<!-- /.container -->
    </div>

</body>

</html>
