<!DOCTYPE html>
<head>
    <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="description" content="">
<meta name="keywords" content=" ">
<title>12. dplyr Environment | systemPipeR Website</title>
<link rel="stylesheet" href="css/syntax.css">


<link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">
<!--<link rel="stylesheet" type="text/css" href="css/bootstrap.min.css">-->
<link rel="stylesheet" href="css/modern-business.css">
<link rel="stylesheet" href="css/lavish-bootstrap.css">
<link rel="stylesheet" href="css/customstyles.css">
<link rel="stylesheet" href="css/theme-blue.css">

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-cookie/1.4.1/jquery.cookie.min.js"></script>
<script src="js/jquery.navgoco.min.js"></script>


<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.4/js/bootstrap.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/2.0.0/anchor.min.js"></script>
<script src="js/toc.js"></script>
<script src="js/customscripts.js"></script>

<script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

<link rel="shortcut icon" href="images/favicon.ico">

<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
<![endif]-->

<link rel="alternate" type="application/rss+xml" title="" href="http://localhost:4000feed.xml">

    <script>
        $(document).ready(function() {
            // Initialize navgoco with default options
            $("#mysidebar").navgoco({
                caretHtml: '',
                accordion: true,
                openClass: 'active', // open
                save: false, // leave false or nav highlighting doesn't work right
                cookie: {
                    name: 'navgoco',
                    expires: false,
                    path: '/'
                },
                slide: {
                    duration: 400,
                    easing: 'swing'
                }
            });

            $("#collapseAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', false);
            });

            $("#expandAll").click(function(e) {
                e.preventDefault();
                $("#mysidebar").navgoco('toggle', true);
            });

        });

    </script>
    <script>
        $(function () {
            $('[data-toggle="tooltip"]').tooltip()
        })
    </script>
    <script>
        $(document).ready(function() {
            $("#tg-sb-link").click(function() {
                $("#tg-sb-sidebar").toggle();
                $("#tg-sb-content").toggleClass('col-md-9');
                $("#tg-sb-content").toggleClass('col-md-12');
                $("#tg-sb-icon").toggleClass('fa-toggle-on');
                $("#tg-sb-icon").toggleClass('fa-toggle-off');
            });
        });
    </script>
    

</head>
<body>
<!-- Navigation -->
<nav class="navbar navbar-inverse navbar-fixed-top">
    <div class="container topnavlinks">
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a class="fa fa-home fa-lg navbar-brand" href="index.html">&nbsp;<span class="projectTitle"> systemPipeR Project Website</span></a>
        </div>
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav navbar-right">
                <!-- toggle sidebar button -->
                <li><a id="tg-sb-link" href="#"><i id="tg-sb-icon" class="fa fa-toggle-on"></i> Nav</a></li>
                <!-- entries without drop-downs appear here -->
                
                
                
                <li><a href="http://bioconductor.org/packages/release/bioc/html/systemPipeR.html" target="_blank">Bioc</a></li>
                
                
                
                <li><a href="https://github.com/tgirke/systemPipeR" target="_blank">GitHub Repo</a></li>
                
                
                
                <!-- entries with drop-downs appear here -->
                <!-- conditional logic to control which topnav appears for the audience defined in the configuration file.-->
                
                
                <li class="dropdown">
                    <a href="#" class="dropdown-toggle" data-toggle="dropdown">Resources<b class="caret"></b></a>
                    <ul class="dropdown-menu">
                        
                        
                        <li><a href="https://cran.r-project.org/" target="_blank">CRAN</a></li>
                        
                        
                        
                        <li><a href="http://bioconductor.org/" target="_blank">Bioconductor</a></li>
                        
                        
                        
                        <li><a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual" target="_blank">R & Bioconductor Manual</a></li>
                        
                        
                        
                        <li><a href="https://guides.github.com/activities/hello-world/" target="_blank">GitHub</a></li>
                        
                        
                        
                        <li><a href="http://hpcc.ucr.edu/" target="_blank">HPC Center (HPCC)</a></li>
                        
                        
                        
                        <li><a href="http://girke.bioinformatics.ucr.edu/" target="_blank">Girke Lab</a></li>
                        
                        
                        
                        <li><a href="http://idratherbewriting.com/documentation-theme-jekyll" target="_blank">Jekyll Doc Theme</a></li>
                        
                        
                    </ul>
                </li>
                
                
                
			<li>



  <a class="email" title="Submit feedback" href="#" onclick="javascript:window.location='mailto:thomas.girke@ucr.edu?subject=systemPipeR Website feedback&body=I have some feedback about the 12. dplyr Environment page: ' + window.location.href;"><i class="fa fa-envelope-o"></i> Feedback</a>

</li>

		
                <!--comment out this block if you want to hide search-->
                <li>
                    <!--start search-->
                    <div id="search-demo-container">
                        <input type="text" id="search-input" placeholder="search...">
                        <ul id="results-container"></ul>
                    </div>
                    <script src="js/jekyll-search.js" type="text/javascript"></script>
                    <script type="text/javascript">
                            SimpleJekyllSearch.init({
                                searchInput: document.getElementById('search-input'),
                                resultsContainer: document.getElementById('results-container'),
                                dataSource: 'search.json',
                                searchResultTemplate: '<li><a href="{url}" title="12. dplyr Environment">{title}</a></li>',
                    noResultsText: 'No results found.',
                            limit: 10,
                            fuzzy: true,
                    })
                    </script>
                    <!--end search-->
                </li>
            </ul>
        </div>
        </div>
        <!-- /.container -->
</nav>

<!-- Page Content -->
<div class="container">
    <div class="col-lg-12">&nbsp;</div>
    <!-- Content Row -->
    <div class="row">
        <!-- Sidebar Column -->
        <div class="col-md-3" id="tg-sb-sidebar">

          












<ul id="mysidebar" class="nav">
    <li class="sidebarTitle"><center><font color="#26466D" size="6">systemPipeR</font></center> <center><font color="#8b9299" size="3">...</font></center></li>
    
    
    
        
    
    <li>
        <a href="#">Home</a>
        <ul>
            
            
            
            <li><a href="index.html">Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_syllabus.html">Syllabus</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Internal</a>
        <ul>
            
            
            
            <li><a href="mydoc_internal.html">Internal Resources</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Schedule</a>
        <ul>
            
            
            
            <li><a href="mydoc_schedule.html">Course Schedule</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Lecture Slides</a>
        <ul>
            
            
            
            <li><a href="mydoc_slides_01.html">Course Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_02.html">Molbiol Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_03.html">Databases and Software</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_04.html">DNA Sequencing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_05.html">Introduction to R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_06.html">Sequence Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_07.html">Programming in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_08.html">Multiple Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_09.html">Short Read Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_10.html">NGS Analysis Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_11.html">Gene Expression Analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_12.html">NGS Workflows</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_13.html">ChIP-Seq Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_14.html">VAR-Seq Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_15.html">Sequence Assembly</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_16.html">Gene Annotation and Ontologies</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_17.html">Cluster Analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_18.html">Profile HMMs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_slides_19.html">Introduction to Phylogenetics</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Homework</a>
        <ul>
            
            
            
            <li><a href="mydoc_homework_01.html">HW1 - Online Tools</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_02.html">HW2 - Biocluster/Linux</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_03.html">HW3 - Introduction to R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_homework_04.html">HW4 - Pairwise Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_09.html">HW5 - Programming in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_08.html">HW6 - Sequence Analysis</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">GitHub</a>
        <ul>
            
            
            
            <li><a href="mydoc_tutorial_01.html">Table of Content</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#github-in-gen242">GitHub in GEN242</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#installing-git">Installation</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#git-basics-from-command-line">Git from Command-line</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#github-basics-from-command-line">GitHub from Command-line</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_01.html#using-github-from-rstudio">GitHub from RStudio</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Linux & HPCC Cluster</a>
        <ul>
            
            
            
            <li><a href="mydoc_tutorial_02.html">Table of Content</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#hpcc-cluster-overview">HPCC Cluster</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#linux-basics">Linux Basics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#software-and-module-system">Software & Module System</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#big-data-storage">Big Data Storage</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#queuing-system-slurm">Queueing System Slurm</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#textcode-editors">Text Editors</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#vim-r-tmux-essentials">Vim-R-Tmux</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_tutorial_02.html#nvim-r-tmux">Nvim-R-Tmux</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Introduction to R</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rbasics_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_02.html">2. R Package Repositories</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_03.html">3. Installation of R Packages</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_04.html">4. Getting Around</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_05.html">5. Basic Syntax</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_06.html">6. Data Types</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_07.html">7. Data Objects</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_08.html">8. Important Utilities</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_09.html">9. Operators and Calculations</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_10.html">10. Reading and Writing External Data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_11.html">11. Useful R Functions</a></li>
            
            
            
            
            
            
            <li class="active"><a href="mydoc_Rbasics_12.html">12. dplyr Environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_13.html">13. SQLite Databases</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_14.html">14. Graphics in R</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_15.html">15. Analysis Routine</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_16.html">16. R Markdown</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_17.html">17. Shiny Web Apps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_18.html">18. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rbasics_19.html">19. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Programming in R</a>
        <ul>
            
            
            
            <li><a href="mydoc_Programming_in_R_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_02.html">2. Control Structures</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_03.html">3. Loops</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_04.html">4. Functions</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_05.html">5. Useful Utilities</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_06.html">6. Running R Scripts</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_07.html">7. Building R Packages</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_08.html">8. Programming Exercises</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_09.html">9. Homework 5</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_10.html">10. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Programming_in_R_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Analysis Basics</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rsequences_01.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_02.html">2. Package Requirements</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_03.html">3. Strings in R Base</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_04.html">4. Sequences in Bioconductor</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_05.html">5. NGS Sequences</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_06.html">6. Range Operations</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_07.html">7. Transcript Ranges</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_08.html">8. Homework 6</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_09.html">9. Session Info</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rsequences_10.html">10. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">NGS Workflow Overview</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeR_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_2.html">2. Getting Started</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_3.html">3. Workflow overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_4.html">4. Workflow templates</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_5.html">5. Version information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeR_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">RNA-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_02.html">2. Samples and environment settings</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_05.html">5. Read quantification</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_06.html">6. Analysis of DEGs</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_07.html">7. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_08.html">8. Clustering and heat maps</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_09.html">9. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_10.html">10. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeRNAseq_11.html">11. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">ChIP-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_05.html">5. Utilities for coverage data</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_06.html">6. Peak calling with MACS2</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_07.html">7. Annotate peaks with genomic context</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_08.html">8. Count reads overlapping peaks</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_09.html">9. Differential binding analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_10.html">10. GO term enrichment analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_11.html">11. Motif analysis</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeChIPseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">VAR-Seq Workflow</a>
        <ul>
            
            
            
            <li><a href="mydoc_systemPipeVARseq_01.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_02.html">2. Workflow environment</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_03.html">3. Read preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_04.html">4. Alignments</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_05.html">5. Variant calling</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_06.html">6. Filter variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_07.html">7. Annotate filtered variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_08.html">8. Combine annotation results among samples</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_09.html">9. Summary statistics of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_10.html">10. Venn diagram of variants</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_11.html">11. Plot variants programmatically</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_12.html">12. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_13.html">13. Funding</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_systemPipeVARseq_14.html">14. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Graphics and Visualization</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rgraphics_1.html">1. Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_2.html">2. Base Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_3.html">3. Grid Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_4.html">4. lattice Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_5.html">5. ggplot2 Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_6.html">6. Specialty Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_7.html">7. Genome Graphics</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rgraphics_8.html">8. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Cluster Analysis</a>
        <ul>
            
            
            
            <li><a href="mydoc_Rclustering_1.html">1. Introduction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_2.html">2. Data Preprocessing</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_3.html">3. Clustering Algorithms</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_4.html">4. Clustering Exercises</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_5.html">5. Version Information</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_Rclustering_6.html">6. References</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Paper Presentations</a>
        <ul>
            
            
            
            <li><a href="mydoc_paper_presentations.html">Paper Presentations</a></li>
            
            
            
            
        </ul>
        
        
    
    <li>
        <a href="#">Course Projects</a>
        <ul>
            
            
            
            <li><a href="mydoc_project_intro.html">Overview</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_chipseq1.html">ChIP-Seq1 - Motif Prediction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_chipseq2.html">ChIP-Seq2 - Peak Callers</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_rnaseq1.html">RNA-Seq1 - Aligners</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_rnaseq2.html">RNA-Seq2 - DEG Methods</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_varseq1.html">VAR-Seq1 - Functional Prediction</a></li>
            
            
            
            
            
            
            <li><a href="mydoc_project_data.html">Project Data</a></li>
            
            
            
            
        </ul>
        
        
        
        <!-- if you aren't using the accordion, uncomment this block:
           <p class="external">
               <a href="#" id="collapseAll">Collapse All</a> | <a href="#" id="expandAll">Expand All</a>
           </p>
           -->
    </li>
</ul>
</div>

<!-- this highlights the active parent class in the navgoco sidebar. this is critical so that the parent expands when you're viewing a page. This must appear below the sidebar code above. Otherwise, if placed inside customscripts.js, the script runs before the sidebar code runs and the class never gets inserted.-->
<script>$("li.active").parents('li').toggleClass("active");</script>

    <!-- Content Column -->
    <div class="col-md-9" id="tg-sb-content">
        <div class="post-header">
   <h1 class="post-title-main">12. dplyr Environment</h1>
</div>



<div class="post-content">

   

    
    
<!-- this handles the automatic toc. use ## for subheads to auto-generate the on-page minitoc. if you use html tags, you must supply an ID for the heading element in order for it to appear in the minitoc. -->
<script>
$( document ).ready(function() {
  // Handler for .ready() called.

$('#toc').toc({ minimumHeaders: 0, listType: 'ul', showSpeed: 0, headers: 'h2,h3,h4' });

/* this offset helps account for the space taken up by the floating toolbar. */
$('#toc').on('click', 'a', function() {
  var target = $(this.getAttribute('href'))
    , scroll_target = target.offset().top

  $(window).scrollTop(scroll_target - 10);
  return false
})
  
});
</script>

<div id="toc"></div>

    


<!--
    

    <a target="_blank" href="https://github.com/tgirke/systemPipeR/blob/gh-pages/pages//mydoc_Rbasics_12.md" class="btn btn-default githubEditButton" role="button"><i class="fa fa-github fa-lg"></i> Edit me</a>
-->
    

  <p>Modern object classes and methods for handling <code class="highlighter-rouge">data.frame</code> like structures
are provided by the <code class="highlighter-rouge">dplyr</code> and <code class="highlighter-rouge">data.table</code> packages. The following gives a
short introduction to the usage and functionalities of the <code class="highlighter-rouge">dplyr</code> package. 
More detailed tutorials on this topic can be found here:</p>

<ul>
  <li><a href="https://rdrr.io/cran/dplyr/">dplyr: A Grammar of Data Manipulation</a></li>
  <li><a href="https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html">Introduction to <code class="highlighter-rouge">dplyr</code></a></li>
  <li><a href="http://genomicsclass.github.io/book/pages/dplyr_tutorial.html">Tutorial on <code class="highlighter-rouge">dplyr</code></a></li>
  <li><a href="http://stat545.com/bit001_dplyr-cheatsheet.html">Cheatsheet for Joins from Jenny Bryan</a></li>
  <li><a href="https://cran.r-project.org/web/packages/tibble/vignettes/tibble.html">Tibbles</a></li>
  <li><a href="https://www.r-bloggers.com/intro-to-the-data-table-package/">Intro to <code class="highlighter-rouge">data.table</code> package</a></li>
  <li><a href="https://www.r-bloggers.com/working-with-large-datasets-with-dplyr-and-data-table/">Big data with <code class="highlighter-rouge">dplyr</code> and <code class="highlighter-rouge">data.table</code></a></li>
  <li><a href="https://www.r-bloggers.com/fast-data-lookups-in-r-dplyr-vs-data-table/">Fast lookups with <code class="highlighter-rouge">dplyr</code> and <code class="highlighter-rouge">data.table</code></a></li>
</ul>

<h2 id="installation">Installation</h2>

<p>The <code class="highlighter-rouge">dplyr</code> environment has evolved into an ecosystem of packages. To simplify
package management, one can install and load the entire collection via the
<code class="highlighter-rouge">tidyverse</code> package. For more details on <code class="highlighter-rouge">tidyverse</code> see
<a href="http://tidyverse.org/">here</a>.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">install.packages</span><span class="p">(</span><span class="s2">"tidyverse"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<h2 id="construct-a-data-frame-tibble">Construct a <code class="highlighter-rouge">data frame</code> (<code class="highlighter-rouge">tibble</code>)</h2>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">tidyverse</span><span class="p">)</span><span class="w">
</span><span class="n">as_data_frame</span><span class="p">(</span><span class="n">iris</span><span class="p">)</span><span class="w"> </span><span class="c1"># coerce data.frame to data frame tbl
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;  &lt;fctr&gt;
## 1           5.1         3.5          1.4         0.2  setosa
## 2           4.9         3.0          1.4         0.2  setosa
## 3           4.7         3.2          1.3         0.2  setosa
## 4           4.6         3.1          1.5         0.2  setosa
## 5           5.0         3.6          1.4         0.2  setosa
## 6           5.4         3.9          1.7         0.4  setosa
## 7           4.6         3.4          1.4         0.3  setosa
## 8           5.0         3.4          1.5         0.2  setosa
## 9           4.4         2.9          1.4         0.2  setosa
## 10          4.9         3.1          1.5         0.1  setosa
## # ... with 140 more rows
</code></pre>
</div>

<p>Alternative functions producing the same result include <code class="highlighter-rouge">as_tibble</code> and <code class="highlighter-rouge">tbl_df</code>:</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">as_tibble</span><span class="p">(</span><span class="n">iris</span><span class="p">)</span><span class="w"> </span><span class="c1"># newer function provided by tibble package
</span><span class="n">tbl_df</span><span class="p">(</span><span class="n">iris</span><span class="p">)</span><span class="w"> </span><span class="c1"># this alternative exists for historical reasons
</span></code></pre>
</div>
<h2 id="reading-and-writing-tabular-files">Reading and writing tabular files</h2>

<p>While the base R read/write utilities can be used for <code class="highlighter-rouge">data frames</code>, best time
performance with the least amount of typing is achieved with the export/import
functions from the <code class="highlighter-rouge">readr</code> package. For very large files the <code class="highlighter-rouge">fread</code> function from 
the <code class="highlighter-rouge">data.table</code> package achieves the best time performance.</p>

<h3 id="import-with-readr">Import with <code class="highlighter-rouge">readr</code></h3>

<p>Import functions provided by <code class="highlighter-rouge">readr</code> include:</p>

<ul>
  <li><code class="highlighter-rouge">read_csv()</code>: comma separated (CSV) files</li>
  <li><code class="highlighter-rouge">read_tsv()</code>: tab separated files</li>
  <li><code class="highlighter-rouge">read_delim()</code>: general delimited files</li>
  <li><code class="highlighter-rouge">read_fwf()</code>: fixed width files</li>
  <li><code class="highlighter-rouge">read_table()</code>: tabular files where colums are separated by white-space.</li>
  <li><code class="highlighter-rouge">read_log()</code>: web log files</li>
</ul>

<p>Create a sample tab delimited file for import</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">write_tsv</span><span class="p">(</span><span class="n">iris</span><span class="p">,</span><span class="w"> </span><span class="s2">"iris.txt"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Creates sample file
</span></code></pre>
</div>

<p>Import with <code class="highlighter-rouge">read_tsv</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">read_tsv</span><span class="p">(</span><span class="s2">"iris.txt"</span><span class="p">)</span><span class="w"> </span><span class="c1"># Import with read_tbv from readr package
</span><span class="n">iris_df</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;
## 1           5.1         3.5          1.4         0.2  setosa
## 2           4.9         3.0          1.4         0.2  setosa
## 3           4.7         3.2          1.3         0.2  setosa
## 4           4.6         3.1          1.5         0.2  setosa
## 5           5.0         3.6          1.4         0.2  setosa
## 6           5.4         3.9          1.7         0.4  setosa
## 7           4.6         3.4          1.4         0.3  setosa
## 8           5.0         3.4          1.5         0.2  setosa
## 9           4.4         2.9          1.4         0.2  setosa
## 10          4.9         3.1          1.5         0.1  setosa
## # ... with 140 more rows
</code></pre>
</div>

<p>To import Google Sheets directly into R, see <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc_Rbasics_10.html">here</a>.</p>

<h3 id="fast-table-import-with-fread">Fast table import with <code class="highlighter-rouge">fread</code></h3>

<p>The <code class="highlighter-rouge">fread</code> function from the <code class="highlighter-rouge">data.table</code> package provides the best time performance for reading large
tabular files into R.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">library</span><span class="p">(</span><span class="n">data.table</span><span class="p">)</span><span class="w">
</span><span class="n">iris_df</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as_data_frame</span><span class="p">(</span><span class="n">fread</span><span class="p">(</span><span class="s2">"iris.txt"</span><span class="p">))</span><span class="w"> </span><span class="c1"># Import with fread and conversion to tibble
</span><span class="n">iris_df</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;
## 1           5.1         3.5          1.4         0.2  setosa
## 2           4.9         3.0          1.4         0.2  setosa
## 3           4.7         3.2          1.3         0.2  setosa
## 4           4.6         3.1          1.5         0.2  setosa
## 5           5.0         3.6          1.4         0.2  setosa
## 6           5.4         3.9          1.7         0.4  setosa
## 7           4.6         3.4          1.4         0.3  setosa
## 8           5.0         3.4          1.5         0.2  setosa
## 9           4.4         2.9          1.4         0.2  setosa
## 10          4.9         3.1          1.5         0.1  setosa
## # ... with 140 more rows
</code></pre>
</div>

<p>Note: to ignore lines starting with comment signs, one can pass on to <code class="highlighter-rouge">fread</code> a shell
command for preprocessing the file. The following example illustrates this option.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">fread</span><span class="p">(</span><span class="s2">"grep -v '^#' iris.txt"</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<h3 id="export-with-readr">Export with <code class="highlighter-rouge">readr</code></h3>

<p>Export function provided by <code class="highlighter-rouge">readr</code> inlcude</p>

<ul>
  <li><code class="highlighter-rouge">write_delim()</code>: general delimited files</li>
  <li><code class="highlighter-rouge">write_csv()</code>: comma separated (CSV) files</li>
  <li><code class="highlighter-rouge">write_excel_csv()</code>: excel style CSV files</li>
  <li><code class="highlighter-rouge">write_tsv()</code>: tab separated files</li>
</ul>

<p>For instance, the <code class="highlighter-rouge">write_tsv</code> function writes a <code class="highlighter-rouge">data frame</code> to a tab delimited file with much nicer
default settings than the base R <code class="highlighter-rouge">write.table</code> function.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">write_tsv</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="s2">"iris.txt"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<h2 id="column-and-row-binds">Column and row binds</h2>

<p>The equivalents to base R’s <code class="highlighter-rouge">rbind</code> and <code class="highlighter-rouge">cbind</code> are <code class="highlighter-rouge">bind_rows</code> and <code class="highlighter-rouge">bind_cols</code>, respectively.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">bind_cols</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">iris_df</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 10
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Sepal.Length Sepal.Width Petal.Length
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;        &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;
## 1           5.1         3.5          1.4         0.2  setosa          5.1         3.5          1.4
## 2           4.9         3.0          1.4         0.2  setosa          4.9         3.0          1.4
## 3           4.7         3.2          1.3         0.2  setosa          4.7         3.2          1.3
## 4           4.6         3.1          1.5         0.2  setosa          4.6         3.1          1.5
## 5           5.0         3.6          1.4         0.2  setosa          5.0         3.6          1.4
## 6           5.4         3.9          1.7         0.4  setosa          5.4         3.9          1.7
## 7           4.6         3.4          1.4         0.3  setosa          4.6         3.4          1.4
## 8           5.0         3.4          1.5         0.2  setosa          5.0         3.4          1.5
## 9           4.4         2.9          1.4         0.2  setosa          4.4         2.9          1.4
## 10          4.9         3.1          1.5         0.1  setosa          4.9         3.1          1.5
## # ... with 140 more rows, and 2 more variables: Petal.Width &lt;dbl&gt;, Species &lt;chr&gt;
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">bind_rows</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">iris_df</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 300 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;
## 1           5.1         3.5          1.4         0.2  setosa
## 2           4.9         3.0          1.4         0.2  setosa
## 3           4.7         3.2          1.3         0.2  setosa
## 4           4.6         3.1          1.5         0.2  setosa
## 5           5.0         3.6          1.4         0.2  setosa
## 6           5.4         3.9          1.7         0.4  setosa
## 7           4.6         3.4          1.4         0.3  setosa
## 8           5.0         3.4          1.5         0.2  setosa
## 9           4.4         2.9          1.4         0.2  setosa
## 10          4.9         3.1          1.5         0.1  setosa
## # ... with 290 more rows
</code></pre>
</div>

<h2 id="extract-column-as-vector">Extract column as vector</h2>

<p>The subsetting operators <code class="highlighter-rouge">[[</code> and <code class="highlighter-rouge">$</code>can be used to extract from a <code class="highlighter-rouge">data frame</code> single columns as vector.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="p">[[</span><span class="m">5</span><span class="p">]][</span><span class="m">1</span><span class="o">:</span><span class="m">12</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa"
## [11] "setosa" "setosa"
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="o">$</span><span class="n">Species</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">12</span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##  [1] "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa"
## [11] "setosa" "setosa"
</code></pre>
</div>

<h2 id="important-dplyr-functions">Important <code class="highlighter-rouge">dplyr</code> functions</h2>

<ol>
  <li><code class="highlighter-rouge">filter()</code> and <code class="highlighter-rouge">slice()</code></li>
  <li><code class="highlighter-rouge">arrange()</code></li>
  <li><code class="highlighter-rouge">select()</code> and <code class="highlighter-rouge">rename()</code></li>
  <li><code class="highlighter-rouge">distinct()</code></li>
  <li><code class="highlighter-rouge">mutate()</code> and <code class="highlighter-rouge">transmute()</code></li>
  <li><code class="highlighter-rouge">summarise()</code></li>
  <li><code class="highlighter-rouge">sample_n()</code> and <code class="highlighter-rouge">sample_frac()</code></li>
</ol>

<h2 id="slice-and-filter-functions">Slice and filter functions</h2>

<h3 id="filter-function">Filter function</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">filter</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">7.5</span><span class="p">,</span><span class="w"> </span><span class="n">Species</span><span class="o">==</span><span class="s2">"virginica"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 6 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width   Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;     &lt;chr&gt;
## 1          7.6         3.0          6.6         2.1 virginica
## 2          7.7         3.8          6.7         2.2 virginica
## 3          7.7         2.6          6.9         2.3 virginica
## 4          7.7         2.8          6.7         2.0 virginica
## 5          7.9         3.8          6.4         2.0 virginica
## 6          7.7         3.0          6.1         2.3 virginica
</code></pre>
</div>

<h3 id="base-r-code-equivalent">Base R code equivalent</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="p">[</span><span class="n">iris_df</span><span class="p">[,</span><span class="w"> </span><span class="s2">"Sepal.Length"</span><span class="p">]</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">7.5</span><span class="w"> </span><span class="o">&amp;</span><span class="w"> </span><span class="n">iris_df</span><span class="p">[,</span><span class="w"> </span><span class="s2">"Species"</span><span class="p">]</span><span class="o">==</span><span class="s2">"virginica"</span><span class="p">,</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 6 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width   Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;     &lt;chr&gt;
## 1          7.6         3.0          6.6         2.1 virginica
## 2          7.7         3.8          6.7         2.2 virginica
## 3          7.7         2.6          6.9         2.3 virginica
## 4          7.7         2.8          6.7         2.0 virginica
## 5          7.9         3.8          6.4         2.0 virginica
## 6          7.7         3.0          6.1         2.3 virginica
</code></pre>
</div>

<h3 id="including-boolean-operators">Including boolean operators</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">filter</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="m">7.5</span><span class="w"> </span><span class="o">|</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="w"> </span><span class="o">&lt;</span><span class="w"> </span><span class="m">5.5</span><span class="p">,</span><span class="w"> </span><span class="n">Species</span><span class="o">==</span><span class="s2">"virginica"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 7 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width   Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;     &lt;chr&gt;
## 1          7.6         3.0          6.6         2.1 virginica
## 2          4.9         2.5          4.5         1.7 virginica
## 3          7.7         3.8          6.7         2.2 virginica
## 4          7.7         2.6          6.9         2.3 virginica
## 5          7.7         2.8          6.7         2.0 virginica
## 6          7.9         3.8          6.4         2.0 virginica
## 7          7.7         3.0          6.1         2.3 virginica
</code></pre>
</div>

<h3 id="subset-rows-by-position">Subset rows by position</h3>

<p><code class="highlighter-rouge">dplyr</code> approach</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">slice</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 2 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
</code></pre>
</div>

<p>Base R code equivalent</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="p">[</span><span class="m">1</span><span class="o">:</span><span class="m">2</span><span class="p">,]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 2 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;
## 1          5.1         3.5          1.4         0.2  setosa
## 2          4.9         3.0          1.4         0.2  setosa
</code></pre>
</div>

<h3 id="subset-rows-by-names">Subset rows by names</h3>

<p>Since <code class="highlighter-rouge">data frames</code> do not contain row names, row wise subsetting via the <code class="highlighter-rouge">[,]</code> operator cannot be used.
However, the corresponding behavior can be achieved by passing to <code class="highlighter-rouge">select</code> a row position index 
obtained by basic R intersect utilities such as <code class="highlighter-rouge">match</code>.</p>

<p>Create a suitable test <code class="highlighter-rouge">data frame</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">df1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bind_cols</span><span class="p">(</span><span class="n">data_frame</span><span class="p">(</span><span class="n">ids1</span><span class="o">=</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"g"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">)),</span><span class="w"> </span><span class="n">as_data_frame</span><span class="p">(</span><span class="n">matrix</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">40</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">dimnames</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"CA"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">)))))</span><span class="w">
</span><span class="n">df1</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 10 × 5
##     ids1   CA1   CA2   CA3   CA4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1     g1     1    11    21    31
## 2     g2     2    12    22    32
## 3     g3     3    13    23    33
## 4     g4     4    14    24    34
## 5     g5     5    15    25    35
## 6     g6     6    16    26    36
## 7     g7     7    17    27    37
## 8     g8     8    18    28    38
## 9     g9     9    19    29    39
## 10   g10    10    20    30    40
</code></pre>
</div>

<p><code class="highlighter-rouge">dplyr</code> approach</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">slice</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span><span class="w"> </span><span class="n">match</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s2">"g10"</span><span class="p">,</span><span class="w"> </span><span class="s2">"g4"</span><span class="p">,</span><span class="w"> </span><span class="s2">"g4"</span><span class="p">),</span><span class="w"> </span><span class="n">df1</span><span class="o">$</span><span class="n">ids1</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 3 × 5
##    ids1   CA1   CA2   CA3   CA4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1   g10    10    20    30    40
## 2    g4     4    14    24    34
## 3    g4     4    14    24    34
</code></pre>
</div>

<p>Base R equivalent</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">df1_old</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">as.data.frame</span><span class="p">(</span><span class="n">df1</span><span class="p">)</span><span class="w">
</span><span class="n">rownames</span><span class="p">(</span><span class="n">df1_old</span><span class="p">)</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">df1_old</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span><span class="w">
</span><span class="n">df1_old</span><span class="p">[</span><span class="nf">c</span><span class="p">(</span><span class="s2">"g10"</span><span class="p">,</span><span class="w"> </span><span class="s2">"g4"</span><span class="p">,</span><span class="w"> </span><span class="s2">"g4"</span><span class="p">),]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>##      ids1 CA1 CA2 CA3 CA4
## g10   g10  10  20  30  40
## g4     g4   4  14  24  34
## g4.1   g4   4  14  24  34
</code></pre>
</div>

<h2 id="sorting-with-arrange">Sorting with <code class="highlighter-rouge">arrange</code></h2>

<p>Row-wise ordering based on specific columns</p>

<p><code class="highlighter-rouge">dplyr</code> approach</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">arrange</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Species</span><span class="p">,</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="p">,</span><span class="w"> </span><span class="n">Sepal.Width</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;
## 1           4.3         3.0          1.1         0.1  setosa
## 2           4.4         2.9          1.4         0.2  setosa
## 3           4.4         3.0          1.3         0.2  setosa
## 4           4.4         3.2          1.3         0.2  setosa
## 5           4.5         2.3          1.3         0.3  setosa
## 6           4.6         3.1          1.5         0.2  setosa
## 7           4.6         3.2          1.4         0.2  setosa
## 8           4.6         3.4          1.4         0.3  setosa
## 9           4.6         3.6          1.0         0.2  setosa
## 10          4.7         3.2          1.3         0.2  setosa
## # ... with 140 more rows
</code></pre>
</div>

<p>For ordering descendingly use <code class="highlighter-rouge">desc()</code> function</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">arrange</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">desc</span><span class="p">(</span><span class="n">Species</span><span class="p">),</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="p">,</span><span class="w"> </span><span class="n">Sepal.Width</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width   Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;     &lt;chr&gt;
## 1           4.9         2.5          4.5         1.7 virginica
## 2           5.6         2.8          4.9         2.0 virginica
## 3           5.7         2.5          5.0         2.0 virginica
## 4           5.8         2.7          5.1         1.9 virginica
## 5           5.8         2.7          5.1         1.9 virginica
## 6           5.8         2.8          5.1         2.4 virginica
## 7           5.9         3.0          5.1         1.8 virginica
## 8           6.0         2.2          5.0         1.5 virginica
## 9           6.0         3.0          4.8         1.8 virginica
## 10          6.1         2.6          5.6         1.4 virginica
## # ... with 140 more rows
</code></pre>
</div>

<p>Base R code equivalent</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">iris_df</span><span class="o">$</span><span class="n">Species</span><span class="p">,</span><span class="w"> </span><span class="n">iris_df</span><span class="o">$</span><span class="n">Sepal.Length</span><span class="p">,</span><span class="w"> </span><span class="n">iris_df</span><span class="o">$</span><span class="n">Sepal.Width</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;
## 1           4.3         3.0          1.1         0.1  setosa
## 2           4.4         2.9          1.4         0.2  setosa
## 3           4.4         3.0          1.3         0.2  setosa
## 4           4.4         3.2          1.3         0.2  setosa
## 5           4.5         2.3          1.3         0.3  setosa
## 6           4.6         3.1          1.5         0.2  setosa
## 7           4.6         3.2          1.4         0.2  setosa
## 8           4.6         3.4          1.4         0.3  setosa
## 9           4.6         3.6          1.0         0.2  setosa
## 10          4.7         3.2          1.3         0.2  setosa
## # ... with 140 more rows
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="p">[</span><span class="n">order</span><span class="p">(</span><span class="n">iris_df</span><span class="o">$</span><span class="n">Species</span><span class="p">,</span><span class="w"> </span><span class="n">decreasing</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">),</span><span class="w"> </span><span class="p">]</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width   Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;     &lt;chr&gt;
## 1           6.3         3.3          6.0         2.5 virginica
## 2           5.8         2.7          5.1         1.9 virginica
## 3           7.1         3.0          5.9         2.1 virginica
## 4           6.3         2.9          5.6         1.8 virginica
## 5           6.5         3.0          5.8         2.2 virginica
## 6           7.6         3.0          6.6         2.1 virginica
## 7           4.9         2.5          4.5         1.7 virginica
## 8           7.3         2.9          6.3         1.8 virginica
## 9           6.7         2.5          5.8         1.8 virginica
## 10          7.2         3.6          6.1         2.5 virginica
## # ... with 140 more rows
</code></pre>
</div>

<h2 id="select-columns-with-select">Select columns with <code class="highlighter-rouge">select</code></h2>

<p>Select specific columns</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">select</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Species</span><span class="p">,</span><span class="w"> </span><span class="n">Petal.Length</span><span class="p">,</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 3
##    Species Petal.Length Sepal.Length
##      &lt;chr&gt;        &lt;dbl&gt;        &lt;dbl&gt;
## 1   setosa          1.4          5.1
## 2   setosa          1.4          4.9
## 3   setosa          1.3          4.7
## 4   setosa          1.5          4.6
## 5   setosa          1.4          5.0
## 6   setosa          1.7          5.4
## 7   setosa          1.4          4.6
## 8   setosa          1.5          5.0
## 9   setosa          1.4          4.4
## 10  setosa          1.5          4.9
## # ... with 140 more rows
</code></pre>
</div>

<p>Select range of columns by name</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">select</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">Petal.Width</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 4
##    Sepal.Length Sepal.Width Petal.Length Petal.Width
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
## 1           5.1         3.5          1.4         0.2
## 2           4.9         3.0          1.4         0.2
## 3           4.7         3.2          1.3         0.2
## 4           4.6         3.1          1.5         0.2
## 5           5.0         3.6          1.4         0.2
## 6           5.4         3.9          1.7         0.4
## 7           4.6         3.4          1.4         0.3
## 8           5.0         3.4          1.5         0.2
## 9           4.4         2.9          1.4         0.2
## 10          4.9         3.1          1.5         0.1
## # ... with 140 more rows
</code></pre>
</div>

<p>Drop specific columns (here range)</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">select</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="o">-</span><span class="p">(</span><span class="n">Sepal.Length</span><span class="w"> </span><span class="o">:</span><span class="w"> </span><span class="n">Petal.Width</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 1
##    Species
##      &lt;chr&gt;
## 1   setosa
## 2   setosa
## 3   setosa
## 4   setosa
## 5   setosa
## 6   setosa
## 7   setosa
## 8   setosa
## 9   setosa
## 10  setosa
## # ... with 140 more rows
</code></pre>
</div>

<h2 id="renaming-columns-with-rename">Renaming columns with <code class="highlighter-rouge">rename</code></h2>

<p><code class="highlighter-rouge">dplyr</code> approach</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">rename</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">new_col_name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Species</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width new_col_name
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;        &lt;chr&gt;
## 1           5.1         3.5          1.4         0.2       setosa
## 2           4.9         3.0          1.4         0.2       setosa
## 3           4.7         3.2          1.3         0.2       setosa
## 4           4.6         3.1          1.5         0.2       setosa
## 5           5.0         3.6          1.4         0.2       setosa
## 6           5.4         3.9          1.7         0.4       setosa
## 7           4.6         3.4          1.4         0.3       setosa
## 8           5.0         3.4          1.5         0.2       setosa
## 9           4.4         2.9          1.4         0.2       setosa
## 10          4.9         3.1          1.5         0.1       setosa
## # ... with 140 more rows
</code></pre>
</div>

<p>Base R code approach</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">colnames</span><span class="p">(</span><span class="n">iris_df</span><span class="p">)[</span><span class="n">colnames</span><span class="p">(</span><span class="n">iris_df</span><span class="p">)</span><span class="o">==</span><span class="s2">"Species"</span><span class="p">]</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="s2">"new_col_names"</span><span class="w">
</span></code></pre>
</div>

<h2 id="obtain-unique-rows-with-distinct">Obtain unique rows with <code class="highlighter-rouge">distinct</code></h2>

<p><code class="highlighter-rouge">dplyr</code> approach</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">distinct</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Species</span><span class="p">,</span><span class="w"> </span><span class="n">.keep_all</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 3 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;      &lt;chr&gt;
## 1          5.1         3.5          1.4         0.2     setosa
## 2          7.0         3.2          4.7         1.4 versicolor
## 3          6.3         3.3          6.0         2.5  virginica
</code></pre>
</div>

<p>Base R code approach</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="p">[</span><span class="o">!</span><span class="n">duplicated</span><span class="p">(</span><span class="n">iris_df</span><span class="o">$</span><span class="n">Species</span><span class="p">),]</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 3 × 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;      &lt;chr&gt;
## 1          5.1         3.5          1.4         0.2     setosa
## 2          7.0         3.2          4.7         1.4 versicolor
## 3          6.3         3.3          6.0         2.5  virginica
</code></pre>
</div>

<h2 id="add-columns">Add columns</h2>

<h3 id="mutate"><code class="highlighter-rouge">mutate</code></h3>

<p>The <code class="highlighter-rouge">mutate</code> function allows to append columns to existing ones.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">mutate</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Ratio</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">Sepal.Width</span><span class="p">,</span><span class="w"> </span><span class="n">Sum</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">Sepal.Width</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 7
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species    Ratio   Sum
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;
## 1           5.1         3.5          1.4         0.2  setosa 1.457143   8.6
## 2           4.9         3.0          1.4         0.2  setosa 1.633333   7.9
## 3           4.7         3.2          1.3         0.2  setosa 1.468750   7.9
## 4           4.6         3.1          1.5         0.2  setosa 1.483871   7.7
## 5           5.0         3.6          1.4         0.2  setosa 1.388889   8.6
## 6           5.4         3.9          1.7         0.4  setosa 1.384615   9.3
## 7           4.6         3.4          1.4         0.3  setosa 1.352941   8.0
## 8           5.0         3.4          1.5         0.2  setosa 1.470588   8.4
## 9           4.4         2.9          1.4         0.2  setosa 1.517241   7.3
## 10          4.9         3.1          1.5         0.1  setosa 1.580645   8.0
## # ... with 140 more rows
</code></pre>
</div>

<h3 id="transmute"><code class="highlighter-rouge">transmute</code></h3>

<p>The <code class="highlighter-rouge">transmute</code> function does the same as <code class="highlighter-rouge">mutate</code> but drops existing columns</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">transmute</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Ratio</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="n">Sepal.Width</span><span class="p">,</span><span class="w"> </span><span class="n">Sum</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Sepal.Length</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">Sepal.Width</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 2
##       Ratio   Sum
##       &lt;dbl&gt; &lt;dbl&gt;
## 1  1.457143   8.6
## 2  1.633333   7.9
## 3  1.468750   7.9
## 4  1.483871   7.7
## 5  1.388889   8.6
## 6  1.384615   9.3
## 7  1.352941   8.0
## 8  1.470588   8.4
## 9  1.517241   7.3
## 10 1.580645   8.0
## # ... with 140 more rows
</code></pre>
</div>

<h3 id="bind_cols"><code class="highlighter-rouge">bind_cols</code></h3>

<p>The <code class="highlighter-rouge">bind_cols</code> function is the equivalent of <code class="highlighter-rouge">cbind</code> in base R. To add rows, use the corresponding 
<code class="highlighter-rouge">bind_rows</code> function.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">bind_cols</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">iris_df</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 150 × 10
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Sepal.Length Sepal.Width Petal.Length
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;        &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;
## 1           5.1         3.5          1.4         0.2  setosa          5.1         3.5          1.4
## 2           4.9         3.0          1.4         0.2  setosa          4.9         3.0          1.4
## 3           4.7         3.2          1.3         0.2  setosa          4.7         3.2          1.3
## 4           4.6         3.1          1.5         0.2  setosa          4.6         3.1          1.5
## 5           5.0         3.6          1.4         0.2  setosa          5.0         3.6          1.4
## 6           5.4         3.9          1.7         0.4  setosa          5.4         3.9          1.7
## 7           4.6         3.4          1.4         0.3  setosa          4.6         3.4          1.4
## 8           5.0         3.4          1.5         0.2  setosa          5.0         3.4          1.5
## 9           4.4         2.9          1.4         0.2  setosa          4.4         2.9          1.4
## 10          4.9         3.1          1.5         0.1  setosa          4.9         3.1          1.5
## # ... with 140 more rows, and 2 more variables: Petal.Width &lt;dbl&gt;, Species &lt;chr&gt;
</code></pre>
</div>

<h2 id="summarize-data">Summarize data</h2>

<p>Summary calculation on single column</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">summarize</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">mean</span><span class="p">(</span><span class="n">Petal.Length</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 1 × 1
##   `mean(Petal.Length)`
##                  &lt;dbl&gt;
## 1                3.758
</code></pre>
</div>

<p>Summary calculation on many columns</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">summarize_all</span><span class="p">(</span><span class="n">iris_df</span><span class="p">[,</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">],</span><span class="w"> </span><span class="n">mean</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 1 × 4
##   Sepal.Length Sepal.Width Petal.Length Petal.Width
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
## 1     5.843333    3.057333        3.758    1.199333
</code></pre>
</div>

<p>Summarize by grouping column</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">summarize</span><span class="p">(</span><span class="n">group_by</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Species</span><span class="p">),</span><span class="w"> </span><span class="n">mean</span><span class="p">(</span><span class="n">Petal.Length</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 3 × 2
##      Species `mean(Petal.Length)`
##        &lt;chr&gt;                &lt;dbl&gt;
## 1     setosa                1.462
## 2 versicolor                4.260
## 3  virginica                5.552
</code></pre>
</div>

<p>Aggregate summaries</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">summarize_all</span><span class="p">(</span><span class="n">group_by</span><span class="p">(</span><span class="n">iris_df</span><span class="p">,</span><span class="w"> </span><span class="n">Species</span><span class="p">),</span><span class="w"> </span><span class="n">mean</span><span class="p">)</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 3 × 5
##      Species Sepal.Length Sepal.Width Petal.Length Petal.Width
##        &lt;chr&gt;        &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
## 1     setosa        5.006       3.428        1.462       0.246
## 2 versicolor        5.936       2.770        4.260       1.326
## 3  virginica        6.588       2.974        5.552       2.026
</code></pre>
</div>

<p>Note: <code class="highlighter-rouge">group_by</code> does the looping for the user similar to <code class="highlighter-rouge">aggregate</code> or <code class="highlighter-rouge">tapply</code>.</p>

<h2 id="merging-data-frames">Merging data frames</h2>

<p>The <code class="highlighter-rouge">dplyr</code> package provides several join functions for merging <code class="highlighter-rouge">data frames</code> by a common key column
similar to the <code class="highlighter-rouge">merge</code> function in base R. These <code class="highlighter-rouge">*_join</code> functions include:</p>

<ul>
  <li><code class="highlighter-rouge">inner_join()</code>: returns join only for rows matching among both <code class="highlighter-rouge">data tables</code></li>
  <li><code class="highlighter-rouge">full_join()</code>: returns join for all (matching and non-matching) rows of two <code class="highlighter-rouge">data tables</code></li>
  <li><code class="highlighter-rouge">left_join()</code>: returns join for all rows in first <code class="highlighter-rouge">data table</code></li>
  <li><code class="highlighter-rouge">right_join()</code>: returns join for all rows in second <code class="highlighter-rouge">data table</code></li>
  <li><code class="highlighter-rouge">anti_join()</code>: returns for first <code class="highlighter-rouge">data table</code> only those rows that have no match in the second one</li>
</ul>

<p>Sample <code class="highlighter-rouge">data frames</code> to illustrate <code class="highlighter-rouge">*.join</code> functions.</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">df1</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bind_cols</span><span class="p">(</span><span class="n">data_frame</span><span class="p">(</span><span class="n">ids1</span><span class="o">=</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"g"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">)),</span><span class="w"> </span><span class="n">as_data_frame</span><span class="p">(</span><span class="n">matrix</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">40</span><span class="p">,</span><span class="w"> </span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">dimnames</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">10</span><span class="p">,</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"CA"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">)))))</span><span class="w">
</span><span class="n">df1</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 10 × 5
##     ids1   CA1   CA2   CA3   CA4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1     g1     1    11    21    31
## 2     g2     2    12    22    32
## 3     g3     3    13    23    33
## 4     g4     4    14    24    34
## 5     g5     5    15    25    35
## 6     g6     6    16    26    36
## 7     g7     7    17    27    37
## 8     g8     8    18    28    38
## 9     g9     9    19    29    39
## 10   g10    10    20    30    40
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">df2</span><span class="w"> </span><span class="o">&lt;-</span><span class="w"> </span><span class="n">bind_cols</span><span class="p">(</span><span class="n">data_frame</span><span class="p">(</span><span class="n">ids2</span><span class="o">=</span><span class="n">paste0</span><span class="p">(</span><span class="s2">"g"</span><span class="p">,</span><span class="w"> </span><span class="nf">c</span><span class="p">(</span><span class="m">2</span><span class="p">,</span><span class="m">5</span><span class="p">,</span><span class="m">11</span><span class="p">,</span><span class="m">12</span><span class="p">))),</span><span class="w"> </span><span class="n">as_data_frame</span><span class="p">(</span><span class="n">matrix</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">16</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">dimnames</span><span class="o">=</span><span class="nf">list</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">,</span><span class="w"> </span><span class="n">paste0</span><span class="p">(</span><span class="s2">"CB"</span><span class="p">,</span><span class="w"> </span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">)))))</span><span class="w">
</span><span class="n">df2</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 4 × 5
##    ids2   CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1    g2     1     5     9    13
## 2    g5     2     6    10    14
## 3   g11     3     7    11    15
## 4   g12     4     8    12    16
</code></pre>
</div>
<h3 id="inner-join">Inner join</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">inner_join</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span><span class="w"> </span><span class="n">df2</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ids1"</span><span class="o">=</span><span class="s2">"ids2"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 2 × 9
##    ids1   CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1    g2     2    12    22    32     1     5     9    13
## 2    g5     5    15    25    35     2     6    10    14
</code></pre>
</div>

<h3 id="left-join">Left join</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">left_join</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span><span class="w"> </span><span class="n">df2</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ids1"</span><span class="o">=</span><span class="s2">"ids2"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 10 × 9
##     ids1   CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1     g1     1    11    21    31    NA    NA    NA    NA
## 2     g2     2    12    22    32     1     5     9    13
## 3     g3     3    13    23    33    NA    NA    NA    NA
## 4     g4     4    14    24    34    NA    NA    NA    NA
## 5     g5     5    15    25    35     2     6    10    14
## 6     g6     6    16    26    36    NA    NA    NA    NA
## 7     g7     7    17    27    37    NA    NA    NA    NA
## 8     g8     8    18    28    38    NA    NA    NA    NA
## 9     g9     9    19    29    39    NA    NA    NA    NA
## 10   g10    10    20    30    40    NA    NA    NA    NA
</code></pre>
</div>

<h3 id="right-join">Right join</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">right_join</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span><span class="w"> </span><span class="n">df2</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ids1"</span><span class="o">=</span><span class="s2">"ids2"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 4 × 9
##    ids1   CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1    g2     2    12    22    32     1     5     9    13
## 2    g5     5    15    25    35     2     6    10    14
## 3   g11    NA    NA    NA    NA     3     7    11    15
## 4   g12    NA    NA    NA    NA     4     8    12    16
</code></pre>
</div>

<h3 id="full-join">Full join</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">full_join</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span><span class="w"> </span><span class="n">df2</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ids1"</span><span class="o">=</span><span class="s2">"ids2"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 12 × 9
##     ids1   CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1     g1     1    11    21    31    NA    NA    NA    NA
## 2     g2     2    12    22    32     1     5     9    13
## 3     g3     3    13    23    33    NA    NA    NA    NA
## 4     g4     4    14    24    34    NA    NA    NA    NA
## 5     g5     5    15    25    35     2     6    10    14
## 6     g6     6    16    26    36    NA    NA    NA    NA
## 7     g7     7    17    27    37    NA    NA    NA    NA
## 8     g8     8    18    28    38    NA    NA    NA    NA
## 9     g9     9    19    29    39    NA    NA    NA    NA
## 10   g10    10    20    30    40    NA    NA    NA    NA
## 11   g11    NA    NA    NA    NA     3     7    11    15
## 12   g12    NA    NA    NA    NA     4     8    12    16
</code></pre>
</div>

<h3 id="anti-join">Anti join</h3>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">anti_join</span><span class="p">(</span><span class="n">df1</span><span class="p">,</span><span class="w"> </span><span class="n">df2</span><span class="p">,</span><span class="w"> </span><span class="n">by</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"ids1"</span><span class="o">=</span><span class="s2">"ids2"</span><span class="p">))</span><span class="w">
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 8 × 5
##    ids1   CA1   CA2   CA3   CA4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1   g10    10    20    30    40
## 2    g9     9    19    29    39
## 3    g8     8    18    28    38
## 4    g7     7    17    27    37
## 5    g6     6    16    26    36
## 6    g4     4    14    24    34
## 7    g3     3    13    23    33
## 8    g1     1    11    21    31
</code></pre>
</div>

<p>For additional join options users want to cosult the <code class="highlighter-rouge">*_join</code> help pages.</p>

<h2 id="chaining">Chaining</h2>

<p>To simplify chaining of serveral operations, <code class="highlighter-rouge">dplyr</code> provides the <code class="highlighter-rouge">%&gt;%</code>
operator. where <code class="highlighter-rouge">x %&gt;% f(y)</code> turns into <code class="highlighter-rouge">f(x, y)</code>. This way one can pipe
together multiple operations by writing them from left-to-right or
top-to-bottom. This makes for easy to type and readable code.</p>

<h3 id="example-1">Example 1</h3>

<p>Series of data manipulations and export</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="c1"># Declare data frame to use 
</span><span class="w">    </span><span class="n">select</span><span class="p">(</span><span class="n">Sepal.Length</span><span class="o">:</span><span class="n">Species</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="c1"># Select columns
</span><span class="w">    </span><span class="n">filter</span><span class="p">(</span><span class="n">Species</span><span class="o">==</span><span class="s2">"setosa"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="c1"># Filter rows by some value
</span><span class="w">    </span><span class="n">arrange</span><span class="p">(</span><span class="n">Sepal.Length</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="c1"># Sort by some column
</span><span class="w">    </span><span class="n">mutate</span><span class="p">(</span><span class="n">Subtract</span><span class="o">=</span><span class="n">Petal.Length</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">Petal.Width</span><span class="p">)</span><span class="w"> </span><span class="c1"># Calculate and append
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 50 × 6
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Subtract
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;   &lt;chr&gt;    &lt;dbl&gt;
## 1           4.3         3.0          1.1         0.1  setosa      1.0
## 2           4.4         2.9          1.4         0.2  setosa      1.2
## 3           4.4         3.0          1.3         0.2  setosa      1.1
## 4           4.4         3.2          1.3         0.2  setosa      1.1
## 5           4.5         2.3          1.3         0.3  setosa      1.0
## 6           4.6         3.1          1.5         0.2  setosa      1.3
## 7           4.6         3.4          1.4         0.3  setosa      1.1
## 8           4.6         3.6          1.0         0.2  setosa      0.8
## 9           4.6         3.2          1.4         0.2  setosa      1.2
## 10          4.7         3.2          1.3         0.2  setosa      1.1
## # ... with 40 more rows
</code></pre>
</div>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="w">    </span><span class="c1"># write_tsv("iris.txt") # Export to file, omitted here to show result 
</span></code></pre>
</div>

<h3 id="example-2">Example 2</h3>

<p>Series of summary calculations for grouped data (<code class="highlighter-rouge">group_by</code>)</p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="c1"># Declare data frame to use 
</span><span class="w">    </span><span class="n">group_by</span><span class="p">(</span><span class="n">Species</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> </span><span class="c1"># Group by species
</span><span class="w">    </span><span class="n">summarize</span><span class="p">(</span><span class="n">Mean_Sepal.Length</span><span class="o">=</span><span class="n">mean</span><span class="p">(</span><span class="n">Sepal.Length</span><span class="p">),</span><span class="w"> 
              </span><span class="n">Max_Sepal.Length</span><span class="o">=</span><span class="nf">max</span><span class="p">(</span><span class="n">Sepal.Length</span><span class="p">),</span><span class="w">
              </span><span class="n">Min_Sepal.Length</span><span class="o">=</span><span class="nf">min</span><span class="p">(</span><span class="n">Sepal.Length</span><span class="p">),</span><span class="w">
              </span><span class="n">SD_Sepal.Length</span><span class="o">=</span><span class="n">sd</span><span class="p">(</span><span class="n">Sepal.Length</span><span class="p">),</span><span class="w">
              </span><span class="n">Total</span><span class="o">=</span><span class="n">n</span><span class="p">())</span><span class="w"> 
</span></code></pre>
</div>

<div class="highlighter-rouge"><pre class="highlight"><code>## # A tibble: 3 × 6
##      Species Mean_Sepal.Length Max_Sepal.Length Min_Sepal.Length SD_Sepal.Length Total
##        &lt;chr&gt;             &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt; &lt;int&gt;
## 1     setosa             5.006              5.8              4.3       0.3524897    50
## 2 versicolor             5.936              7.0              4.9       0.5161711    50
## 3  virginica             6.588              7.9              4.9       0.6358796    50
</code></pre>
</div>

<h3 id="example-3">Example 3</h3>

<p>Combining <code class="highlighter-rouge">dplyr</code> chaining with <code class="highlighter-rouge">ggplot</code></p>

<div class="language-r highlighter-rouge"><pre class="highlight"><code><span class="n">iris_df</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
    </span><span class="n">group_by</span><span class="p">(</span><span class="n">Species</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
    </span><span class="n">summarize_all</span><span class="p">(</span><span class="n">mean</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w"> 
    </span><span class="n">reshape2</span><span class="o">::</span><span class="n">melt</span><span class="p">(</span><span class="n">id.vars</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s2">"Species"</span><span class="p">),</span><span class="w"> </span><span class="n">variable.name</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s2">"Samples"</span><span class="p">,</span><span class="w"> </span><span class="n">value.name</span><span class="o">=</span><span class="s2">"Values"</span><span class="p">)</span><span class="w"> </span><span class="o">%&gt;%</span><span class="w">
    </span><span class="n">ggplot</span><span class="p">(</span><span class="n">aes</span><span class="p">(</span><span class="n">Samples</span><span class="p">,</span><span class="w"> </span><span class="n">Values</span><span class="p">,</span><span class="w"> </span><span class="n">fill</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">Species</span><span class="p">))</span><span class="w"> </span><span class="o">+</span><span class="w"> 
        </span><span class="n">geom_bar</span><span class="p">(</span><span class="n">position</span><span class="o">=</span><span class="s2">"dodge"</span><span class="p">,</span><span class="w"> </span><span class="n">stat</span><span class="o">=</span><span class="s2">"identity"</span><span class="p">)</span><span class="w">
</span></code></pre>
</div>

<p><img src="./pages/mydoc/Rbasics_files/plyr_chaining3-1.png" width="672" /></p>

<p><br /><br /><center><a href="mydoc_Rbasics_11.html"><img src="images/left_arrow.png" alt="Previous page." /></a>Previous Page &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Next Page
<a href="mydoc_Rbasics_13.html"><img src="images/right_arrow.png" alt="Next page." /></a></center></p>


    <div class="tags">
        
    </div>

    

</div>

<hr class="shaded"/>

<footer>
            <div class="row">
                <div class="col-lg-12 footer">
               &copy;2017 UC Riverside. All rights reserved. <br />
<span>Page last updated:</span> Sun Jun 25 17:41:27 2017<br/> Site last generated: Oct 8, 2017 <br />
<!--
<p><img src="images/company_logo.png" alt="Company logo"/></p>
-->
</div>
            </div>
</footer>


    </div>
    <!-- /.row -->
</div>
<!-- /.container -->
    </div>

</body>

</html>
