---
title: 14. References
last_updated: Mon Nov 13 16:23:52 2017
sidebar: mydoc_sidebar
permalink: mydoc_systemPipeChIPseq_14.html
---
 
1. H Backman, Tyler W, Girke, Thomas (2016) systemPipeR: NGS workflow and report generation environment. BMC Bioinformatics, 17: 388; [URL](http://dx.doi.org/10.1186/s12859-016-1241-0)
2. Langmead, Ben, Salzberg, Steven L (2012) Fast gapped-read alignment with Bowtie 2. Nat. Methods, 9: 357--359; [URL](http://dx.doi.org/10.1038/nmeth.1923)
3. Love, Michael, Huber, Wolfgang, Anders, Simon (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol., 15: 550; [URL](http://genomebiology.com/2014/15/12/550)
4. Robinson, M D, McCarthy, D J, Smyth, G K (2010) edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 26: 139--140; [URL](http://dx.doi.org/10.1093/bioinformatics/btp616)
5. Yu, Guangchuang, Wang, Li-Gen, He, Qing-Yu (2015) ChIPseeker: an R/Bioconductor package for ChIP peak annotation, comparison and visualization. Bioinformatics, 31: 2382--2383; [URL](http://dx.doi.org/10.1093/bioinformatics/btv145)
6. Zhang, Y, Liu, T, Meyer, C A, Eeckhoute, J, Johnson, D S, Bernstein, B E, Nussbaum, C, Myers, R M, Brown, M, Li, W, Liu, X S (2008) Model-based analysis of ChIP-Seq (MACS). Genome Biol., 9: NA; [URL](http://dx.doi.org/10.1186/gb-2008-9-9-r137)
7. Zhu, Lihua J, Gazin, Claude, Lawson, Nathan D, Pag\`{e}s, Herv\'{e}, Lin, Simon M, Lapointe, David S, Green, Michael R (2010) ChIPpeakAnno: a Bioconductor package to annotate ChIP-seq and ChIP-chip data. BMC Bioinformatics, 11: 237; [URL](http://dx.doi.org/10.1186/1471-2105-11-237)
<br><br><center><a href="mydoc_systemPipeChIPseq_13.html"><img src="images/left_arrow.png" alt="Previous page."></a>Previous Page &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Next Page
<a href="mydoc_systemPipeChIPseq_01.html"><img src="images/right_arrow.png" alt="Next page."></a></center>
