---
title: 8. References
last_updated: Thu Jan 28 14:24:39 2021
sidebar: mydoc_sidebar
permalink: mydoc_systemPipeR_8.html
---
 
1. H Backman, Tyler W, Girke, Thomas (2016) systemPipeR: NGS workflow and report generation environment. BMC Bioinformatics, 17: 388; [URL](http://dx.doi.org/10.1186/s12859-016-1241-0)
2. Howard, Brian E, Hu, Qiwen, Babaoglu, Ahmet Can, Chandra, Manan, Borghi, Monica, Tan, Xiaoping, He, Luyan, Winter-Sederoff, Heike, Gassmann, Walter, Veronese, Paola, Heber, Steffen (2013) High-throughput RNA sequencing of pseudomonas-infected Arabidopsis reveals hidden transcriptome complexity and novel splice variants. PLoS One, 8: e74183; [URL](http://dx.doi.org/10.1371/journal.pone.0074183)
3. Kim, Daehwan, Langmead, Ben, Salzberg, Steven L (2015) HISAT: a fast spliced aligner with low memory requirements. Nat. Methods, 12: 357--360; [URL](NA)
4. Kim, Daehwan, Pertea, Geo, Trapnell, Cole, Pimentel, Harold, Kelley, Ryan, Salzberg, Steven L (2013) TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions. Genome Biol., 14: R36; [URL](http://dx.doi.org/10.1186/gb-2013-14-4-r36)
5. Langmead, Ben, Salzberg, Steven L (2012) Fast gapped-read alignment with Bowtie 2. Nat. Methods, 9: 357--359; [URL](http://dx.doi.org/10.1038/nmeth.1923)
6. Lawrence, Michael, Huber, Wolfgang, Pag\`{e}s, Herv\'{e}, Aboyoun, Patrick, Carlson, Marc, Gentleman, Robert, Morgan, Martin T, Carey, Vincent J (2013) Software for computing and annotating genomic ranges. PLoS Comput. Biol., 9: e1003118; [URL](http://dx.doi.org/10.1371/journal.pcbi.1003118)
7. Li, H, Durbin, R (2009) Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25: 1754--1760; [URL](http://dx.doi.org/10.1093/bioinformatics/btp324)
8. Li, Heng (2013) Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM. arXiv [q-bio.GN], NA: NA; [URL](http://arxiv.org/abs/1303.3997)
9. Liao, Yang, Smyth, Gordon K, Shi, Wei (2013) The Subread aligner: fast, accurate and scalable read mapping by seed-and-vote. Nucleic Acids Res., 41: e108; [URL](http://dx.doi.org/10.1093/nar/gkt214)
10. Love, Michael, Huber, Wolfgang, Anders, Simon (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biol., 15: 550; [URL](http://genomebiology.com/2014/15/12/550)
11. Martin Morgan, Herv\'e Pag\`es, Valerie Obenchain, Nathaniel Hayden (2019) Rsamtools: Binary alignment (BAM), FASTA, variant call (BCF), and tabix file import. NA, NA: NA; [URL](http://bioconductor.org/packages/Rsamtools)
12. Robinson, M D, McCarthy, D J, Smyth, G K (2010) edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics, 26: 139--140; [URL](http://dx.doi.org/10.1093/bioinformatics/btp616)
13. Wu, T D, Nacu, S (2010) Fast and SNP-tolerant detection of complex variants and splicing in short reads. Bioinformatics, 26: 873--881; [URL](http://dx.doi.org/10.1093/bioinformatics/btq057)
<br><br><center><a href="mydoc_systemPipeR_7.html"><img src="images/left_arrow.png" alt="Previous page."></a>Previous Page &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Next Page
<a href="mydoc_systemPipeR_1.html"><img src="images/right_arrow.png" alt="Next page."></a></center>
